{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h1>Docking pipeline</h1>\n",
    "This notebook performs the following tasks:\n",
    "<ol>\n",
    "<li>Installing and importing necessary modules and packages, setting environments, and giving permissions to scripts and license. </li>\n",
    "<li>Read a lit of PDB IDs and download the PDB file to the current directory using a Biopython module.</li>\n",
    "<li>Ligand extraction from a PDB file, using OpenEye (OE), which requires the conversion to an OE Design Unit before writing out the ligand in PDB format.</li>\n",
    "<li>Receptor (protein) extraction using a customisable Biopython module, which excludes heteroatoms and water. This writes out the receptor in PDB format.</li>\n",
    "<li><font color='blue'>(Optional and not currently used)</font> Center of Mass (COM) and Center of Geometry (COG) calculation for the ligand. This is usually used for defining a docking search grid but not required for a score-only vina process using the --autobox flag within vina.</li>\n",
    "<li>Docking file preparation of the extracted PDB ligand and receptor, for AutoDock Vina, which saves the prepped files in PDBQT format. </li>\n",
    "<li>Scoring using Autodock Vina, using the --autobox and the --score-only flags. This means that no ligand poses are returned. An output log is saved in OUT format, and the energy is passed onto the next step to save to a CSV file.</li>\n",
    "<li>The dG score obtained is written to a csv file under the 'computed_dG' column under the PDB ID label. Extra information (ligand SMILES, protein sequence and experimental binding affinity) are also included in the CSV file.</li>\n",
    "<li>All data are moved to the 'data' directory except the CSV file, which continuously collects data in the current directory, and a new folder is created for each PDB ID.</li>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Installing and configuring required packages\n",
    "\n",
    "<b><font color='red'> (only run this if you don't have them already installed. If you git cloned this repo then no installation is required.) </font></b>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # !conda install -c conda-forge -c openbiosim biosimspace --yes\n",
    "# !conda install -c conda-forge -c bioconda mgltools openbabel zlib ncurses --yes\n",
    "# !conda install py3Dmol --yes\n",
    "# !conda install biopython --yes\n",
    "# !conda install pdb2pqr --yes\n",
    "# !conda install MDAnalysis --yes\n",
    "# # !pip install -q condacolab\n",
    "# # import condacolab # only needed for running on Google Colab\n",
    "# # condacolab.install_miniconda()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# avoid using pip install at all costs...\n",
    "# !pip3 -q install rdkit-pypi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# %%bash\n",
    "# # we wont be using this script directly, but its function is used in the ExtractLigandFromDU function.\n",
    "# wget -q https://docs.eyesopen.com/toolkits/python/_downloads/42fee916a9b875e2fc0a6e18e42c8701/extract_ligand_oedu.py\n",
    "# chmod +x ./extract_ligand_oedu.py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# %%bash\n",
    "# # this shell script will install ADFRsuite in the directory ADFRsuite, it is used to prepare the receptor and ligand for docking\n",
    "# mkdir ADFRsuite\n",
    "# cd ADFRsuite\n",
    "# wget -q https://ccsb.scripps.edu/adfr/download/1038 -O ADFRsuite_x86_64Linux_1.0.tar.gz\n",
    "# tar -xzf ADFRsuite_x86_64Linux_1.0.tar.gz\n",
    "# chmod a+x ADFRsuite_x86_64Linux_1.0\n",
    "# cd ADFRsuite_x86_64Linux_1.0\n",
    "# # this will install ADFRsuite in the current directory, the pipeline used will answer y to all questions prompted by the installer\n",
    "# yes Y | ./install.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# %%bash\n",
    "# # Install AutoDock Vina\n",
    "# mkdir -p vina\n",
    "# cd vina\n",
    "# wget -q https://github.com/ccsb-scripps/AutoDock-Vina/releases/download/v1.2.5/vina_1.2.5_linux_x86_64\n",
    "# chmod +x ./vina_1.2.5_linux_x86_64\n",
    "# # AutoDock GPU:\n",
    "# # wget -q https://github.com/ccsb-scripps/AutoDock-GPU/releases/download/v1.5.3/adgpu_analysis\n",
    "# # wget -q https://github.com/ccsb-scripps/AutoDock-GPU/releases/download/v1.5.3/adgpu-v1.5.3_linux_ocl_128wi\n",
    "# cd ..\n",
    "# alias vina='./vina/vina_1.2.5_linux_x86_64'\n",
    "# # OpenEye Tools:\n",
    "# # Protein Prep: \n",
    "# mkdir -p OpenEye\n",
    "# cd OpenEye\n",
    "# wget -q https://docs.eyesopen.com/toolkits/python/_downloads/99240d9bd18a29490f003efe389f1319/proteinprep.py\n",
    "# wget -q https://docs.eyesopen.com/toolkits/python/_downloads/42fee916a9b875e2fc0a6e18e42c8701/extract_ligand_oedu.py\n",
    "# chmod +x ./extract_ligand_oedu.py\n",
    "# chmod +x ./proteinprep.py\n",
    "# cd ..\n",
    "# alias proteinprep='./OpenEye/proteinprep.py'\n",
    "# alias extract_ligand='./OpenEye/extract_ligand_oedu.py'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We also need the asapdiscovery-modelling module to extract ligands. I've commented out the code as I've already git cloned it into my cwd."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# %%bash\n",
    "# we will use asap discovery to prepare the ligand for docking\n",
    "# follow guidance in https://github.com/asapdiscovery/asapdiscovery/tree/main to clone asapdiscovery\n",
    "# git clone https://github.com/choderalab/asapdiscovery.git\n",
    "# cd asapdiscovery\n",
    "# change the platform to the platform you are using\n",
    "# mamba env create -f devtools/conda-envs/asapdiscovery-{platform}.yml\n",
    "# conda activate asapdiscovery\n",
    "# pip install asapdiscovery-data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h1><b><font color='red'>(Run codes from this point on.) </font></b></h1></br>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You will also need your OpenEye license AND set it up correctly using ```\n",
    "os.environ['OE_LICENSE'] = '/path/to/license'``` before importing any OpenEye or ASAP Discovery modules, or OpenEye will kill your kernel!<br>\n",
    "\n",
    "Click <a href=https://docs.eyesopen.com/toolkits/python/quickstart-python/license.html>here</a> to see how to install licence for OpenEye Toolkits.\n",
    "\n",
    "On a LINUX commandline, you can run: <br>\n",
    "```export OE_LICENSE=/home/USERNAME/oe_license.txt```<br>\n",
    "to set your license path, and verify that it's been correctly set up by running:<br>\n",
    "```echo $OE_LICENSE```<br>\n",
    "```cat $OE_LICENSE```<br>\n",
    "\n",
    "These commands should print out the path to the license and the content of the license.\n",
    "\n",
    "ADFRsuite is used to prepare receptors and ligands for AutoDock Vina, and permission needs to be granted to scripts within `./ADFRsuite/ADFRsuite_x86_64Linux_1.0` "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "# giving permissions to run vina and ADFR scripts, change the directories of vina as required\n",
    "os.chmod('./vina/vina_1.2.5_linux_x86_64', 0o755) # giving permissions to run vina scripts\n",
    "os.chmod('./vina/vina_split_1.2.5_linux_x86_64', 0o755) # giving permissions to run vina scripts\n",
    "os.chmod('./extract_ligand.sh', 0o755) # giving permissions to run extract_ligand.sh\n",
    "os.chmod('./ADFRsuite/ADFRsuite_x86_64Linux_1.0', 0o755) # giving permissions to run ADFRsuite\n",
    "\n",
    "# make sure to set the OE_LICENSE environment variable, the full path should be included, or else openeye will kill your kernel!\n",
    "os.environ['OE_LICENSE'] = '/home/ian/oe_license.txt'\n",
    "os.chmod('/home/ian/oe_license.txt', 0o755)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from Bio.PDB import PDBList, PDBParser, Select, PDBIO\n",
    "from subprocess import Popen, PIPE\n",
    "import re\n",
    "import logging \n",
    "from pathlib import Path\n",
    "import contextlib\n",
    "import subprocess\n",
    "from rdkit import Chem\n",
    "from rdkit.Chem import rdMolTransforms as rdmt\n",
    "import MDAnalysis as mda\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from tqdm import tqdm \n",
    "from asapdiscovery.data.backend.openeye import (\n",
    "    oechem,\n",
    "    oedocking,\n",
    "    oegrid,\n",
    "    oespruce,\n",
    "    openeye_perceive_residues,\n",
    ")\n",
    "from asapdiscovery.modeling.schema import MoleculeComponent, MoleculeFilter\n",
    "from asapdiscovery.modeling.modeling import split_openeye_mol, make_design_unit\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import openeye.oechem as oechem\n",
    "from pdbfixer import PDBFixer\n",
    "from openmm.app import PDBFile\n",
    "from rdkit import Chem\n",
    "from rdkit.Chem import SDWriter\n",
    "from ase import Atoms\n",
    "from ase.io.sdf import read_sdf\n",
    "from ase.io import read\n",
    "from iodata import load_one\n",
    "from iodata.utils import angstrom\n",
    "\n",
    "from openmm.app.element import zinc, iron, calcium  # Import other metals as needed\n",
    "import openmm.unit as unit\n",
    "\n",
    "# Set up the working directory\n",
    "cwd = os.getcwd()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# We start by defining a function to change the directories temporarily, and another to run the shell commands, within built-in error traps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to change directories\n",
    "@contextlib.contextmanager\n",
    "def set_directory(dirname: os.PathLike, mkdir: bool = False):\n",
    "    pwd = os.getcwd()\n",
    "    path = Path(dirname).resolve()\n",
    "    if mkdir:\n",
    "        path.mkdir(exist_ok=True, parents=True)\n",
    "    os.chdir(path)\n",
    "    yield path\n",
    "    os.chdir(pwd)\n",
    "\n",
    "# Function to run shell commands\n",
    "def run_command(cmd, raise_error=True, input=None, timeout=None, **kwargs):\n",
    "    \"\"\"Run a shell command and handle possible errors.\"\"\"\n",
    "    # Popen is used to run the command\n",
    "    sub = Popen(cmd, shell=True, stdin=PIPE, stdout=PIPE, stderr=PIPE, **kwargs)\n",
    "    if input is not None:\n",
    "        sub.stdin.write(bytes(input, encoding='utf-8'))\n",
    "    try:\n",
    "        out, err = sub.communicate(timeout=timeout)\n",
    "        return_code = sub.poll()\n",
    "    # if the command times out, kill the process\n",
    "    except subprocess.TimeoutExpired:\n",
    "        sub.kill()\n",
    "        print(f\"Command {cmd} timeout after {timeout} seconds\")\n",
    "        return 999, \"\", \"\"  # 999 is a special return code for timeout\n",
    "    out = out.decode('utf-8')\n",
    "    err = err.decode('utf-8')\n",
    "    if raise_error and return_code != 0:\n",
    "        raise CommandExecuteError(f\"Command {cmd} failed: \\n{err}\")\n",
    "    return return_code, out, err\n",
    "\n",
    "# Exception for command execution errors\n",
    "class CommandExecuteError(Exception):\n",
    "    def __init__(self, msg):\n",
    "        self.msg = msg\n",
    "    def __str__(self):\n",
    "        return self.msg"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Using OpenEye Toolkits to extract the ligand (1st method)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# from https://github.com/choderalab/perses/blob/main/examples/moonshot-mainseries/00-prep-receptor.py#L19-L35\n",
    "def read_pdb_file(pdb_file):\n",
    "    print(f'Reading receptor from {pdb_file}...')\n",
    "\n",
    "    from openeye import oechem\n",
    "    ifs = oechem.oemolistream()\n",
    "    #ifs.SetFlavor(oechem.OEFormat_PDB, oechem.OEIFlavor_PDB_Default | oechem.OEIFlavor_PDB_DATA | oechem.OEIFlavor_PDB_ALTLOC)  # Causes extra protons on VAL73 for x1425\n",
    "    ifs.SetFlavor(oechem.OEFormat_PDB, oechem.OEIFlavor_PDB_Default | oechem.OEIFlavor_PDB_DATA )\n",
    "\n",
    "    if not ifs.open(pdb_file):\n",
    "        oechem.OEThrow.Fatal(\"Unable to open %s for reading.\" % pdb_file)\n",
    "\n",
    "    mol = oechem.OEGraphMol()\n",
    "    if not oechem.OEReadMolecule(ifs, mol):\n",
    "        oechem.OEThrow.Fatal(\"Unable to read molecule from %s.\" % pdb_file)\n",
    "    ifs.close()\n",
    "\n",
    "    return (mol)\n",
    "\n",
    "# script from extract_ligand_oedu.py: https://docs.eyesopen.com/toolkits/python/oechemtk/oebio_examples/oebio_example_extract_ligand.html#section-example-oebio-extract-ligand\n",
    "def ExtractLigandFromDU(du, ofs):\n",
    "# @ <SNIPPET-ExtractLigandFromDesignUnit>\n",
    "    lig = oechem.OEGraphMol()\n",
    "    if not du.GetLigand(lig):\n",
    "        oechem.OEThrow.Fatal(\"Error: Could not extract ligand from the OEDesignUnit.\")\n",
    "    oechem.OEWriteMolecule(ofs,lig) \n",
    "# @ </SNIPPET-ExtractLigandFromDesignUnit>\n",
    "\n",
    "    ofs.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Using OpenEye Toolkits to extract the ligand (2nd method in case the 1st fails)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to read a molecular complex/ligand/protein from a file\n",
    "def oe_read_molecule(filename):\n",
    "    ifs = oechem.oemolistream()\n",
    "    if not ifs.open(filename):\n",
    "        oechem.OEThrow.Fatal(\"Unable to open file %s\" % filename)\n",
    "    mol = oechem.OEGraphMol()\n",
    "    oechem.OEReadMolecule(ifs, mol)\n",
    "    ifs.close()\n",
    "    return mol\n",
    "\n",
    "def oe_split_complex(input_filename, ligand_output_filename, protein_output_filename=None):\n",
    "    # Read the complex\n",
    "    complex_molecule = oe_read_molecule(input_filename)\n",
    "\n",
    "    # Split the complex\n",
    "    protein = oechem.OEGraphMol()\n",
    "    ligand = oechem.OEGraphMol()\n",
    "    water = oechem.OEGraphMol()\n",
    "    other = oechem.OEGraphMol()\n",
    "\n",
    "    # Split the complex\n",
    "    oechem.OESplitMolComplex(ligand, protein, water, other, complex_molecule)\n",
    "\n",
    "    # Write the components to files\n",
    "    oechem.OEWriteMolecule(oechem.oemolostream(ligand_output_filename), ligand)\n",
    "    \n",
    "    # we dont need the water and other components\n",
    "    # oechem.OEWriteMolecule(oechem.oemolostream(protein_output_filename), protein)\n",
    "    # oechem.OEWriteMolecule(oechem.oemolostream(output_basename + \"_water.pdb\"), water)\n",
    "    # oechem.OEWriteMolecule(oechem.oemolostream(output_basename + \"_other.pdb\"), other)\n",
    "\n",
    "# Example usage\n",
    "# input_filename = \"./Mpro-x0072_0.pdb\"  # Replace with your file path\n",
    "# ligand_output_filename = \"./Mpro-x0072_0_ligand.pdb\"\n",
    "# protein_output_filename = \"./Mpro-x0072_0_protein.pdb\"\n",
    "# oe_split_complex(input_filename, ligand_output_filename, protein_output_filename)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Define some classes and functions to download, rename and split the ligand and the receptor from a complex.<br>\n",
    "For ligand, we are using OpenEye to extract, for receptors (proteins), we use a customisable biopython module. (Please do not use biopython to extract ligands, as it does not retain the position of the ligand atoms)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# using Biopython to extract the receptor\n",
    "class ReceptorSelect(Select):\n",
    "    '''This class is used to select only the receptor residues from a PDB file, by excluding any HETATM and any water.'''\n",
    "\n",
    "    def __init__(self, preserve_water=False, preserve_metal=True):\n",
    "        self.preserve_water = preserve_water\n",
    "        # List of common heavy metal ions in protein structures\n",
    "        self.preserve_metal = preserve_metal\n",
    "        # I deleted the heavy metal ions as they are not a valid AutoDock type! Be cautious here!\n",
    "        self.heavy_metal_ions = ['ZN', 'FE', 'MG', 'CA', 'MN', 'CO', 'CU', 'NI', 'MO', 'W','YB', 'K', 'NA']  # TODO: Add more metal ions as needed\n",
    "\n",
    "\n",
    "    def accept_residue(self, residue):\n",
    "        # Exclude water molecules, assumed to have residue name 'HOH' or 'WAT'\n",
    "        #TODO: What if there's water in the binding site?\n",
    "        resname = residue.get_resname()\n",
    "        # print(f\"Processing residue: {resname}\")\n",
    "\n",
    "        if resname in ['HOH', 'WAT']:\n",
    "            if self.preserve_water:\n",
    "                # print(\"Preserving water molecule\")\n",
    "                return True\n",
    "            else:\n",
    "                # print(\"Excluding water molecule\")\n",
    "                return False\n",
    "\n",
    "        # Include heavy metal ions\n",
    "        if resname in self.heavy_metal_ions:\n",
    "            if self.preserve_metal:\n",
    "                # print(\"Preserving metal ion\")\n",
    "                return True\n",
    "            else:\n",
    "                # print(\"Excluding metal ion\")\n",
    "                return False\n",
    "        \n",
    "        # Check if any atom in the residue is from an ATOM record\n",
    "        for atom in residue:\n",
    "            '''in biopython, atom.get_full_id()[3] is accessing the fourth element of the full ID. This element represents the atom name, atom.get_full_id()[3][0] is accessing the first character of the atom name. This character represents the element symbol of the atom. The condition atom.get_full_id()[3][0] == ' ' checks whether the first character of the atom name is a space. If it is a space, then the atom is from an ATOM record, otherwise it is from a HETATM record.'''\n",
    "            if atom.get_full_id()[3][0] == ' ':\n",
    "                return True\n",
    "            \n",
    "        # print(\"Excluding residue\")\n",
    "        return False\n",
    "\n",
    "def download_pdb_file(pdb_id):\n",
    "    \"\"\"Download PDB file using PDB ID.\"\"\"\n",
    "    pdbl = PDBList()\n",
    "    filename = pdbl.retrieve_pdb_file(pdb_id, file_format='pdb', pdir='.', overwrite=True)\n",
    "    return filename\n",
    "\n",
    "def get_structure(filename):\n",
    "    \"\"\"Parse the structure from a PDB file.\"\"\"\n",
    "    parser = PDBParser()\n",
    "    structure = parser.get_structure('structure', filename)\n",
    "    return structure\n",
    "\n",
    "def save_receptor(structure, receptor_file, preserve_water=False, preserve_metal=False):\n",
    "    \"\"\"Save the receptor part of the PDB file.\"\"\"\n",
    "    io = PDBIO()\n",
    "    io.set_structure(structure)\n",
    "    io.save(receptor_file, ReceptorSelect(preserve_water=preserve_water,\n",
    "    preserve_metal = preserve_metal)) \n",
    "\n",
    "# Example usage:\n",
    "# pdb_id = '1abc'\n",
    "# filename = download_pdb_file(pdb_id)\n",
    "# structure = get_structure(filename)\n",
    "# save_receptor(structure, 'receptor.pdb', preserve_water=True)\n",
    "\n",
    "def fix_receptor_file(input_receptor_file):\n",
    "    '''This function must be used for complexes without crystal structurers on PDB, as there could be places where things should be fixed before docking. This must be done or else the docking software might report errors. '''\n",
    "\n",
    "    #TODO: Fix this function - it unwantedly removes metal ions and water from the structure\n",
    "\n",
    "    # Initialize the PDBFixer with the input file\n",
    "    fixer = PDBFixer(filename=input_receptor_file)\n",
    "    \n",
    "    # Find and fix missing residues, atoms, and hydrogens\n",
    "    fixer.findMissingResidues()\n",
    "    fixer.findMissingAtoms()\n",
    "    fixer.addMissingAtoms()\n",
    "    fixer.addMissingHydrogens(7.0)\n",
    "    \n",
    "    # Define the output file name\n",
    "    output_receptor_file = input_receptor_file.replace('.pdb', '_fixed.pdb')\n",
    "    \n",
    "    # Save the fixed receptor file\n",
    "    with open(output_receptor_file, 'w') as output_file:\n",
    "        PDBFile.writeFile(fixer.topology, fixer.positions, output_file)\n",
    "    \n",
    "    print(f\"Fixed receptor file saved as: {output_receptor_file}\")\n",
    "    \n",
    "    return output_receptor_file\n",
    "\n",
    "# def fix_receptor_file(input_receptor_file):\n",
    "#     '''This function must be used for complexes without crystal structures on PDB, as there could be places where things should be fixed before docking. This must be done or else the docking software might report errors.'''\n",
    "\n",
    "#     # Initialize the PDBFixer with the input file\n",
    "#     fixer = PDBFixer(filename=input_receptor_file)\n",
    "    \n",
    "#     # Preserve metal ions\n",
    "#     metals = ['ZN', 'FE','CA']  # Add other metal symbols as needed\n",
    "#     metal_atoms = []\n",
    "#     for chain in fixer.topology.chains():\n",
    "#         for residue in chain.residues():\n",
    "#             if residue.name in metals:\n",
    "#                 for atom in residue.atoms():\n",
    "#                     metal_atoms.append((atom, atom.element, atom.name, residue.id, chain.id))\n",
    "\n",
    "#     # Find and fix missing residues, atoms, and hydrogens\n",
    "#     fixer.findMissingResidues()\n",
    "#     fixer.findMissingAtoms()\n",
    "#     fixer.addMissingAtoms()\n",
    "#     fixer.addMissingHydrogens(7.0)\n",
    "\n",
    "#     # Add metal ions back\n",
    "#     for atom, element, name, residue_id, chain_id in metal_atoms:\n",
    "#         chain = next(chain for chain in fixer.topology.chains() if chain.id == chain_id)\n",
    "#         residue = chain.residue(residue_id)\n",
    "#         if not any(a.name == name for a in residue.atoms()):\n",
    "#             new_atom = residue.addAtom(name, element)\n",
    "#             fixer.positions.append((atom.element.mass*unit.dalton).value_in_unit(unit.angstrom))\n",
    "\n",
    "#     # Define the output file name\n",
    "#     output_receptor_file = input_receptor_file.replace('.pdb', '_fixed.pdb')\n",
    "\n",
    "#     # Save the fixed receptor file\n",
    "#     with open(output_receptor_file, 'w') as output_file:\n",
    "#         PDBFile.writeFile(fixer.topology, fixer.positions, output_file)\n",
    "    \n",
    "#     print(f\"Fixed receptor file saved as: {output_receptor_file}\")\n",
    "    \n",
    "#     return output_receptor_file\n",
    "\n",
    "# Example usage:\n",
    "# fixed_file = fix_receptor_file('Mpro-x0072_0_receptor.pdb')\n",
    "\n",
    "def rename_file(old_filename, new_filename):\n",
    "    \"\"\"Rename the downloaded file to a standard name.\"\"\"\n",
    "    if os.path.exists(old_filename) and not os.path.exists(new_filename):\n",
    "        os.rename(old_filename, new_filename)\n",
    "    return new_filename\n",
    "\n",
    "def extract_ligand(filename, ligand_file):\n",
    "    \"\"\"Extract the ligand from the complex.\"\"\"\n",
    "    print(f'Extracting ligand from {filename}...')\n",
    "    complex = read_pdb_file(filename)\n",
    "    success, du = make_design_unit(complex)\n",
    "    \n",
    "    if success:\n",
    "        oeoutfile = oechem.oemolostream(ligand_file)\n",
    "        ExtractLigandFromDU(du, oeoutfile)\n",
    "    else:\n",
    "        print(f'Failed to extract ligand from {filename} using the OpenEye Toolkit,' \n",
    "              'trying a different script written using OESplitMolComplex...')\n",
    "        # Extract ligand using shell script\n",
    "        complex_id = filename.rsplit('.')[0]\n",
    "\n",
    "        # script written using OESplitMolComplex:\n",
    "        oe_split_complex(input_filename = filename, \n",
    "                         ligand_output_filename = f\"{ligand_file}\")\n",
    "        \n",
    "        # this is my shell script - not guaranteed to work for everything and hence not used!\n",
    "        # run_command(f'./extract_ligand.sh {complex_id}') \n",
    "\n",
    "def pdb_to_prot_lig(pdb_id, filename, preserve_water=False, preserve_metal=False):\n",
    "    \"\"\"Main function to handle the splitting of protein and ligand.\"\"\"\n",
    "    structure = get_structure(filename)\n",
    "\n",
    "    # naming the output files\n",
    "    receptor_file = f\"{pdb_id}_receptor.pdb\"\n",
    "    ligand_file = f\"{pdb_id}_ligand.pdb\"\n",
    "    complex_file = rename_file(filename, f\"{pdb_id}.pdb\")\n",
    "\n",
    "    # ---old codes (uses Biopython for receptor extraction, and ASAP for ligand extraction - slower)---\n",
    "    save_receptor(structure, receptor_file, preserve_water, preserve_metal)\n",
    "    receptor_file = fix_receptor_file(receptor_file)\n",
    "    extract_ligand(complex_file, ligand_file)\n",
    "\n",
    "    # ---new codes (uses OE for ligand and receptor extraction)---\n",
    "    # split the complex into protein and ligand\n",
    "    # oe_split_complex(input_filename = filename, ligand_output_filename = ligand_file, protein_output_filename = receptor_file)\n",
    "    \n",
    "    print(f'The complex file has been saved as {complex_file}')\n",
    "    print(f'The receptor file has been saved as {receptor_file}')\n",
    "    print(f'The ligand file has been saved as {ligand_file}')\n",
    "    return complex_file, receptor_file, ligand_file\n",
    "\n",
    "# Example usage:\n",
    "# filename = download_pdb_file(pdb_id) # if starting from a PDB ID\n",
    "# receptor_file, ligand_file = pdb_to_prot_lig(pdb_id, filename)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Prepare the protein and ligands for Vina, the prepped files end in .pdbqt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "class DockingPrepper:\n",
    "    '''This class is used to prepare the ligand and protein files for docking using AutoDock Vina. It uses the ADFRsuite to convert the ligand and protein files to pdbqt format.\n",
    "    # Example usage:\n",
    "    prep = DockingPrepper(folder='path/to/folder', pdb_id = 'pdb_id.pdb', lig_file='ligand.pdb', prot_file='protein.pdb')\n",
    "    \n",
    "    # ---- AutoDock Vina Preparation ----\n",
    "    prep.vina_process()\n",
    "    \n",
    "    # Alternatively...\n",
    "    # ---- OpenEye Preparation ----\n",
    "    prep.oe_process()\n",
    "    '''\n",
    "    def __init__(self, folder, pdb_id, lig_file, prot_file, lig_out_name=None, prot_out_name=None, add_h=True, preserve_water=False, preserve_metal=True):\n",
    "        self.folder = folder\n",
    "        self.pdb_id = pdb_id\n",
    "        self.lig_file = lig_file\n",
    "        self.prot_file = prot_file\n",
    "        self.lig_out_name = lig_out_name or f\"{lig_file.split('.')[0]}.pdbqt\"\n",
    "        self.prot_out_name = prot_out_name or f\"{prot_file.split('.')[0]}.pdbqt\"\n",
    "        self.add_h = add_h\n",
    "        self.preserve_water = preserve_water\n",
    "        self.preserve_metal = preserve_metal #TODO/WARMING: This is not implemented yet!\n",
    "\n",
    "    # ----------------- AutoDock Vina Preparation -----------------\n",
    "    def vina_prepare_ligand(self):\n",
    "        \"\"\"Prepare the ligand file for Vina.\"\"\"\n",
    "        if self.add_h:\n",
    "            cmd = f'./ADFRsuite/ADFRsuite_x86_64Linux_1.0/bin/prepare_ligand -l {self.lig_file} -o {self.lig_out_name} -A hydrogens'\n",
    "        else:\n",
    "            cmd = f'./ADFRsuite/ADFRsuite_x86_64Linux_1.0/bin/prepare_ligand -l {self.lig_file} -o {self.lig_out_name}'\n",
    "        self._run_command(cmd, f'The ligand file has been saved as {self.lig_out_name}')\n",
    "\n",
    "    def vina_prepare_protein(self):\n",
    "        \"\"\"Prepare the protein file for Vina.\"\"\"\n",
    "        if self.preserve_water:\n",
    "            cmd = f'./ADFRsuite/ADFRsuite_x86_64Linux_1.0/bin/prepare_receptor -r {self.prot_file} -o {self.prot_out_name} -A checkhydrogens -U nphs_lps_nonstdres'\n",
    "            '''        \n",
    "            [-U]  cleanup type:\n",
    "             'nphs': merge charges and remove non-polar hydrogens\n",
    "             'lps': merge charges and remove lone pairs\n",
    "             'waters': remove water residues\n",
    "             'nonstdres': remove chains composed entirely of residues of\n",
    "                      types other than the standard 20 amino acids\n",
    "             'deleteAltB': remove XX@B atoms and rename XX@A atoms->XX\n",
    "             (default is 'nphs_lps_waters_nonstdres') \n",
    "             '''\n",
    "        else:\n",
    "            cmd = f'./ADFRsuite/ADFRsuite_x86_64Linux_1.0/bin/prepare_receptor -r {self.prot_file} -o {self.prot_out_name} -A checkhydrogens'\n",
    "        self._run_command(cmd, f'The protein file {self.prot_file} has been converted to pdbqt format and saved in {self.prot_out_name}')\n",
    "\n",
    "    def _run_command(self, cmd, success_message):\n",
    "        \"\"\"Run a shell command and handle possible errors.\"\"\"\n",
    "        try:\n",
    "            subprocess.run(cmd, shell=True, check=True, timeout=999)\n",
    "            print(success_message)\n",
    "        except subprocess.CalledProcessError as e:\n",
    "            print(f\"Error occurred: {e.output}\")\n",
    "            raise e\n",
    "        except subprocess.TimeoutExpired:\n",
    "            print(f\"Command '{cmd}' timed out.\")\n",
    "            raise e\n",
    "\n",
    "    def vina_process(self):\n",
    "        \"\"\"Process the ligand and protein files.\"\"\"\n",
    "        with set_directory(self.folder):\n",
    "        # if the lig_file and prot_file are not in pdb format, convert them to pdb format using obabel\n",
    "            if self.lig_file.split('.')[-1] != 'pdb':\n",
    "                lig_name = self.lig_file.split('.')[0]\n",
    "                cmd = f'obabel {self.lig_file} -O {lig_name}.pdb'\n",
    "                self._run_command(cmd, f'The ligand file has been converted to {lig_name}.pdb')\n",
    "                self.lig_file = f'{lig_name}.pdb'\n",
    "\n",
    "            if self.prot_file.split('.')[-1] != 'pdb':\n",
    "                prot_name = self.prot_file.split('.')[0]\n",
    "                cmd = f'obabel {self.prot_file} -O {prot_name}.pdb'\n",
    "                self._run_command(cmd, f'The protein file has been converted to {prot_name}.pdb')\n",
    "                self.prot_file = f'{prot_name}.pdb'\n",
    "                \n",
    "            self.vina_prepare_ligand()\n",
    "            self.vina_prepare_protein()\n",
    "    \n",
    "    # ----------------- OpenEye Preparation -----------------\n",
    "    def oe_make_design_unit(self):\n",
    "        ''' \n",
    "        This function makes design units from a protein-ligand complex.\n",
    "        \n",
    "        Input: input_file - the protein-ligand complex file\n",
    "        Saved to file: {input_file_basename}_DU_{i}.oedu \n",
    "        Output: output_files - a list of the design unit files\n",
    "        '''\n",
    "        input_file = f\"{self.pdb_id}.pdb\"\n",
    "    \n",
    "        os.system(f\"python ./OpenEye/make_design_units.py {input_file}\")\n",
    "        \n",
    "        # this is the output pattern for the design unit files\n",
    "        output_pattern = os.path.basename(input_file)[:-4] + \"_DU_{}.oedu\"\n",
    "        # List the files in the current directory and filter for the expected output pattern\n",
    "        output_files = []\n",
    "        i = 0\n",
    "        while True:\n",
    "            output_file = output_pattern.format(i)\n",
    "            if os.path.exists(output_file):\n",
    "                output_files.append(output_file)\n",
    "                i += 1\n",
    "            else:\n",
    "                break\n",
    "\n",
    "        # output_files now contains the names of the output files\n",
    "        print(f\"Design unit was successfully made for {input_file}, output is saved to {output_files}.\")\n",
    "        return output_files\n",
    "\n",
    "    def oe_make_receptor(self, input_file):\n",
    "\n",
    "        oe_output_files = [] \n",
    "        for ifile in input_file:\n",
    "            output_basename = os.path.basename(ifile)[:-5]\n",
    "            os.system(f\"python ./OpenEye/MakeReceptor.py -in {ifile} -out {output_basename}_receptor.oedu\") # this will output {ifile}_receptor.oedu\n",
    "\n",
    "            oe_output_files.append(f\"{output_basename}_receptor.oedu\")\n",
    "            print(f'The receptor design unit file has been saved as {output_basename}_receptor.oedu')\n",
    "\n",
    "        return oe_output_files \n",
    "    \n",
    "    # Example usage:\n",
    "    # input_file = \"6EQ2.pdb\"\n",
    "    # complex_DU = oe_make_design_unit(input_file) # this will output ['6EQ2_DU_0.oedu']\n",
    "    # receptor_DU = oe_make_receptor(complex_DU) # this will output ['6EQ2_DU_0_receptor.oedu']\n",
    "\n",
    "    def oe_process(self):\n",
    "        '''Process the ligand and protein files using OpenEye.'''\n",
    "        print(f'Preparing the ligand and protein files using OpenEye Toolkits...')\n",
    "        with set_directory(self.folder):\n",
    "            complex_DU = self.oe_make_design_unit()\n",
    "            oe_output_files = self.oe_make_receptor(input_file=complex_DU)\n",
    "        \n",
    "        # self.oe_output_files = [f\"{self.pdb_id}_DU_0_receptor.oedu\"]\n",
    "        return oe_output_files"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Define a function to get the COM of the ligand file and hence a config file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_COM(file):\n",
    "        if file.endswith('mol2') or file.endswith('xyz'):\n",
    "            mol = load_one(file)\n",
    "            ase_mol = Atoms(numbers=mol.atnums, positions=mol.atcoords / angstrom)\n",
    "        elif file.endswith('sdf'):\n",
    "            ase_mol = read_sdf(file)\n",
    "        elif file.endswith('pdb'):\n",
    "            ase_mol = read(file)\n",
    "        else:\n",
    "            raise NotImplementedError(f\"File extension not supported for {file}\")\n",
    "\n",
    "        return ase_mol.get_center_of_mass()\n",
    "    \n",
    "def write_config_vina(lig_pdbqt,prot_pdbqt,center, config_fp = \"config.txt\", weights=None, boxsize=50, exhaustiveness=32, num_modes=1, energy_range=30, **kwargs):\n",
    "    '''\n",
    "    Write the config file for AutoDock Vina docking\n",
    "    :param exhaustiveness: int, the exhaustiveness of the docking\n",
    "    :param num_modes: int, the number of modes (conformations) to be generated\n",
    "    :param energy_range: int, the energy range of the docking\n",
    "    '''\n",
    "\n",
    "    lines = [\"receptor = {}\".format(prot_pdbqt),\n",
    "            \"ligand = {}\".format(lig_pdbqt),\n",
    "            \"scoring = vina\",\n",
    "            \"\",\n",
    "            \"center_x = {}\".format(center[0]),\n",
    "            \"center_y = {}\".format(center[1]),\n",
    "            \"center_z = {}\".format(center[2]),\n",
    "            \"\",\n",
    "            \"size_x = {}\".format(boxsize),\n",
    "            \"size_y = {}\".format(boxsize),\n",
    "            \"size_z = {}\".format(boxsize),\n",
    "            \"\",\n",
    "            # \"exhaustiveness = {}\".format(exhaustiveness),\n",
    "            # \"num_modes = {}\".format(num_modes),\n",
    "            # \"energy_range = {}\".format(energy_range),\n",
    "            ]\n",
    "    if weights is not None:\n",
    "        assert len(weights) == 6, \"Autodock vina needs 6 weights\"\n",
    "        # --weight_gauss1 1 --weight_gauss2 0 --weight_repulsion 0  --weight_hydrophobic 0 --weight_hydrogen 0 --weight_rot 0\"\n",
    "        lines.extend([\n",
    "            f\"weight_gauss1 = {weights[0]}\",\n",
    "            f\"weight_gauss2 = {weights[1]}\",\n",
    "            f\"weight_repulsion = {weights[2]}\",\n",
    "            f\"weight_hydrophobic = {weights[3]}\",\n",
    "            f\"weight_hydrogen = {weights[4]}\",\n",
    "            f\"weight_rot = {weights[5]}\",\n",
    "        ])\n",
    "    with open(config_fp, \"w\") as f:\n",
    "        f.write(\"\\n\".join(lines))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Defining a function to run the AutoDock Vina and obtain the docking score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "class DockingScorer:\n",
    "    '''This class is used to score the ligand using AutoDock Vina. It uses the prepared ligand and protein files to run Vina and extract the docking score.\n",
    "    Example usage:\n",
    "    scorer = DockingScorer(folder='path/to/folder', lig_file='ligand.pdbqt', prot_file='protein.pdbqt', save_out_file=True)\n",
    "    docking_score = scorer.vina_score_ligand()\n",
    "    print(f\"Docking Score: {docking_score}\")\n",
    "    '''\n",
    "    def __init__(self, \n",
    "                 folder, \n",
    "                 lig_file, \n",
    "                 prot_file,\n",
    "                 complex_file=None,\n",
    "                 get_vina_poses=False,\n",
    "                 weights=None, \n",
    "                 save_out_file=True, \n",
    "                 protein_sequence=None, \n",
    "                 smiles=None, \n",
    "                 exp_binding_affinity=None,\n",
    "                 from_pdb=True,\n",
    "                 csv_out_file='docking_data_playground_from_pdb.csv',\n",
    "                 receptor_DU=None):\n",
    "        self.folder = folder\n",
    "        self.lig_file = lig_file\n",
    "        self.prot_file = prot_file\n",
    "        self.complex_file = complex_file\n",
    "        self.lig_name = os.path.splitext(lig_file)[0]\n",
    "        self.prot_name = os.path.splitext(prot_file)[0]\n",
    "        self.complex_name = os.path.splitext(complex_file)[0]\n",
    "        self.pdb_id = self.prot_name.split('_')[0]\n",
    "        self.get_vina_poses = get_vina_poses\n",
    "        self.weights = weights\n",
    "        self.save_out_file = save_out_file\n",
    "        #TODO: more work required here to extract the protein sequence, ligand SMILES string and exp dG value, from an online db (low priority)\n",
    "        self.protein_sequence = protein_sequence\n",
    "        self.smiles = smiles\n",
    "        self.exp_binding_affinity = exp_binding_affinity\n",
    "        self.from_pdb = from_pdb\n",
    "        self.csv_out_file = csv_out_file\n",
    "        self.receptor_DU = receptor_DU\n",
    "\n",
    "    def check_files(self):\n",
    "        '''Check and download the necessary files if needed'''\n",
    "        if not os.path.isfile(f\"{self.folder}/{self.prot_name}.pdbqt\") or not os.path.isfile(f\"{self.folder}/{self.lig_name}.pdbqt\"):\n",
    "            filename = download_pdb_file(self.pdb_id)\n",
    "            self.complex_file, self.prot_file, self.lig_file = pdb_to_prot_lig(self.pdb_id, filename)\n",
    "    #TODO: add a functionality to use Vina GPU instead (low priority)\n",
    "    \n",
    "    # ----------------- AutoDock Vina Scoring -----------------\n",
    "    def run_vina(self):\n",
    "        '''Run Vina with the prepared files and return the output.'''\n",
    "        \n",
    "        with set_directory(self.folder):\n",
    "            \n",
    "            # extract a reference ligand from the protein-ligand complex\n",
    "            ref_ligand = f\"{self.complex_name}_ref_ligand.pdb\"\n",
    "            extract_ligand(self.complex_file, ref_ligand)\n",
    "\n",
    "            # get the COM from the reference ligand so we can define a docking grid\n",
    "            ligand_COM = get_COM(ref_ligand)\n",
    "            write_config_vina(f'{self.lig_name}.pdbqt', f'{self.prot_name}.pdbqt',ligand_COM,config_fp = f\"{self.lig_name}_{self.prot_name}_config.txt\", weights=None)\n",
    "            cmd = f\"./vina/vina_1.2.5_linux_x86_64 --config {self.lig_name}_{self.prot_name}_config.txt --score_only\" #TODO: make it so that this can also output poses\n",
    "\n",
    "            if self.get_vina_poses == True:\n",
    "                cmd = f\"./vina/vina_1.2.5_linux_x86_64 --config {self.lig_name}_{self.prot_name}_config.txt\"\n",
    "\n",
    "            try:\n",
    "                out, err = \"\", \"\"  # Initialize out and err\n",
    "                code, out, err = run_command(cmd, timeout=100)\n",
    "                if code != 0:\n",
    "                    raise CommandExecuteError(f\"Command failed with return code {code}\")\n",
    "                \n",
    "                return out\n",
    "            except CommandExecuteError as e:\n",
    "                print(f\"Error in {self.pdb_id}: {e}\")\n",
    "                print(\"out: \", out)\n",
    "                print(\"err: \", err)\n",
    "                raise e\n",
    "\n",
    "    def extract_vina_score(self, out):\n",
    "        '''Extract the docking score from Vina output.'''\n",
    "        if not self.get_vina_poses:\n",
    "            strings = re.split('Estimated Free Energy of Binding   :', out)\n",
    "            line = strings[1].split('\\n')[0]\n",
    "            energy = float(line.strip().split()[0])\n",
    "            return energy\n",
    "        else:\n",
    "            # TODO: if get_vina_poses is True, return the output and the energy\n",
    "            energy = []\n",
    "            for line in out.split('\\n'):\n",
    "                match = re.search(r'^\\s*\\d+\\s+(-\\d+\\.\\d+)', line)\n",
    "                if match:\n",
    "                    energy.append(float(match.group(1)))\n",
    "            return min(energy) if energy else None\n",
    "\n",
    "    \n",
    "    # ----------------- OpenEye Toolkits Scoring -----------------\n",
    "    def oe_clean_then_dock(self):\n",
    "        ''' This function docks a ligand to a receptor using OpenEye's CleanThenDockMolecules.py script.\n",
    "        Input: lig_file - the ligand file to be docked\n",
    "            receptor_DU - the receptor file to dock the ligand to\n",
    "        Output: output_files - a list of the docked ligand files, which also contains the chemgauss4 score.'''\n",
    "        \n",
    "        output_files = [] \n",
    "\n",
    "        if self.receptor_DU is None:\n",
    "            print(\"No receptor design unit file provided. Please provide a receptor file to dock the ligand.\")\n",
    "            raise ValueError(\"No receptor design unit file provided.\")\n",
    "        for receptor_du in self.receptor_DU:   \n",
    "            output_basename = f\"{receptor_du.split('.')[0]}_{self.lig_name}\"\n",
    "            os.system(f'python ./OpenEye/CleanThenDockMolecules.py -in {self.lig_file} -out {output_basename}_docked.sdf -receptor {receptor_du}') # output score is Chemgauss4, contained in the sdf file\n",
    "            \n",
    "            output_files.append(f\"{output_basename}_docked.sdf\")\n",
    "        \n",
    "        print(f'The docked ligand file has been saved as {output_files}')\n",
    "        return output_files    \n",
    "    \n",
    "\n",
    "    ################## combined\n",
    "    # # TODO: modify this function to calculate and extract the other scores from the docked ligands\n",
    "    def extract_chemgauss4_scores(self, docked_ligand):\n",
    "        ''' This function extracts the Chemgauss4 scores from the docked ligands using RDKit and falls back to regex if needed.'''\n",
    "        \n",
    "        # Handle if docked_ligand is passed as a list\n",
    "        if isinstance(docked_ligand, list):\n",
    "            docked_ligand = docked_ligand[0]\n",
    "\n",
    "        chemgauss4_scores = []\n",
    "\n",
    "        try:\n",
    "            # Try using RDKit-based approach\n",
    "            supplier = Chem.SDMolSupplier(docked_ligand)\n",
    "            for mol in supplier:\n",
    "                if mol is None:\n",
    "                    raise ValueError(\"Error: Invalid molecule found, try the regex-based approach.\")\n",
    "                \n",
    "                if mol.HasProp('Chemgauss4'):\n",
    "                    energy = mol.GetProp('Chemgauss4')\n",
    "                    energy = float(energy)\n",
    "                    chemgauss4_scores.append(energy)\n",
    "\n",
    "                    if energy is None or chemgauss4_scores is None:\n",
    "                        raise ValueError(\"Energy or chemgauss4_scores is None\")\n",
    "\n",
    "        except Exception as e:\n",
    "            print(f\"RDKit approach failed: {e}, trying regex-based approach...\")\n",
    "            # Fallback to regex-based approach\n",
    "            chemgauss4_pattern = re.compile(r'> <Chemgauss4>\\s+(-?\\d+\\.\\d+)')\n",
    "            try:\n",
    "                with open(docked_ligand, 'r') as file:\n",
    "                    sdf_content = file.read()\n",
    "\n",
    "                matches = chemgauss4_pattern.findall(sdf_content)\n",
    "                chemgauss4_scores = [float(score) for score in matches]\n",
    "\n",
    "            except FileNotFoundError:\n",
    "                print(f\"Error: The file {docked_ligand} was not found.\")\n",
    "            except Exception as e:\n",
    "                print(f\"An error occurred while extracting Chemgauss4 scores: {e}\")\n",
    "\n",
    "        return chemgauss4_scores\n",
    "\n",
    "\n",
    "\n",
    "    # # Example usage\n",
    "    # docked_ligand = oe_clean_then_dock(ligand_file, receptor_DU) # this will output ['6EQ2_ligand_docked.sdf'], suppose ligand_file = '6EQ2_ligand.pdb'\n",
    "    # Extracting Chemgauss4 scores from the docked ligands\n",
    "    # scores = extract_chemgauss4_scores(docked_ligand)\n",
    "\n",
    "    # ----------------- Putting things together -----------------\n",
    "    # make the output into a pandas dataframe\n",
    "    # add pdb_id, protein sequence, ligand SMILES string, dG, exp_dG as a column in the dataframe\n",
    "    def extract_data_from_leakypdb(self, df):\n",
    "        '''Extract the protein sequence, ligand SMILES string, and experimental binding affinity from the specified dataframe (leakypdb_test.csv).'''\n",
    "\n",
    "        # # Read the DataFrame (must be done so that df is correctly recognised as a DataFrame, not a string.)    \n",
    "        # df = pd.read_csv(df)\n",
    "\n",
    "        # Filter the DataFrame for the specific PDB ID\n",
    "        filtered_df = df[df['pdb_id'] == self.pdb_id] \n",
    "\n",
    "        # Check if the specific columns exist in the DataFrame\n",
    "        required_columns = ['pdb_id','smiles', 'protein_sequence', 'binding_affinity']\n",
    "        missing_columns = [col for col in required_columns if col not in df.columns]\n",
    "        if missing_columns:\n",
    "            return f\"Missing columns in the data: {', '.join(missing_columns)}\"\n",
    "        # Extract the needed information\n",
    "        result = filtered_df[required_columns]\n",
    "        return result    \n",
    "\n",
    "    def save_output(self, out, energy):\n",
    "        '''Save the vina output to a file if required.'''\n",
    "        if self.save_out_file:\n",
    "            if self.from_pdb:\n",
    "                with open(f\"{self.pdb_id}.out\", 'w') as f:\n",
    "                    # Convert list to string if 'out' is a list\n",
    "                    if isinstance(out, list):\n",
    "                        out = \"\\n\".join(out)\n",
    "                    f.write(out)\n",
    "                print(f\"Output saved as {self.pdb_id}.out\\n\")\n",
    "\n",
    "                df = self.extract_data_from_leakypdb(leakypdb) # define the dataframe to extract data from\n",
    "\n",
    "                # add data to the dataframe: Assign the energy value to a new column 'computed_dG' for the filtered rows\n",
    "                df.loc[df['pdb_id'] == self.pdb_id, 'computed_dG'] = energy\n",
    "                updated_rows = df[df['pdb_id'] == self.pdb_id]\n",
    "\n",
    "                if not os.path.isfile(f'{self.csv_out_file}'):\n",
    "                    updated_rows.to_csv(f'{self.csv_out_file}', index=False)\n",
    "                else: # else it exists so append without writing the header\n",
    "                    updated_rows.to_csv(f'{self.csv_out_file}', mode='a', header=False, index=False)\n",
    "                print(f\"Data saved to {self.csv_out_file}\")\n",
    "                return updated_rows\n",
    "            else:\n",
    "                if isinstance(energy, list): \n",
    "                    for i, e in enumerate(energy):\n",
    "                        with open(f\"{self.prot_name}_{self.lig_name}_{i}.out\", 'w') as f:\n",
    "                            # Convert list to string if 'out' is a list\n",
    "                            if isinstance(out, list):\n",
    "                                out = \"\\n\".join(out)\n",
    "                            f.write(out)\n",
    "                        print(f\"Output saved as {self.prot_name}_{self.lig_name}_{i}.out\\n\")  \n",
    "                        \n",
    "                        lig_name_with_index = f'{self.lig_name}_{i}'\n",
    "\n",
    "                        df = pd.DataFrame({'ligand_name': [lig_name_with_index], \n",
    "                                           'protein_name': [self.prot_name], \n",
    "                                           'computed_dG': [e],\n",
    "                                           'error_message': [None]})\n",
    "                                           \n",
    "                        if not os.path.isfile(f'{self.csv_out_file}'):\n",
    "                            df.to_csv(f'{self.csv_out_file}', index=False)\n",
    "                        else: # else it exists so append without writing the header\n",
    "                            df.to_csv(f'{self.csv_out_file}', mode='a', header=False, index=False)\n",
    "                        print(f\"Data saved to {self.csv_out_file}\")\n",
    "\n",
    "                else: \n",
    "                    with open(f\"{self.prot_name}_{self.lig_name}.out\", 'w') as f:\n",
    "                        # Convert list to string if 'out' is a list\n",
    "                        if isinstance(out, list):\n",
    "                            out = \"\\n\".join(out)\n",
    "                        f.write(out)\n",
    "                    print(f\"Output saved as {self.prot_name}_{self.lig_name}.out\\n\")\n",
    "                    \n",
    "                    df = pd.DataFrame({'ligand_name': [self.lig_name], \n",
    "                                    'protein_name': [self.prot_name], \n",
    "                                    'computed_dG': [energy],\n",
    "                                    'error_message': [None]})\n",
    "\n",
    "                    if not os.path.isfile(f'{self.csv_out_file}'):\n",
    "                        df.to_csv(f'{self.csv_out_file}', index=False)\n",
    "                    else: # else it exists so append without writing the header\n",
    "                        df.to_csv(f'{self.csv_out_file}', mode='a', header=False, index=False)\n",
    "                    print(f\"Data saved to {self.csv_out_file}\")\n",
    "                return df\n",
    "            \n",
    "    ######################## FINAL FUNCTIONS ########################\n",
    "    def vina_score_ligand(self):\n",
    "        \"\"\"Main method to score the ligand using Vina.\"\"\"\n",
    "        if self.from_pdb:\n",
    "            self.check_files()\n",
    "        try:\n",
    "            out = self.run_vina()\n",
    "            energy = self.extract_vina_score(out)\n",
    "        except Exception as e:\n",
    "            print(f\"Error in {self.pdb_id}: {e}\")\n",
    "            out = f\"Error in {self.pdb_id}: {e}\"\n",
    "            energy = None\n",
    "            raise e\n",
    "        \n",
    "        print(f\"{self.pdb_id}: Estimated Free Energy of Binding = {energy} kcal/mol\")\n",
    "        self.save_output(out, energy)\n",
    "        return energy\n",
    "\n",
    "    def oe_score_ligand(self):\n",
    "        '''Main method to score the ligand using OpenEye Toolkits.'''\n",
    "        # if self.from_pdb:\n",
    "        #     self.check_files()\n",
    "        try:\n",
    "            docked_ligand = self.oe_clean_then_dock()\n",
    "            # TODO: out should be the output on the commandline.... not the docked_ligand\n",
    "            out = docked_ligand\n",
    "            energies = []\n",
    "            if isinstance(docked_ligand, list) and len(docked_ligand) > 1:\n",
    "                for i, lig in enumerate(docked_ligand):\n",
    "                    energy = self.extract_chemgauss4_scores(lig)\n",
    "                    # debug line\n",
    "                    # print('debug line:', energy)\n",
    "                    if energy is not None:\n",
    "                        self.save_output(out, energy)\n",
    "                        if isinstance(energy, list): \n",
    "                            for j, e in enumerate(energy):\n",
    "                                print(f\"Docked ligand {lig}, molecule {j+1}: Chemgauss4 score = {e:.2f}\")\n",
    "                                # self.save_output(out, e)\n",
    "                                energies.append(e)\n",
    "                        else: \n",
    "                            print(f\"Docked ligand {lig}, molecule {i+1}: Chemgauss4 score = {energy:.2f}\")\n",
    "                            # self.save_output(out, energy)\n",
    "                            energies.append(energy)\n",
    "            else:\n",
    "                if isinstance(docked_ligand, list):\n",
    "                    docked_ligand = docked_ligand[0]\n",
    "                    \n",
    "                energy = self.extract_chemgauss4_scores(docked_ligand)\n",
    "                if energy is not None:\n",
    "                    self.save_output(out, energy)\n",
    "                    if isinstance(energy, list): \n",
    "                        for j, e in enumerate(energy):\n",
    "                            print(f\"Docked ligand {docked_ligand}, molecule {j+1}: Chemgauss4 score = {e:.2f}\")\n",
    "                            # self.save_output(out, e)\n",
    "                            energies.append(e)\n",
    "                    else: \n",
    "                        print(f\"Docked ligand {docked_ligand}, molecule {i+1}: Chemgauss4 score = {energy:.2f}\")\n",
    "                        # self.save_output(out, energy)\n",
    "                        energies.append(energy)\n",
    "                else:\n",
    "                    self.save_output(out, energy)\n",
    "                    energies = None\n",
    "                    raise ValueError(f\"{self.pdb_id}: No valid energy value available.\")            \n",
    "            # debug line\n",
    "            # print(energies)\n",
    "        except Exception as e:\n",
    "            print(f\"Error in {self.pdb_id}: {e}\")\n",
    "            out = f\"Error in {self.pdb_id}: {e}\"\n",
    "            energies = None\n",
    "            print(f\"{self.pdb_id}: No valid energy value available.\")\n",
    "            # self.save_output(out, energies)\n",
    "            raise e\n",
    "    \n",
    "        return energies"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Full Pipeline: From protein-ligand complex to dG scores \n",
    "Using the vina pipeline first:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# download the PDB file using the PDB ID\n",
    "# filename = download_pdb_file(pdb_id)\n",
    "# # split the protein and ligand from the PDB file\n",
    "# receptor_file, ligand_file = pdb_to_prot_lig(pdb_id, filename)\n",
    "# this is for when lig_files and prot_files are lists...\n",
    "# for lig, prot in tqdm(zip(lig_files, prot_files)):\n",
    "#         vina_process_lig_prot(lig, prot)\n",
    "\n",
    "def vina_process_lig_prot(lig_file, prot_file, complex_file, preserve_water=False, preserve_metal=True, csv_out_file='vina_docking_data.csv'):\n",
    "    '''Final function to process each protein and ligand pair, prepare them for docking, score the ligand and save the output files.\n",
    "    Both lig_file, prot_file and complex_file must be strings, not lists. If you want to process multiple ligands and proteins (i.e. LISTS), you should use this function in the following manner: \n",
    "\n",
    "    for lig, prot in tqdm(zip(lig_files, prot_files)):\n",
    "        vina_process_lig_prot(lig, prot, complex_file)\n",
    "        ...\n",
    "\n",
    "    The reference complex file needs to be a protein-ligand complex, with a docked ligand in the binding site of the protein of interest.\n",
    "    '''\n",
    "    # TODO: handling lists..?\n",
    "    if isinstance(lig_file, list) or isinstance(prot_file, list) or isinstance(complex_file, list):\n",
    "        raise ValueError(\"lig_file and prot_file must be strings, not lists. If you are trying to process multiple ligands and proteins, you should use this function in a loop.\")\n",
    "    \n",
    "    # check if any of the input is none\n",
    "    if lig_file is None or prot_file is None or complex_file is None:\n",
    "        raise ValueError(\"lig_file, prot_file and complex_file must not be None. You risk deleting everything in your current directory if any of these is None!\")\n",
    "\n",
    "    lig_name = os.path.splitext(lig_file)[0]\n",
    "    prot_name = os.path.splitext(prot_file)[0]\n",
    "    complex_name = os.path.splitext(complex_file)[0]\n",
    "\n",
    "    try:      \n",
    "        os.environ['OE_LICENSE'] = '/home/ian/oe_license.txt' # change this to your OE_LICENSE path\n",
    "\n",
    "        # prepare the ligand and receptor for Vina i.e. convert to pdbqt format\n",
    "        prep = DockingPrepper('.',\n",
    "                                lig_file=lig_file, \n",
    "                                prot_file=prot_file,\n",
    "                                pdb_id=complex_name,\n",
    "                                preserve_water=preserve_water,\n",
    "                                preserve_metal=preserve_metal) # this is not really the pdb_id but more of a basename for the files\n",
    "        # for vina preparation\n",
    "        prep.vina_process()\n",
    "                \n",
    "        # using the DockingScorer class to get the docking score\n",
    "        scorer = DockingScorer('.', \n",
    "                                lig_file, \n",
    "                                prot_file,\n",
    "                                complex_file=complex_file, # reference complex file, required for the vina process \n",
    "                                get_vina_poses=True, # this only works for vina, there's no choice for OE - LOL!\n",
    "                                save_out_file=True,\n",
    "                                from_pdb=False,\n",
    "                                csv_out_file=csv_out_file,\n",
    "                                receptor_DU=None) # None because this is for Vina, but for OE, this will be the receptor DU (required)\n",
    "\n",
    "        docking_score = scorer.vina_score_ligand()\n",
    "\n",
    "    except Exception as e:\n",
    "        # ------- START OF CODES: save the error message to the .out file, None as the energy score, and write both the energy and the error message to a csv file -------\n",
    "        # Set up logging\n",
    "        logging.basicConfig(filename=f'{lig_name}_{prot_name}error_log.txt', level=logging.ERROR, format='%(asctime)s - %(levelname)s - %(message)s')\n",
    "        logging.error(f\"Error processing {lig_file}: {str(e)}\")\n",
    "        print(f\"Skipping {lig_file} due to an error: {str(e)}\")\n",
    "        \n",
    "        # saving the error message to the .txt file and None as the energy score \n",
    "        energy = None\n",
    "        df = pd.DataFrame({'ligand_name': [lig_file], \n",
    "                                'protein_name': [prot_file], \n",
    "                                'computed_dG': [energy],\n",
    "                                'error_message' : [str(e)]}\n",
    "                                )\n",
    "    \n",
    "        if not os.path.isfile(csv_out_file):\n",
    "            df.to_csv(csv_out_file, index=True)\n",
    "        else: # else it exists so append without writing the header\n",
    "            df.to_csv(csv_out_file, mode='a', header=False, index=False)   \n",
    "        print(f\"Error message saved to {csv_out_file}\") \n",
    "        # -------END OF CODES: save the error message to the .out file, None as the energy score, and write both the energy and the error message to a csv file -------\n",
    "        pass\n",
    "\n",
    "        \n",
    "    new_dir = os.path.join(cwd, f'{lig_name}_{prot_name}')\n",
    "    os.makedirs(new_dir, exist_ok=True)\n",
    "    \n",
    "    # copy the files to the new directory, this will overwrite the files if they already exist\n",
    "    os.system(f'mv -f *{lig_name}* {new_dir}')\n",
    "    os.system(f'mv -f *{prot_name}* {new_dir}')\n",
    "    os.system(f'mv -f *{complex_name}* {new_dir}')\n",
    "\n",
    "    # if the new directory already exists, forcibly remove it so we can overwrite it\n",
    "    if os.path.exists(f'data/{new_dir}'):\n",
    "        os.system(f'rm -rf data/{new_dir}')\n",
    "    # move the output dir to the data storage dir 'data'\n",
    "    os.system(f'mv -f {new_dir} ./data/')\n",
    "    print(f\"Data saved in ./data/{new_dir}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### OpenEye Toolkits Pipeline\n",
    "\n",
    "With this pipeline, you need to provide an additional parameter, a protein-ligand complex, as a reference of where the active site is."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Use the following code if you need to download the PDB file using the PDB ID\n",
    "# download the PDB file using the PDB ID\n",
    "# filename = download_pdb_file(pdb_id)\n",
    "# # split the protein and ligand from the PDB file\n",
    "# receptor_file, ligand_file = pdb_to_prot_lig(pdb_id, filename)\n",
    "# this is for when lig_files and prot_files are lists...\n",
    "# for lig, prot in tqdm(zip(lig_files, prot_files)):\n",
    "#         vina_process_lig_prot(lig, prot)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "def oe_process_lig_prot(lig_file, prot_file, complex_file, preserve_water=False, preserve_metal=True, csv_out_file='oe_docking_data.csv'):\n",
    "    '''Final function to process each protein and ligand pair, prepare them for docking, score the ligand and save the output files.\n",
    "    Both lig_file and prot_file must be strings, not lists. If you want to process multiple ligands and proteins (i.e. LISTS), you should use this function in the following manner: \n",
    "    for lig, prot in tqdm(zip(lig_files, prot_files)):\n",
    "        oe_process_lig_prot(lig, prot)\n",
    "        ...\n",
    "    '''\n",
    "\n",
    "    # TODO: handling lists..?\n",
    "    if isinstance(lig_file, list) or isinstance(prot_file, list) or isinstance(complex_file, list):\n",
    "        raise ValueError(\"lig_file and prot_file must be strings, not lists. If you are trying to process multiple ligands and proteins, you should use this function in a loop.\")\n",
    "    \n",
    "    # check if any of the input is none\n",
    "    if lig_file is None or prot_file is None or complex_file is None:\n",
    "        raise ValueError(\"lig_file, prot_file and complex_file must not be None. You risk deleting everything in your current directory if any of these is None!\")\n",
    "    \n",
    "    lig_name = os.path.splitext(lig_file)[0]\n",
    "    prot_name = os.path.splitext(prot_file)[0]\n",
    "    complex_name = os.path.splitext(complex_file)[0]\n",
    "\n",
    "    try:\n",
    "        os.environ['OE_LICENSE'] = '/home/ian/oe_license.txt' # change this to your OE_LICENSE path\n",
    "        \n",
    "        # prepare the ligand and receptor\n",
    "        prep = DockingPrepper('.',\n",
    "                                lig_file=lig_file, \n",
    "                                prot_file=prot_file,\n",
    "                                pdb_id=complex_name,\n",
    "                                preserve_water=preserve_water,\n",
    "                                preserve_metal=preserve_metal) # this is not really the pdb_id but more of a basename for the files\n",
    "\n",
    "        # OE preparation\n",
    "        DU = prep.oe_process()\n",
    "        \n",
    "        # using the DockingScorer class to get the docking score\n",
    "        scorer = DockingScorer('.', \n",
    "                            lig_file, \n",
    "                            prot_file, \n",
    "                            save_out_file=True,\n",
    "                            from_pdb=False,\n",
    "                            complex_file=complex_file,\n",
    "                            csv_out_file=csv_out_file,\n",
    "                            receptor_DU=DU) # for OE, this will be the receptor DU produced from DockingPrepper.oe_process (required)\n",
    "                            \n",
    "        docking_score = scorer.oe_score_ligand()\n",
    "\n",
    "    except Exception as e:\n",
    "        # ------- START OF CODES: save the error message to the .out file, None as the energy score, and write both the energy and the error message to a csv file -------\n",
    "        # Set up logging\n",
    "        logging.basicConfig(filename=f'{lig_name}_{prot_name}error_log.txt', level=logging.ERROR, format='%(asctime)s - %(levelname)s - %(message)s')\n",
    "        logging.error(f\"Error processing {lig_file}: {str(e)}\")\n",
    "        print(f\"Skipping {lig_file} due to an error: {str(e)}\")\n",
    "        \n",
    "        # saving the error message to the .txt file and None as the energy score \n",
    "        energy = None\n",
    "        df = pd.DataFrame({'ligand_name': [lig_file], \n",
    "                                'protein_name': [prot_file], \n",
    "                                'Chemgauss4': [energy],\n",
    "                                'error_message' : [str(e)]}\n",
    "                                )\n",
    "    \n",
    "        if not os.path.isfile(csv_out_file):\n",
    "            df.to_csv(csv_out_file, index=True)\n",
    "        else: # else it exists so append without writing the header\n",
    "            df.to_csv(csv_out_file, mode='a', header=False, index=False)   \n",
    "        print(f\"Error message saved to {csv_out_file}\") \n",
    "        # -------END OF CODES: save the error message to the .out file, None as the energy score, and write both the energy and the error message to a csv file -------\n",
    "\n",
    "        pass  # This will skip the current lig and proceed with the next one\n",
    "\n",
    "    new_dir = os.path.join(cwd, f'{lig_name}_{prot_name}')\n",
    "    os.makedirs(new_dir, exist_ok=True)\n",
    "    \n",
    "    # copy the files to the new directory, this will overwrite the files if they already exist\n",
    "    os.system(f'mv -f *{lig_name}* {new_dir}')\n",
    "    os.system(f'mv -f *{prot_name}* {new_dir}')\n",
    "    os.system(f'mv -f *{complex_name}* {new_dir}')\n",
    "\n",
    "    # if the new directory already exists, forcibly remove it so we can overwrite it\n",
    "    if os.path.exists(f'data/{new_dir}'):\n",
    "        os.system(f'rm -rf data/{new_dir}')\n",
    "    # move the output dir to the data storage dir 'data'\n",
    "    os.system(f'mv -f {new_dir} ./data/')\n",
    "    print(f\"Data saved in ./data/{new_dir}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# If the input SDF file has more than 1 molecules, then it needs to be split. For example, SDF files generated by diffusion model typically have more than 1 molecules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "def split_sdf(input_sdf):\n",
    "    # Load the multi-molecule SDF file\n",
    "    supplier = Chem.SDMolSupplier(input_sdf)\n",
    "    output_name = input_sdf.split('.')[0]\n",
    "    \n",
    "    # Count the number of molecules\n",
    "    molecules = [mol for mol in supplier if mol is not None]\n",
    "    num_molecules = len(molecules)\n",
    "    \n",
    "    output = []\n",
    "    # Check if the SDF file contains more than one molecule\n",
    "    if num_molecules > 1:\n",
    "        # Loop over each molecule and write it to a separate file\n",
    "        for i, mol in enumerate(molecules):\n",
    "            output_sdf = f\"{output_name}_{i+1}.sdf\"\n",
    "            output.append(output_sdf)\n",
    "\n",
    "            writer = SDWriter(output_sdf)\n",
    "            writer.write(mol)\n",
    "            writer.close()\n",
    "        print(f\"{num_molecules} molecules have been successfully separated into individual SDF files.\")\n",
    "    else:\n",
    "        print(f\"The SDF file {input_sdf} contains one or no molecules. No splitting is required.\")\n",
    "        output = str(input_sdf)\n",
    "    return output\n",
    "\n",
    "# Replace with the actual path to your SDF file\n",
    "# input_sdf = \"molecules_bx_2024_06_18_154410_1.sdf\"\n",
    "# lig_file = split_sdf(input_sdf)\n",
    "# print(lig_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Loading the leakyPDB dataset - we use this dataset to test out our pipeline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: 'leakypdb_test.csv'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[20], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m# loading LeakyPDB dataset\u001b[39;00m\n\u001b[0;32m----> 2\u001b[0m leakypdb_test \u001b[38;5;241m=\u001b[39m \u001b[43mpd\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread_csv\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mleakypdb_test.csv\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m      3\u001b[0m leakypdb_test\u001b[38;5;241m.\u001b[39mrename(columns\u001b[38;5;241m=\u001b[39m{\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mUnnamed: 0\u001b[39m\u001b[38;5;124m'\u001b[39m: \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mpdb_id\u001b[39m\u001b[38;5;124m'\u001b[39m,\n\u001b[1;32m      4\u001b[0m                           \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mheader\u001b[39m\u001b[38;5;124m'\u001b[39m: \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mprotein_family\u001b[39m\u001b[38;5;124m'\u001b[39m, \n\u001b[1;32m      5\u001b[0m                           \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mseq\u001b[39m\u001b[38;5;124m'\u001b[39m:\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mprotein_sequence\u001b[39m\u001b[38;5;124m'\u001b[39m,\n\u001b[1;32m      6\u001b[0m                           \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mkd/ki\u001b[39m\u001b[38;5;124m'\u001b[39m:\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mbinding_affinity\u001b[39m\u001b[38;5;124m'\u001b[39m,\n\u001b[1;32m      7\u001b[0m                           \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mvalue\u001b[39m\u001b[38;5;124m'\u001b[39m:\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mpkd/pki\u001b[39m\u001b[38;5;124m'\u001b[39m}, inplace\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m)\n\u001b[1;32m      8\u001b[0m leakypdb_test[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mpdb_id\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m leakypdb_test[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mpdb_id\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m.\u001b[39mstr\u001b[38;5;241m.\u001b[39mupper()\n",
      "File \u001b[0;32m~/micromamba/envs/asapdiscovery/lib/python3.10/site-packages/pandas/util/_decorators.py:211\u001b[0m, in \u001b[0;36mdeprecate_kwarg.<locals>._deprecate_kwarg.<locals>.wrapper\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    209\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    210\u001b[0m         kwargs[new_arg_name] \u001b[38;5;241m=\u001b[39m new_arg_value\n\u001b[0;32m--> 211\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfunc\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/micromamba/envs/asapdiscovery/lib/python3.10/site-packages/pandas/util/_decorators.py:331\u001b[0m, in \u001b[0;36mdeprecate_nonkeyword_arguments.<locals>.decorate.<locals>.wrapper\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    325\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(args) \u001b[38;5;241m>\u001b[39m num_allow_args:\n\u001b[1;32m    326\u001b[0m     warnings\u001b[38;5;241m.\u001b[39mwarn(\n\u001b[1;32m    327\u001b[0m         msg\u001b[38;5;241m.\u001b[39mformat(arguments\u001b[38;5;241m=\u001b[39m_format_argument_list(allow_args)),\n\u001b[1;32m    328\u001b[0m         \u001b[38;5;167;01mFutureWarning\u001b[39;00m,\n\u001b[1;32m    329\u001b[0m         stacklevel\u001b[38;5;241m=\u001b[39mfind_stack_level(),\n\u001b[1;32m    330\u001b[0m     )\n\u001b[0;32m--> 331\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfunc\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/micromamba/envs/asapdiscovery/lib/python3.10/site-packages/pandas/io/parsers/readers.py:950\u001b[0m, in \u001b[0;36mread_csv\u001b[0;34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, squeeze, prefix, mangle_dupe_cols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, encoding_errors, dialect, error_bad_lines, warn_bad_lines, on_bad_lines, delim_whitespace, low_memory, memory_map, float_precision, storage_options)\u001b[0m\n\u001b[1;32m    935\u001b[0m kwds_defaults \u001b[38;5;241m=\u001b[39m _refine_defaults_read(\n\u001b[1;32m    936\u001b[0m     dialect,\n\u001b[1;32m    937\u001b[0m     delimiter,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    946\u001b[0m     defaults\u001b[38;5;241m=\u001b[39m{\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdelimiter\u001b[39m\u001b[38;5;124m\"\u001b[39m: \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m,\u001b[39m\u001b[38;5;124m\"\u001b[39m},\n\u001b[1;32m    947\u001b[0m )\n\u001b[1;32m    948\u001b[0m kwds\u001b[38;5;241m.\u001b[39mupdate(kwds_defaults)\n\u001b[0;32m--> 950\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_read\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilepath_or_buffer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/micromamba/envs/asapdiscovery/lib/python3.10/site-packages/pandas/io/parsers/readers.py:605\u001b[0m, in \u001b[0;36m_read\u001b[0;34m(filepath_or_buffer, kwds)\u001b[0m\n\u001b[1;32m    602\u001b[0m _validate_names(kwds\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mnames\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m))\n\u001b[1;32m    604\u001b[0m \u001b[38;5;66;03m# Create the parser.\u001b[39;00m\n\u001b[0;32m--> 605\u001b[0m parser \u001b[38;5;241m=\u001b[39m \u001b[43mTextFileReader\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilepath_or_buffer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    607\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m chunksize \u001b[38;5;129;01mor\u001b[39;00m iterator:\n\u001b[1;32m    608\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m parser\n",
      "File \u001b[0;32m~/micromamba/envs/asapdiscovery/lib/python3.10/site-packages/pandas/io/parsers/readers.py:1442\u001b[0m, in \u001b[0;36mTextFileReader.__init__\u001b[0;34m(self, f, engine, **kwds)\u001b[0m\n\u001b[1;32m   1439\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39moptions[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhas_index_names\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m kwds[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhas_index_names\u001b[39m\u001b[38;5;124m\"\u001b[39m]\n\u001b[1;32m   1441\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles: IOHandles \u001b[38;5;241m|\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[0;32m-> 1442\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_engine \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_make_engine\u001b[49m\u001b[43m(\u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mengine\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/micromamba/envs/asapdiscovery/lib/python3.10/site-packages/pandas/io/parsers/readers.py:1735\u001b[0m, in \u001b[0;36mTextFileReader._make_engine\u001b[0;34m(self, f, engine)\u001b[0m\n\u001b[1;32m   1733\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mb\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m mode:\n\u001b[1;32m   1734\u001b[0m         mode \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mb\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m-> 1735\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles \u001b[38;5;241m=\u001b[39m \u001b[43mget_handle\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1736\u001b[0m \u001b[43m    \u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1737\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmode\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1738\u001b[0m \u001b[43m    \u001b[49m\u001b[43mencoding\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mencoding\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1739\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcompression\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mcompression\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1740\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmemory_map\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mmemory_map\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1741\u001b[0m \u001b[43m    \u001b[49m\u001b[43mis_text\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mis_text\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1742\u001b[0m \u001b[43m    \u001b[49m\u001b[43merrors\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mencoding_errors\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mstrict\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1743\u001b[0m \u001b[43m    \u001b[49m\u001b[43mstorage_options\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mstorage_options\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1744\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1745\u001b[0m \u001b[38;5;28;01massert\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m   1746\u001b[0m f \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles\u001b[38;5;241m.\u001b[39mhandle\n",
      "File \u001b[0;32m~/micromamba/envs/asapdiscovery/lib/python3.10/site-packages/pandas/io/common.py:856\u001b[0m, in \u001b[0;36mget_handle\u001b[0;34m(path_or_buf, mode, encoding, compression, memory_map, is_text, errors, storage_options)\u001b[0m\n\u001b[1;32m    851\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(handle, \u001b[38;5;28mstr\u001b[39m):\n\u001b[1;32m    852\u001b[0m     \u001b[38;5;66;03m# Check whether the filename is to be opened in binary mode.\u001b[39;00m\n\u001b[1;32m    853\u001b[0m     \u001b[38;5;66;03m# Binary mode does not support 'encoding' and 'newline'.\u001b[39;00m\n\u001b[1;32m    854\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m ioargs\u001b[38;5;241m.\u001b[39mencoding \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mb\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m ioargs\u001b[38;5;241m.\u001b[39mmode:\n\u001b[1;32m    855\u001b[0m         \u001b[38;5;66;03m# Encoding\u001b[39;00m\n\u001b[0;32m--> 856\u001b[0m         handle \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mopen\u001b[39;49m\u001b[43m(\u001b[49m\n\u001b[1;32m    857\u001b[0m \u001b[43m            \u001b[49m\u001b[43mhandle\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    858\u001b[0m \u001b[43m            \u001b[49m\u001b[43mioargs\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmode\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    859\u001b[0m \u001b[43m            \u001b[49m\u001b[43mencoding\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mioargs\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mencoding\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    860\u001b[0m \u001b[43m            \u001b[49m\u001b[43merrors\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43merrors\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    861\u001b[0m \u001b[43m            \u001b[49m\u001b[43mnewline\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    862\u001b[0m \u001b[43m        \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    863\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    864\u001b[0m         \u001b[38;5;66;03m# Binary mode\u001b[39;00m\n\u001b[1;32m    865\u001b[0m         handle \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mopen\u001b[39m(handle, ioargs\u001b[38;5;241m.\u001b[39mmode)\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: 'leakypdb_test.csv'"
     ]
    }
   ],
   "source": [
    "# loading LeakyPDB dataset\n",
    "leakypdb_test = pd.read_csv('leakypdb_test.csv')\n",
    "leakypdb_test.rename(columns={'Unnamed: 0': 'pdb_id',\n",
    "                          'header': 'protein_family', \n",
    "                          'seq':'protein_sequence',\n",
    "                          'kd/ki':'binding_affinity',\n",
    "                          'value':'pkd/pki'}, inplace=True)\n",
    "leakypdb_test['pdb_id'] = leakypdb_test['pdb_id'].str.upper()\n",
    "leakypdb_test\n",
    "#TODO: kd/ki column may not be all experimentally determined, some may be predicted, need to check this"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'leakypdb_test' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[21], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m leakypdb_ids\u001b[38;5;241m=\u001b[39m\u001b[43mleakypdb_test\u001b[49m[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mpdb_id\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m.\u001b[39mtolist()\n\u001b[1;32m      2\u001b[0m leakypdb_ids \u001b[38;5;241m=\u001b[39m [leakypdb_id\u001b[38;5;241m.\u001b[39mupper() \u001b[38;5;28;01mfor\u001b[39;00m leakypdb_id \u001b[38;5;129;01min\u001b[39;00m leakypdb_ids]\n\u001b[1;32m      3\u001b[0m leakypdb_ids\n",
      "\u001b[0;31mNameError\u001b[0m: name 'leakypdb_test' is not defined"
     ]
    }
   ],
   "source": [
    "leakypdb_ids=leakypdb_test['pdb_id'].tolist()\n",
    "leakypdb_ids = [leakypdb_id.upper() for leakypdb_id in leakypdb_ids]\n",
    "leakypdb_ids"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Testing the vina pipeline\n",
    "<font color='orange'>Testing the vina pipeline without try except</font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Preparing the ligand and protein files using OpenEye Toolkits...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "DPI: 0.12, RFree: 0.23, Resolution: 1.65\n",
      "Processing BU # 1 with title: ---, chains AB, alt: A\n",
      "Warning: OEPlaceHydrogens: mol DesignUnit Components has heavy atoms ' O   HOH 231 A B 1  ' and ' O   HOH 231 B B 1  ' too close 0.11 < 0.5A\n",
      "Warning: Unable to add protons to DesignUnit: ---(AB)altA__DU__biounit\n",
      "Warning: Due to lack of protons, unable to add partial charges and radii to DesignUnit: ---(AB)altA__DU__biounit\n",
      "Processing BU # 2 with title: ---, chains AB, alt: B\n",
      "Warning: OEPlaceHydrogens: mol DesignUnit Components has heavy atoms ' O   HOH 231 A B 1  ' and ' O   HOH 231 B B 1  ' too close 0.11 < 0.5A\n",
      "Warning: Unable to add protons to DesignUnit: ---(AB)altB__DU__biounit\n",
      "Warning: Due to lack of protons, unable to add partial charges and radii to DesignUnit: ---(AB)altB__DU__biounit\n",
      "Warning: Hydrogens have not been placed, can not add OEInteractionHints\n",
      "Warning: Unable to add interactions to Design Unit: ---(AB)altA > LIG(A-1101)\n",
      "Warning: Hydrogens have not been placed, can not add OEInteractionHints\n",
      "Warning: Unable to add interactions to Design Unit: ---(AB)altA > LIG(B-1101)\n",
      "Warning: Hydrogens have not been placed, can not add OEInteractionHints\n",
      "Warning: Unable to add interactions to Design Unit: ---(AB)altB > LIG(A-1101)\n",
      "Warning: Hydrogens have not been placed, can not add OEInteractionHints\n",
      "Warning: Unable to add interactions to Design Unit: ---(AB)altB > LIG(B-1101)\n",
      "Skipping redundant DU with alts outside the site of interest, renaming existing to collapse alts\n",
      "Discarding redundant alt DU with title ---(AB)altB > LIG(A-1101)\n",
      "Skipping redundant DU with alts outside the site of interest, renaming existing to collapse alts\n",
      "Discarding redundant alt DU with title ---(AB)altB > LIG(B-1101)\n",
      "Warning: Could not find alignment that matches constraints, try again without constraints\n",
      "Superposition - RMSD: 0.00, Ref: , Fit: , SeqScore: 3102\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Design unit was successfully made for Mpro-x0072_0.pdb, output is saved to ['Mpro-x0072_0_DU_0.oedu', 'Mpro-x0072_0_DU_1.oedu'].\n",
      "The receptor design unit file has been saved as Mpro-x0072_0_DU_0_receptor.oedu\n",
      "The receptor design unit file has been saved as Mpro-x0072_0_DU_1_receptor.oedu\n",
      "cleaning \n",
      "docking \n",
      "cleaning \n",
      "docking \n",
      "The docked ligand file has been saved as ['Mpro-x0072_0_DU_0_receptor_Updated_Mpro_data_2044_docked.sdf', 'Mpro-x0072_0_DU_1_receptor_Updated_Mpro_data_2044_docked.sdf']\n",
      "RDKit approach failed: Error: Invalid molecule found, try the regex-based approach., trying regex-based approach...\n",
      "Output saved as Mpro-protein_Updated_Mpro_data_2044_0.out\n",
      "\n",
      "Data saved to oe_docking_data.csv\n",
      "Docked ligand Mpro-x0072_0_DU_0_receptor_Updated_Mpro_data_2044_docked.sdf, molecule 1: Chemgauss4 score = -1.02\n",
      "RDKit approach failed: Error: Invalid molecule found, try the regex-based approach., trying regex-based approach...\n",
      "Output saved as Mpro-protein_Updated_Mpro_data_2044_0.out\n",
      "\n",
      "Data saved to oe_docking_data.csv\n",
      "Docked ligand Mpro-x0072_0_DU_1_receptor_Updated_Mpro_data_2044_docked.sdf, molecule 1: Chemgauss4 score = -1.36\n",
      "Preparing the ligand and protein files using OpenEye Toolkits...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[22:09:36] Explicit valence for atom # 4 N, 4, is greater than permitted\n",
      "[22:09:36] ERROR: Could not sanitize molecule ending on line 162\n",
      "[22:09:36] ERROR: Explicit valence for atom # 4 N, 4, is greater than permitted\n",
      "[22:09:36] Explicit valence for atom # 4 N, 4, is greater than permitted\n",
      "[22:09:36] ERROR: Could not sanitize molecule ending on line 162\n",
      "[22:09:36] ERROR: Explicit valence for atom # 4 N, 4, is greater than permitted\n",
      "DPI: 0.12, RFree: 0.23, Resolution: 1.65\n",
      "Processing BU # 1 with title: ---, chains AB, alt: A\n",
      "Warning: OEPlaceHydrogens: mol DesignUnit Components has heavy atoms ' O   HOH 231 A B 1  ' and ' O   HOH 231 B B 1  ' too close 0.11 < 0.5A\n",
      "Warning: Unable to add protons to DesignUnit: ---(AB)altA__DU__biounit\n",
      "Warning: Due to lack of protons, unable to add partial charges and radii to DesignUnit: ---(AB)altA__DU__biounit\n",
      "Processing BU # 2 with title: ---, chains AB, alt: B\n",
      "Warning: OEPlaceHydrogens: mol DesignUnit Components has heavy atoms ' O   HOH 231 A B 1  ' and ' O   HOH 231 B B 1  ' too close 0.11 < 0.5A\n",
      "Warning: Unable to add protons to DesignUnit: ---(AB)altB__DU__biounit\n",
      "Warning: Due to lack of protons, unable to add partial charges and radii to DesignUnit: ---(AB)altB__DU__biounit\n",
      "Warning: Hydrogens have not been placed, can not add OEInteractionHints\n",
      "Warning: Unable to add interactions to Design Unit: ---(AB)altA > LIG(A-1101)\n",
      "Warning: Hydrogens have not been placed, can not add OEInteractionHints\n",
      "Warning: Unable to add interactions to Design Unit: ---(AB)altA > LIG(B-1101)\n",
      "Warning: Hydrogens have not been placed, can not add OEInteractionHints\n",
      "Warning: Unable to add interactions to Design Unit: ---(AB)altB > LIG(A-1101)\n",
      "Warning: Hydrogens have not been placed, can not add OEInteractionHints\n",
      "Warning: Unable to add interactions to Design Unit: ---(AB)altB > LIG(B-1101)\n",
      "Skipping redundant DU with alts outside the site of interest, renaming existing to collapse alts\n",
      "Discarding redundant alt DU with title ---(AB)altB > LIG(A-1101)\n",
      "Skipping redundant DU with alts outside the site of interest, renaming existing to collapse alts\n",
      "Discarding redundant alt DU with title ---(AB)altB > LIG(B-1101)\n",
      "Warning: Could not find alignment that matches constraints, try again without constraints\n",
      "Superposition - RMSD: 0.00, Ref: , Fit: , SeqScore: 3102\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Design unit was successfully made for Mpro-x0072_0.pdb, output is saved to ['Mpro-x0072_0_DU_0.oedu', 'Mpro-x0072_0_DU_1.oedu'].\n",
      "The receptor design unit file has been saved as Mpro-x0072_0_DU_0_receptor.oedu\n",
      "The receptor design unit file has been saved as Mpro-x0072_0_DU_1_receptor.oedu\n",
      "cleaning \n",
      "docking \n",
      "cleaning \n",
      "docking \n",
      "The docked ligand file has been saved as ['Mpro-x0072_0_DU_0_receptor_Updated_Mpro_data_1687_docked.sdf', 'Mpro-x0072_0_DU_1_receptor_Updated_Mpro_data_1687_docked.sdf']\n",
      "RDKit approach failed: Error: Invalid molecule found, try the regex-based approach., trying regex-based approach...\n",
      "Output saved as Mpro-protein_Updated_Mpro_data_1687_0.out\n",
      "\n",
      "Data saved to oe_docking_data.csv\n",
      "Docked ligand Mpro-x0072_0_DU_0_receptor_Updated_Mpro_data_1687_docked.sdf, molecule 1: Chemgauss4 score = -8.06\n",
      "RDKit approach failed: Error: Invalid molecule found, try the regex-based approach., trying regex-based approach...\n",
      "Output saved as Mpro-protein_Updated_Mpro_data_1687_0.out\n",
      "\n",
      "Data saved to oe_docking_data.csv\n",
      "Docked ligand Mpro-x0072_0_DU_1_receptor_Updated_Mpro_data_1687_docked.sdf, molecule 1: Chemgauss4 score = -7.87\n",
      "Preparing the ligand and protein files using OpenEye Toolkits...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[22:11:12] Explicit valence for atom # 22 N, 4, is greater than permitted\n",
      "[22:11:12] ERROR: Could not sanitize molecule ending on line 99\n",
      "[22:11:12] ERROR: Explicit valence for atom # 22 N, 4, is greater than permitted\n",
      "[22:11:12] Explicit valence for atom # 22 N, 4, is greater than permitted\n",
      "[22:11:12] ERROR: Could not sanitize molecule ending on line 99\n",
      "[22:11:12] ERROR: Explicit valence for atom # 22 N, 4, is greater than permitted\n",
      "DPI: 0.12, RFree: 0.23, Resolution: 1.65\n",
      "Processing BU # 1 with title: ---, chains AB, alt: A\n",
      "Warning: OEPlaceHydrogens: mol DesignUnit Components has heavy atoms ' O   HOH 231 A B 1  ' and ' O   HOH 231 B B 1  ' too close 0.11 < 0.5A\n",
      "Warning: Unable to add protons to DesignUnit: ---(AB)altA__DU__biounit\n",
      "Warning: Due to lack of protons, unable to add partial charges and radii to DesignUnit: ---(AB)altA__DU__biounit\n",
      "Processing BU # 2 with title: ---, chains AB, alt: B\n"
     ]
    }
   ],
   "source": [
    "mpro_oe_water = pd.read_csv('results/mpro_oe/with_water/mpro_oe_w_water_data.csv')\n",
    "mpro_oe_water_error_ligands = mpro_oe_water[mpro_oe_water['error_message'].notnull()]['ligand_name'].tolist()\n",
    "import glob\n",
    "all_mpro_ligands = []\n",
    "for pattern in mpro_oe_water_error_ligands:\n",
    "    matched_files = glob.glob(f'Updated_Mpro_ligands/{pattern}')\n",
    "    all_mpro_ligands.extend(matched_files)\n",
    "\n",
    "for ligand_file in all_mpro_ligands:\n",
    "    \n",
    "    os.system(f'cp {ligand_file} .')\n",
    "    # ligand_file name without path name\n",
    "    ligand_file = os.path.basename(ligand_file)\n",
    "\n",
    "    receptor_file = 'Mpro-protein.pdb'\n",
    "    complex_file = 'Mpro-x0072_0.pdb'\n",
    "    os.system(f'cp Mpro_complexes/{receptor_file} .')\n",
    "    os.system(f'cp Mpro_complexes/{complex_file} .')\n",
    "\n",
    "    # receptor_file = '6YNQ_receptor_fixed.pdb'\n",
    "    # complex_file = '6YNQ.pdb'             \n",
    "    # os.system(f'cp Mpro_active_site/{complex_file} .')\n",
    "    # os.system(f'cp Mpro_active_site/{receptor_file} .')\n",
    "\n",
    "    \n",
    "\n",
    "    # oe_process_lig_prot(ligand_file, \n",
    "    #                     receptor_file, \n",
    "    #                     complex_file, \n",
    "    #                     preserve_water=True, \n",
    "    #                     preserve_metal=True, \n",
    "    #                     csv_out_file='mpro_oe_w_water_data.csv')\n",
    "\n",
    "# def oe_process_lig_prot(lig_file, prot_file, complex_file, preserve_water=False, preserve_metal=False, csv_out_file='oe_docking_data.csv'):\n",
    "#     '''Final function to process each protein and ligand pair, prepare them for docking, score the ligand and save the output files.\n",
    "#     Both lig_file and prot_file must be strings, not lists. If you want to process multiple ligands and proteins (i.e. LISTS), you should use this function in the following manner: \n",
    "#     for lig, prot in tqdm(zip(lig_files, prot_files)):\n",
    "#         oe_process_lig_prot(lig, prot)\n",
    "#         ...\n",
    "#     '''\n",
    "    lig_file = ligand_file\n",
    "    prot_file = receptor_file\n",
    "    complex_file = complex_file\n",
    "    preserve_water=True \n",
    "    preserve_metal=True \n",
    "    csv_out_file='oe_docking_data.csv'\n",
    "\n",
    "    # TODO: handling lists..?\n",
    "    if isinstance(lig_file, list) or isinstance(prot_file, list) or isinstance(complex_file, list):\n",
    "        raise ValueError(\"lig_file and prot_file must be strings, not lists. If you are trying to process multiple ligands and proteins, you should use this function in a loop.\")\n",
    "    \n",
    "    # check if any of the input is none\n",
    "    if lig_file is None or prot_file is None or complex_file is None:\n",
    "        raise ValueError(\"lig_file, prot_file and complex_file must not be None. You risk deleting everything in your current directory if any of these is None!\")\n",
    "    \n",
    "    lig_name = os.path.splitext(lig_file)[0]\n",
    "    prot_name = os.path.splitext(prot_file)[0]\n",
    "    complex_name = os.path.splitext(complex_file)[0]\n",
    "\n",
    "    # try:\n",
    "    os.environ['OE_LICENSE'] = '/home/ian/oe_license.txt' # change this to your OE_LICENSE path\n",
    "    \n",
    "    # prepare the ligand and receptor\n",
    "    prep = DockingPrepper('.',\n",
    "                            lig_file=lig_file, \n",
    "                            prot_file=prot_file,\n",
    "                            pdb_id=complex_name,\n",
    "                            preserve_water=preserve_water,\n",
    "                            preserve_metal=preserve_metal) # this is not really the pdb_id but more of a basename for the files\n",
    "\n",
    "    # OE preparation\n",
    "    DU = prep.oe_process()\n",
    "    \n",
    "    # using the DockingScorer class to get the docking score\n",
    "    scorer = DockingScorer('.', \n",
    "                        lig_file, \n",
    "                        prot_file, \n",
    "                        save_out_file=True,\n",
    "                        from_pdb=False,\n",
    "                        complex_file=complex_file,\n",
    "                        csv_out_file=csv_out_file,\n",
    "                        receptor_DU=DU) # for OE, this will be the receptor DU produced from DockingPrepper.oe_process (required)\n",
    "                        \n",
    "    docking_score = scorer.oe_score_ligand()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "1 molecule converted\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The ligand file has been converted to L3001_0.pdb\n",
      "The ligand file has been saved as L3001_0.pdbqt\n",
      "WARNING! 5S9L_receptor_w_water_metal has 144  atoms with alternate locations named ['A', 'B', 'C']!\n",
      "\n",
      "WARNING! using alternate location A\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A: CA901:CA\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A: ZN907:ZN\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A: CA909:CA\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1001:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1002:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1003:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1004:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1005:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1006:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1007:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1008:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1009:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1010:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1011:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1012:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1013:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1014:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1015:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1016:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1017:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1018:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1019:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1020:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1021:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1022:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1023:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1024:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1025:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1026:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1027:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1028:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1029:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1030:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1031:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1032:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1033:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1034:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1035:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1036:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1037:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1038:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1039:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1040:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1041:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1042:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1043:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1044:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1045:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1046:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1047:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1048:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1049:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1050:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1051:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1052:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1053:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1054:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1055:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1056:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1057:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1058:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1059:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1060:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1061:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1062:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1063:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1064:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1065:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1066:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1067:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1068:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1069:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1070:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1071:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1072:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1073:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1074:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1075:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1076:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1077:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1078:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1079:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1080:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1081:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1082:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1083:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1084:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1085:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1086:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1087:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1088:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1089:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1090:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1091:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1092:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1093:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1094:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1095:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1096:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1097:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1098:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1099:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1100:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1101:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1102:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1103:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1104:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1105:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1106:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1107:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1108:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1109:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1110:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1111:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1112:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1113:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1114:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1115:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1116:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1117:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1118:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1119:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1120:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1121:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1122:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1123:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1124:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1125:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1126:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1127:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1128:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1129:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1130:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1131:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1132:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1133:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1134:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1135:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1136:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1137:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1138:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1139:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1140:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1141:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1142:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1143:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1144:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1145:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1146:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1147:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1148:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1149:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1150:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1151:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1152:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1153:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1154:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1155:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1156:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1157:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1158:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1159:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1160:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1161:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1162:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1163:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1164:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1165:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1166:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1167:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1168:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1169:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1170:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1171:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1172:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1173:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1174:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1175:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1176:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1177:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1178:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1179:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1180:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1181:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1182:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1183:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1184:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1185:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1186:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1187:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1188:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1189:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1190:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1191:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1192:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1193:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1194:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1195:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1196:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1197:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1198:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1199:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1200:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1201:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1202:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1203:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1204:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1205:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1206:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1207:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1208:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1209:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1210:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1211:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1212:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1213:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1214:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1215:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1216:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1217:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1218:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1219:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1220:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1221:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1222:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1223:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1224:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1225:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1226:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1227:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1228:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1229:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1230:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1231:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1232:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1233:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1234:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1235:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1236:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1237:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1238:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1239:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1240:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1241:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1242:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1243:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1244:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1245:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1246:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1247:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1248:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1249:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1250:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1251:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1252:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1253:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1254:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1255:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1256:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1257:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1258:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1259:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1260:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1261:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1262:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1263:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1264:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1265:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1266:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1267:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1268:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1269:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1270:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1271:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1272:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1273:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1274:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1275:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1276:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1277:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1278:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1279:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1280:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1281:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1282:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1283:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1284:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1285:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1286:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1287:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1288:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1289:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1290:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1291:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1292:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1293:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1294:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1295:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1296:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1297:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1298:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1299:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1300:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1301:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1302:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1303:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1304:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1305:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1306:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1307:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1308:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1309:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1310:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1311:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1312:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1313:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1314:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1315:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1316:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1317:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1318:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1319:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1320:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1321:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1322:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1323:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1324:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1325:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1326:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1327:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1328:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1329:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1330:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1331:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1332:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1333:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1334:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1335:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1336:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1337:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1338:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1339:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1340:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1341:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1342:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1343:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1344:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1345:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1346:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1347:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1348:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1349:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1350:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1351:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1352:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1353:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1354:O@A\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1355:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1356:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1357:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1358:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1359:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1360:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1361:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1362:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1363:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1364:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1365:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1366:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1367:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1368:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1369:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1370:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1371:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1372:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1373:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1374:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1375:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1376:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1377:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1378:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1379:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1380:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1381:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1382:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1383:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1384:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1385:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1386:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1387:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1388:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1389:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1390:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1391:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1392:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1393:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1394:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1395:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1396:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1397:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1398:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1399:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1400:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1401:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1402:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1403:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1404:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1405:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1406:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1407:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1408:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1409:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1410:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1411:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1412:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1413:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1414:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1415:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1416:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1417:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1418:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1419:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1420:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1421:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1422:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1423:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1424:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1426:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1427:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1428:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1429:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1430:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1431:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1432:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1433:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1434:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1435:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1436:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1437:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1438:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1439:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1440:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1441:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1442:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1443:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1444:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1445:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1446:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1447:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1448:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1449:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1450:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1451:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1452:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1453:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1454:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1455:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1456:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1457:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1458:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1459:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1460:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1461:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1462:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1463:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1464:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1465:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1466:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1467:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1468:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1469:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1470:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1471:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1472:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1473:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1474:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1475:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1476:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1477:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1478:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1479:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1480:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1481:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1482:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1483:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1484:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1485:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1486:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1487:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1488:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1489:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1490:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1491:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1492:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1493:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1494:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1495:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1496:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1497:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1498:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1499:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1500:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1501:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1502:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1503:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1504:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1505:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1506:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1507:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1508:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_w_water_metal:A:HOH1509:O\n",
      "The protein file 5S9L_receptor_w_water_metal.pdb has been converted to pdbqt format and saved in 5S9L_receptor_w_water_metal.pdbqt\n",
      "Extracting ligand from 5S9L.pdb...\n",
      "Reading receptor from 5S9L.pdb...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "DPI: 0.12, RFree: 0.19, Resolution: 1.90\n",
      "Processing BU # 1 with title: ISOFORM 2 OF ECTONUCLEOTIDE, chains AB\n",
      "Warning: No rotamer of residue: GLU 573   A 1   passed a clash test, cluster will be skipped\n",
      "Found unresolved N-terminal with 23 residues before TRP 51   A 1  , with sequence FTASRIKRAEWDEGPPTVLSDSP\n",
      "Found 4 residue gap between LYS 463   A 1   and CYS 468   A 1  , with sequence PSGK\n",
      "Found 2 residue gap between GLU 573   A 1   and ASN 576   A 1  , with sequence PK\n",
      "Found unresolved C-terminal with 12 residues after GLU 861   A 1  , with sequence IGGRHHHHHHHH\n",
      "Warning: There was a problem building some missing pieces, built as much as was possible\n",
      "   Falling back to charging protein with OEMMFF94Charges\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Error in 5S9L: Command ./vina/vina_1.2.5_linux_x86_64 --config L3001_0_5S9L_receptor_w_water_metal_config.txt failed: \n",
      "\n",
      "\n",
      "PDBQT parsing error: Coordinate \"7  65.14\" is not valid.\n",
      " > ATOM   3946  HH1A1ARG A 450       8.039   7.307  65.141  1.00  0.00     0.174 HD\n",
      "\n",
      "out:  \n",
      "err:  \n",
      "Error in 5S9L: Command ./vina/vina_1.2.5_linux_x86_64 --config L3001_0_5S9L_receptor_w_water_metal_config.txt failed: \n",
      "\n",
      "\n",
      "PDBQT parsing error: Coordinate \"7  65.14\" is not valid.\n",
      " > ATOM   3946  HH1A1ARG A 450       8.039   7.307  65.141  1.00  0.00     0.174 HD\n",
      "\n"
     ]
    },
    {
     "ename": "CommandExecuteError",
     "evalue": "Command ./vina/vina_1.2.5_linux_x86_64 --config L3001_0_5S9L_receptor_w_water_metal_config.txt failed: \n\n\nPDBQT parsing error: Coordinate \"7  65.14\" is not valid.\n > ATOM   3946  HH1A1ARG A 450       8.039   7.307  65.141  1.00  0.00     0.174 HD\n",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mCommandExecuteError\u001b[0m                       Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[52], line 47\u001b[0m\n\u001b[1;32m     36\u001b[0m \u001b[38;5;66;03m# using the DockingScorer class to get the docking score\u001b[39;00m\n\u001b[1;32m     37\u001b[0m scorer \u001b[38;5;241m=\u001b[39m DockingScorer(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m.\u001b[39m\u001b[38;5;124m'\u001b[39m, \n\u001b[1;32m     38\u001b[0m                         lig_file, \n\u001b[1;32m     39\u001b[0m                         prot_file,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     44\u001b[0m                         csv_out_file\u001b[38;5;241m=\u001b[39mcsv_out_file,\n\u001b[1;32m     45\u001b[0m                         receptor_DU\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m) \u001b[38;5;66;03m# None because this is for Vina, but for OE, this will be the receptor DU (required)\u001b[39;00m\n\u001b[0;32m---> 47\u001b[0m docking_score \u001b[38;5;241m=\u001b[39m \u001b[43mscorer\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mvina_score_ligand\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     49\u001b[0m new_dir_name \u001b[38;5;241m=\u001b[39m \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mlig_name\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m_\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mprot_name\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m'\u001b[39m\n\u001b[1;32m     50\u001b[0m new_dir \u001b[38;5;241m=\u001b[39m os\u001b[38;5;241m.\u001b[39mpath\u001b[38;5;241m.\u001b[39mjoin(cwd, new_dir_name)\n",
      "Cell \u001b[0;32mIn[39], line 244\u001b[0m, in \u001b[0;36mDockingScorer.vina_score_ligand\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    242\u001b[0m     out \u001b[38;5;241m=\u001b[39m \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mError in \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mpdb_id\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m: \u001b[39m\u001b[38;5;132;01m{\u001b[39;00me\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    243\u001b[0m     energy \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[0;32m--> 244\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m e\n\u001b[1;32m    246\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mpdb_id\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m: Estimated Free Energy of Binding = \u001b[39m\u001b[38;5;132;01m{\u001b[39;00menergy\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m kcal/mol\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m    247\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39msave_output(out, energy)\n",
      "Cell \u001b[0;32mIn[39], line 238\u001b[0m, in \u001b[0;36mDockingScorer.vina_score_ligand\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    236\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mcheck_files()\n\u001b[1;32m    237\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 238\u001b[0m     out \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrun_vina\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    239\u001b[0m     energy \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mextract_vina_score(out)\n\u001b[1;32m    240\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m e:\n",
      "Cell \u001b[0;32mIn[39], line 77\u001b[0m, in \u001b[0;36mDockingScorer.run_vina\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m     75\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mout: \u001b[39m\u001b[38;5;124m\"\u001b[39m, out)\n\u001b[1;32m     76\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124merr: \u001b[39m\u001b[38;5;124m\"\u001b[39m, err)\n\u001b[0;32m---> 77\u001b[0m \u001b[38;5;28;01mraise\u001b[39;00m e\n",
      "Cell \u001b[0;32mIn[39], line 68\u001b[0m, in \u001b[0;36mDockingScorer.run_vina\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m     66\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m     67\u001b[0m     out, err \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m\"\u001b[39m  \u001b[38;5;66;03m# Initialize out and err\u001b[39;00m\n\u001b[0;32m---> 68\u001b[0m     code, out, err \u001b[38;5;241m=\u001b[39m \u001b[43mrun_command\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcmd\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtimeout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m100\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m     69\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m code \u001b[38;5;241m!=\u001b[39m \u001b[38;5;241m0\u001b[39m:\n\u001b[1;32m     70\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m CommandExecuteError(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mCommand failed with return code \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mcode\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m)\n",
      "Cell \u001b[0;32mIn[24], line 30\u001b[0m, in \u001b[0;36mrun_command\u001b[0;34m(cmd, raise_error, input, timeout, **kwargs)\u001b[0m\n\u001b[1;32m     28\u001b[0m err \u001b[38;5;241m=\u001b[39m err\u001b[38;5;241m.\u001b[39mdecode(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mutf-8\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[1;32m     29\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m raise_error \u001b[38;5;129;01mand\u001b[39;00m return_code \u001b[38;5;241m!=\u001b[39m \u001b[38;5;241m0\u001b[39m:\n\u001b[0;32m---> 30\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m CommandExecuteError(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mCommand \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mcmd\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m failed: \u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;132;01m{\u001b[39;00merr\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m     31\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m return_code, out, err\n",
      "\u001b[0;31mCommandExecuteError\u001b[0m: Command ./vina/vina_1.2.5_linux_x86_64 --config L3001_0_5S9L_receptor_w_water_metal_config.txt failed: \n\n\nPDBQT parsing error: Coordinate \"7  65.14\" is not valid.\n > ATOM   3946  HH1A1ARG A 450       8.039   7.307  65.141  1.00  0.00     0.174 HD\n"
     ]
    }
   ],
   "source": [
    "lig_file = 'L3001_0.sdf' \n",
    "prot_file = '5S9L_receptor_w_water_metal.pdb'\n",
    "complex_file = '5S9L.pdb'\n",
    "csv_out_file='vina_playground_docking_data.csv'\n",
    "\n",
    "'''Final function to process each protein and ligand pair, prepare them for docking, score the ligand and save the output files.\n",
    "Both lig_file and prot_file must be strings, not lists. If you want to process multiple ligands and proteins (i.e. LISTS), you should use this function in the following manner: \n",
    "for lig, prot in tqdm(zip(lig_files, prot_files)):\n",
    "    vina_process_lig_prot(lig, prot)\n",
    "    ...\n",
    "'''\n",
    "# TODO: handling lists..?\n",
    "if isinstance(lig_file, list) or isinstance(prot_file, list):\n",
    "    raise ValueError(\"lig_file and prot_file must be strings, not lists.\")\n",
    "\n",
    "destination_dir = f'data/{lig_file.split(\".\")[0]}_with_{prot_file.split(\".\")[0]}'\n",
    "os.makedirs(destination_dir, exist_ok=True)\n",
    "\n",
    "lig_name = os.path.splitext(lig_file)[0]\n",
    "prot_name = os.path.splitext(prot_file)[0]\n",
    "complex_name = os.path.splitext(complex_file)[0]\n",
    "\n",
    "# try:      \n",
    "os.environ['OE_LICENSE'] = '/home/ian/oe_license.txt' # change this to your OE_LICENSE path\n",
    "\n",
    "# prepare the ligand and receptor for Vina i.e. convert to pdbqt format\n",
    "prep = DockingPrepper('.',\n",
    "                        lig_file=lig_file, \n",
    "                        prot_file=prot_file,\n",
    "                        pdb_id=complex_name, # this is not really the pdb_id but more of a basename for the files\n",
    "# for vina preparation\n",
    "                        preserve_water = True,\n",
    "                        preserve_metal = True) \n",
    "prep.vina_process()\n",
    "        \n",
    "# using the DockingScorer class to get the docking score\n",
    "scorer = DockingScorer('.', \n",
    "                        lig_file, \n",
    "                        prot_file,\n",
    "                        complex_file=complex_file, # reference complex file, required for the vina process \n",
    "                        get_vina_poses=True, # this only works for vina, there's no choice for OE - HA!\n",
    "                        save_out_file=True,\n",
    "                        from_pdb=False,\n",
    "                        csv_out_file=csv_out_file,\n",
    "                        receptor_DU=None) # None because this is for Vina, but for OE, this will be the receptor DU (required)\n",
    "\n",
    "docking_score = scorer.vina_score_ligand()\n",
    "\n",
    "new_dir_name = f'{lig_name}_{prot_name}'\n",
    "new_dir = os.path.join(cwd, new_dir_name)\n",
    "os.makedirs(new_dir, exist_ok=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# <font color='orange'>Testing with complexes produced from donwloading PDB ID, then plitting the complex files</font>\n",
    "\n",
    "Rescoring crystal structures"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|          | 0/6 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Downloading PDB structure '5s9l'...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ian/micromamba/envs/asapdiscovery/lib/python3.10/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 7409.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fixed receptor file saved as: 5S9L_receptor_fixed.pdb\n",
      "Extracting ligand from 5S9L.pdb...\n",
      "Reading receptor from 5S9L.pdb...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "DPI: 0.12, RFree: 0.19, Resolution: 1.90\n",
      "Processing BU # 1 with title: ISOFORM 2 OF ECTONUCLEOTIDE, chains AB\n",
      "Warning: No rotamer of residue: GLU 573   A 1   passed a clash test, cluster will be skipped\n",
      "Found unresolved N-terminal with 23 residues before TRP 51   A 1  , with sequence FTASRIKRAEWDEGPPTVLSDSP\n",
      "Found 4 residue gap between LYS 463   A 1   and CYS 468   A 1  , with sequence PSGK\n",
      "Found 2 residue gap between GLU 573   A 1   and ASN 576   A 1  , with sequence PK\n",
      "Found unresolved C-terminal with 12 residues after GLU 861   A 1  , with sequence IGGRHHHHHHHH\n",
      "Warning: There was a problem building some missing pieces, built as much as was possible\n",
      "   Falling back to charging protein with OEMMFF94Charges\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The complex file has been saved as 5S9L.pdb\n",
      "The receptor file has been saved as 5S9L_receptor_fixed.pdb\n",
      "The ligand file has been saved as 5S9L_ligand.pdb\n",
      "The ligand file has been saved as 5S9L_ligand.pdbqt\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_fixed:A:SER751:OG\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_fixed:A: CA806:CA\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_fixed:A:  K807:K\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_fixed:A: ZN808:ZN\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_fixed:A: CA810:CA\n",
      "Sorry, there are no Gasteiger parameters available for atom 5S9L_receptor_fixed:A:HOH1235:O\n",
      "The protein file 5S9L_receptor_fixed.pdb has been converted to pdbqt format and saved in 5S9L_receptor_fixed.pdbqt\n",
      "Extracting ligand from 5S9L.pdb...\n",
      "Reading receptor from 5S9L.pdb...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "DPI: 0.12, RFree: 0.19, Resolution: 1.90\n",
      "Processing BU # 1 with title: ISOFORM 2 OF ECTONUCLEOTIDE, chains AB\n",
      "Warning: No rotamer of residue: GLU 573   A 1   passed a clash test, cluster will be skipped\n",
      "Found unresolved N-terminal with 23 residues before TRP 51   A 1  , with sequence FTASRIKRAEWDEGPPTVLSDSP\n",
      "Found 4 residue gap between LYS 463   A 1   and CYS 468   A 1  , with sequence PSGK\n",
      "Found 2 residue gap between GLU 573   A 1   and ASN 576   A 1  , with sequence PK\n",
      "Found unresolved C-terminal with 12 residues after GLU 861   A 1  , with sequence IGGRHHHHHHHH\n",
      "Warning: There was a problem building some missing pieces, built as much as was possible\n",
      "   Falling back to charging protein with OEMMFF94Charges\n",
      "mv: cannot move '5S9L_ligand_5S9L_receptor_fixed' to a subdirectory of itself, '/home/ian/msc_project/tmp_exp_docking/5S9L_ligand_5S9L_receptor_fixed/5S9L_ligand_5S9L_receptor_fixed'\n",
      "mv: cannot move '5S9L_ligand_5S9L_receptor_fixed' to a subdirectory of itself, '/home/ian/msc_project/tmp_exp_docking/5S9L_ligand_5S9L_receptor_fixed/5S9L_ligand_5S9L_receptor_fixed'\n",
      "mv: cannot move '5S9L_ligand_5S9L_receptor_fixed' to a subdirectory of itself, '/home/ian/msc_project/tmp_exp_docking/5S9L_ligand_5S9L_receptor_fixed/5S9L_ligand_5S9L_receptor_fixed'\n",
      " 17%|        | 1/6 [01:49<09:07, 109.47s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5S9L: Estimated Free Energy of Binding = -11.19 kcal/mol\n",
      "Output saved as 5S9L_receptor_fixed_5S9L_ligand.out\n",
      "\n",
      "Data saved to AutoDock_w_water_data.csv\n",
      "Data saved in ./data//home/ian/msc_project/tmp_exp_docking/5S9L_ligand_5S9L_receptor_fixed\n",
      "Downloading PDB structure '1klt'...\n",
      "Fixed receptor file saved as: 1KLT_receptor_fixed.pdb\n",
      "Extracting ligand from 1KLT.pdb...\n",
      "Reading receptor from 1KLT.pdb...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "DPI: 0.22, RFree: 0.26, Resolution: 1.90\n",
      "Processing BU # 1 with title: CHYMASE, chains A\n",
      "Warning: For residue LYS 109   A 1   removing clashing solvent molecule HOH 533   A 2  \n",
      "Warning: For residue LYS 109   A 1   removing clashing solvent molecule HOH 531   A 2  \n",
      "Warning: For residue ARG 137   A 1   removing clashing solvent molecule HOH 591   A 2  \n",
      "Warning: For 'PMS', reference state application resulted in 3 results in 3D\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The complex file has been saved as 1KLT.pdb\n",
      "The receptor file has been saved as 1KLT_receptor_fixed.pdb\n",
      "The ligand file has been saved as 1KLT_ligand.pdb\n",
      "The ligand file has been saved as 1KLT_ligand.pdbqt\n",
      "adding gasteiger charges to peptide\n",
      "Sorry, there are no Gasteiger parameters available for atom 1KLT_receptor_fixed:A:GLU71:OE2\n",
      "Unable to assign MAP type to atom N\n",
      "Sorry, there are no Gasteiger parameters available for atom 1KLT_receptor_fixed:A:LYS173B:NZ\n",
      "Sorry, there are no Gasteiger parameters available for atom 1KLT_receptor_fixed:A:HOH367:O\n",
      "Warning: hydrogens,  ['H1', 'H2']  , with no bonds!\n",
      "The protein file 1KLT_receptor_fixed.pdb has been converted to pdbqt format and saved in 1KLT_receptor_fixed.pdbqt\n",
      "Extracting ligand from 1KLT.pdb...\n",
      "Reading receptor from 1KLT.pdb...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "DPI: 0.22, RFree: 0.26, Resolution: 1.90\n",
      "Processing BU # 1 with title: CHYMASE, chains A\n",
      "Warning: For residue LYS 109   A 1   removing clashing solvent molecule HOH 533   A 2  \n",
      "Warning: For residue LYS 109   A 1   removing clashing solvent molecule HOH 531   A 2  \n",
      "Warning: For residue ARG 137   A 1   removing clashing solvent molecule HOH 591   A 2  \n",
      "Warning: For 'PMS', reference state application resulted in 3 results in 3D\n",
      "mv: cannot move '1KLT_ligand_1KLT_receptor_fixed' to a subdirectory of itself, '/home/ian/msc_project/tmp_exp_docking/1KLT_ligand_1KLT_receptor_fixed/1KLT_ligand_1KLT_receptor_fixed'\n",
      "mv: cannot move '1KLT_ligand_1KLT_receptor_fixed' to a subdirectory of itself, '/home/ian/msc_project/tmp_exp_docking/1KLT_ligand_1KLT_receptor_fixed/1KLT_ligand_1KLT_receptor_fixed'\n",
      "mv: cannot move '1KLT_ligand_1KLT_receptor_fixed' to a subdirectory of itself, '/home/ian/msc_project/tmp_exp_docking/1KLT_ligand_1KLT_receptor_fixed/1KLT_ligand_1KLT_receptor_fixed'\n",
      " 33%|      | 2/6 [02:56<05:38, 84.58s/it] "
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1KLT: Estimated Free Energy of Binding = -4.271 kcal/mol\n",
      "Output saved as 1KLT_receptor_fixed_1KLT_ligand.out\n",
      "\n",
      "Data saved to AutoDock_w_water_data.csv\n",
      "Data saved in ./data//home/ian/msc_project/tmp_exp_docking/1KLT_ligand_1KLT_receptor_fixed\n",
      "Downloading PDB structure '1a42'...\n",
      "Fixed receptor file saved as: 1A42_receptor_fixed.pdb\n",
      "Extracting ligand from 1A42.pdb...\n",
      "Reading receptor from 1A42.pdb...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning: Cannot set Rfree, due to lack of R value for CARBONIC ANHYDRASE II\n",
      "DPI: 0.00, RFree: 0.00, Resolution: 2.25\n",
      "Processing BU # 1 with title: CARBONIC ANHYDRASE II, chains A\n",
      "Found unresolved N-terminal with 2 residues before HIS 4   A 1  , with sequence SH\n",
      "Found unresolved C-terminal with 1 residues after PHE 260   A 1  , with sequence K\n",
      "Warning: There was a problem building some missing pieces, built as much as was possible\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The complex file has been saved as 1A42.pdb\n",
      "The receptor file has been saved as 1A42_receptor_fixed.pdb\n",
      "The ligand file has been saved as 1A42_ligand.pdb\n",
      "The ligand file has been saved as 1A42_ligand.pdbqt\n",
      "Sorry, there are no Gasteiger parameters available for atom 1A42_receptor_fixed:A:PHE256:O\n",
      "Sorry, there are no Gasteiger parameters available for atom 1A42_receptor_fixed:A: ZN257:ZN\n",
      "The protein file 1A42_receptor_fixed.pdb has been converted to pdbqt format and saved in 1A42_receptor_fixed.pdbqt\n",
      "Extracting ligand from 1A42.pdb...\n",
      "Reading receptor from 1A42.pdb...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning: Cannot set Rfree, due to lack of R value for CARBONIC ANHYDRASE II\n",
      "DPI: 0.00, RFree: 0.00, Resolution: 2.25\n",
      "Processing BU # 1 with title: CARBONIC ANHYDRASE II, chains A\n",
      "Found unresolved N-terminal with 2 residues before HIS 4   A 1  , with sequence SH\n",
      "Found unresolved C-terminal with 1 residues after PHE 260   A 1  , with sequence K\n",
      "Warning: There was a problem building some missing pieces, built as much as was possible\n",
      "mv: cannot move '1A42_ligand_1A42_receptor_fixed' to a subdirectory of itself, '/home/ian/msc_project/tmp_exp_docking/1A42_ligand_1A42_receptor_fixed/1A42_ligand_1A42_receptor_fixed'\n",
      "mv: cannot move '1A42_ligand_1A42_receptor_fixed' to a subdirectory of itself, '/home/ian/msc_project/tmp_exp_docking/1A42_ligand_1A42_receptor_fixed/1A42_ligand_1A42_receptor_fixed'\n",
      "mv: cannot move '1A42_ligand_1A42_receptor_fixed' to a subdirectory of itself, '/home/ian/msc_project/tmp_exp_docking/1A42_ligand_1A42_receptor_fixed/1A42_ligand_1A42_receptor_fixed'\n",
      " 50%|     | 3/6 [03:48<03:29, 69.83s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1A42: Estimated Free Energy of Binding = -6.809 kcal/mol\n",
      "Output saved as 1A42_receptor_fixed_1A42_ligand.out\n",
      "\n",
      "Data saved to AutoDock_w_water_data.csv\n",
      "Data saved in ./data//home/ian/msc_project/tmp_exp_docking/1A42_ligand_1A42_receptor_fixed\n",
      "Downloading PDB structure '4b3d'...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ian/micromamba/envs/asapdiscovery/lib/python3.10/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 4092.\n",
      "  warnings.warn(\n",
      "/home/ian/micromamba/envs/asapdiscovery/lib/python3.10/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain C is discontinuous at line 4117.\n",
      "  warnings.warn(\n",
      "/home/ian/micromamba/envs/asapdiscovery/lib/python3.10/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 4142.\n",
      "  warnings.warn(\n",
      "/home/ian/micromamba/envs/asapdiscovery/lib/python3.10/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain C is discontinuous at line 4491.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fixed receptor file saved as: 4B3D_receptor_fixed.pdb\n",
      "Extracting ligand from 4B3D.pdb...\n",
      "Reading receptor from 4B3D.pdb...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "DPI: 0.09, RFree: 0.21, Resolution: 1.59\n",
      "Processing BU # 1 with title: DNA REPAIR AND RECOMBINATION PROTEIN RADA, chains A\n",
      "Found unresolved N-terminal with 1 residues before ALA 108   A 1  , with sequence M\n",
      "Found 11 residue gap between VAL 285   A 1   and ALA 309   A 1  , with sequence QANGGHILAHS\n",
      "Found 6 residue gap between ASP 329   A 1   and GLY 336   A 1  , with sequence APHLPE\n",
      "Warning: There was a problem building some missing pieces, built as much as was possible\n",
      "   Falling back to charging protein with OEMMFF94Charges\n",
      "Processing BU # 2 with title: DNA REPAIR AND RECOMBINATION PROTEIN RADA, chains C\n",
      "Found unresolved N-terminal with 1 residues before ALA 108   C 2  , with sequence M\n",
      "Found 10 residue gap between VAL 285   C 2   and SER 308   C 2  , with sequence QANGGHILAH\n",
      "Found 5 residue gap between ASP 329   C 2   and GLU 335   C 2  , with sequence APHLP\n",
      "Warning: There was a problem building some missing pieces, built as much as was possible\n",
      "   Falling back to charging protein with OEMMFF94Charges\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The complex file has been saved as 4B3D.pdb\n",
      "The receptor file has been saved as 4B3D_receptor_fixed.pdb\n",
      "The ligand file has been saved as 4B3D_ligand.pdb\n",
      "The ligand file has been saved as 4B3D_ligand.pdbqt\n",
      "adding gasteiger charges to peptide\n",
      "The protein file 4B3D_receptor_fixed.pdb has been converted to pdbqt format and saved in 4B3D_receptor_fixed.pdbqt\n",
      "Extracting ligand from 4B3D.pdb...\n",
      "Reading receptor from 4B3D.pdb...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "DPI: 0.09, RFree: 0.21, Resolution: 1.59\n",
      "Processing BU # 1 with title: DNA REPAIR AND RECOMBINATION PROTEIN RADA, chains A\n",
      "Found unresolved N-terminal with 1 residues before ALA 108   A 1  , with sequence M\n",
      "Found 11 residue gap between VAL 285   A 1   and ALA 309   A 1  , with sequence QANGGHILAHS\n",
      "Found 6 residue gap between ASP 329   A 1   and GLY 336   A 1  , with sequence APHLPE\n",
      "Warning: There was a problem building some missing pieces, built as much as was possible\n",
      "   Falling back to charging protein with OEMMFF94Charges\n",
      "Processing BU # 2 with title: DNA REPAIR AND RECOMBINATION PROTEIN RADA, chains C\n",
      "Found unresolved N-terminal with 1 residues before ALA 108   C 2  , with sequence M\n",
      "Found 10 residue gap between VAL 285   C 2   and SER 308   C 2  , with sequence QANGGHILAH\n",
      "Found 5 residue gap between ASP 329   C 2   and GLU 335   C 2  , with sequence APHLP\n",
      "Warning: There was a problem building some missing pieces, built as much as was possible\n",
      "   Falling back to charging protein with OEMMFF94Charges\n",
      "mv: cannot move '4B3D_ligand_4B3D_receptor_fixed' to a subdirectory of itself, '/home/ian/msc_project/tmp_exp_docking/4B3D_ligand_4B3D_receptor_fixed/4B3D_ligand_4B3D_receptor_fixed'\n",
      "mv: cannot move '4B3D_ligand_4B3D_receptor_fixed' to a subdirectory of itself, '/home/ian/msc_project/tmp_exp_docking/4B3D_ligand_4B3D_receptor_fixed/4B3D_ligand_4B3D_receptor_fixed'\n",
      "mv: cannot move '4B3D_ligand_4B3D_receptor_fixed' to a subdirectory of itself, '/home/ian/msc_project/tmp_exp_docking/4B3D_ligand_4B3D_receptor_fixed/4B3D_ligand_4B3D_receptor_fixed'\n",
      " 67%|   | 4/6 [04:49<02:12, 66.19s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4B3D: Estimated Free Energy of Binding = -5.427 kcal/mol\n",
      "Output saved as 4B3D_receptor_fixed_4B3D_ligand.out\n",
      "\n",
      "Data saved to AutoDock_w_water_data.csv\n",
      "Data saved in ./data//home/ian/msc_project/tmp_exp_docking/4B3D_ligand_4B3D_receptor_fixed\n",
      "Downloading PDB structure '3gvu'...\n",
      "Fixed receptor file saved as: 3GVU_receptor_fixed.pdb\n",
      "Extracting ligand from 3GVU.pdb...\n",
      "Reading receptor from 3GVU.pdb...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "DPI: 0.15, RFree: 0.23, Resolution: 2.05\n",
      "Processing BU # 1 with title: TYROSINE-PROTEIN KINASE ABL2, chains A\n",
      "Warning: None of the residue conformations passed a clash test\n",
      "Warning: For residue MET 278   A 1   removing clashing solvent molecule HOH 203   A 2  \n",
      "Warning: For residue LYS 402   A 1   removing clashing solvent molecule HOH 231   A 2  \n",
      "Warning: For residue LYS 402   A 1   removing clashing solvent molecule HOH 136   A 2  \n",
      "Warning: For residue SER 546   A 1   removing clashing solvent molecule HOH 549   A 2  \n",
      "Warning: Sequence alignment created an incorrect match for chain A, skipping gap detection. Try again with adjusted gap and extend penalties.\n",
      "Warning: There was a problem building some missing pieces, built as much as was possible\n",
      "   Falling back to charging protein with OEMMFF94Charges\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The complex file has been saved as 3GVU.pdb\n",
      "The receptor file has been saved as 3GVU_receptor_fixed.pdb\n",
      "The ligand file has been saved as 3GVU_ligand.pdb\n",
      "The ligand file has been saved as 3GVU_ligand.pdbqt\n",
      "adding gasteiger charges to peptide\n",
      "The protein file 3GVU_receptor_fixed.pdb has been converted to pdbqt format and saved in 3GVU_receptor_fixed.pdbqt\n",
      "Extracting ligand from 3GVU.pdb...\n",
      "Reading receptor from 3GVU.pdb...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "DPI: 0.15, RFree: 0.23, Resolution: 2.05\n",
      "Processing BU # 1 with title: TYROSINE-PROTEIN KINASE ABL2, chains A\n",
      "Warning: None of the residue conformations passed a clash test\n",
      "Warning: For residue MET 278   A 1   removing clashing solvent molecule HOH 203   A 2  \n",
      "Warning: For residue LYS 402   A 1   removing clashing solvent molecule HOH 231   A 2  \n",
      "Warning: For residue LYS 402   A 1   removing clashing solvent molecule HOH 136   A 2  \n",
      "Warning: For residue SER 546   A 1   removing clashing solvent molecule HOH 549   A 2  \n",
      "Warning: Sequence alignment created an incorrect match for chain A, skipping gap detection. Try again with adjusted gap and extend penalties.\n",
      "Warning: There was a problem building some missing pieces, built as much as was possible\n",
      "   Falling back to charging protein with OEMMFF94Charges\n",
      "mv: cannot move '3GVU_ligand_3GVU_receptor_fixed' to a subdirectory of itself, '/home/ian/msc_project/tmp_exp_docking/3GVU_ligand_3GVU_receptor_fixed/3GVU_ligand_3GVU_receptor_fixed'\n",
      "mv: cannot move '3GVU_ligand_3GVU_receptor_fixed' to a subdirectory of itself, '/home/ian/msc_project/tmp_exp_docking/3GVU_ligand_3GVU_receptor_fixed/3GVU_ligand_3GVU_receptor_fixed'\n",
      "mv: cannot move '3GVU_ligand_3GVU_receptor_fixed' to a subdirectory of itself, '/home/ian/msc_project/tmp_exp_docking/3GVU_ligand_3GVU_receptor_fixed/3GVU_ligand_3GVU_receptor_fixed'\n",
      " 83%| | 5/6 [06:47<01:24, 84.78s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3GVU: Estimated Free Energy of Binding = -9.571 kcal/mol\n",
      "Output saved as 3GVU_receptor_fixed_3GVU_ligand.out\n",
      "\n",
      "Data saved to AutoDock_w_water_data.csv\n",
      "Data saved in ./data//home/ian/msc_project/tmp_exp_docking/3GVU_ligand_3GVU_receptor_fixed\n",
      "Downloading PDB structure '6c7q'...\n",
      "Fixed receptor file saved as: 6C7Q_receptor_fixed.pdb\n",
      "Extracting ligand from 6C7Q.pdb...\n",
      "Reading receptor from 6C7Q.pdb...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "DPI: 0.08, RFree: 0.23, Resolution: 1.51\n",
      "Processing BU # 1 with title: BROMODOMAIN-CONTAINING PROTEIN 4, chains A\n",
      "Warning: For residue GLU 383   A 1   removing clashing solvent molecule HOH 686   A 2  \n",
      "Found unresolved N-terminal with 19 residues before LYS 349   A 1  , with sequence SNAKDVPDSQQHPAPEKSS\n",
      "Warning: There was a problem building some missing pieces, built as much as was possible\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The complex file has been saved as 6C7Q.pdb\n",
      "The receptor file has been saved as 6C7Q_receptor_fixed.pdb\n",
      "The ligand file has been saved as 6C7Q_ligand.pdb\n",
      "The ligand file has been saved as 6C7Q_ligand.pdbqt\n",
      "adding gasteiger charges to peptide\n",
      "The protein file 6C7Q_receptor_fixed.pdb has been converted to pdbqt format and saved in 6C7Q_receptor_fixed.pdbqt\n",
      "Extracting ligand from 6C7Q.pdb...\n",
      "Reading receptor from 6C7Q.pdb...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "DPI: 0.08, RFree: 0.23, Resolution: 1.51\n",
      "Processing BU # 1 with title: BROMODOMAIN-CONTAINING PROTEIN 4, chains A\n",
      "Warning: For residue GLU 383   A 1   removing clashing solvent molecule HOH 686   A 2  \n",
      "Found unresolved N-terminal with 19 residues before LYS 349   A 1  , with sequence SNAKDVPDSQQHPAPEKSS\n",
      "Warning: There was a problem building some missing pieces, built as much as was possible\n",
      "mv: cannot move '6C7Q_ligand_6C7Q_receptor_fixed' to a subdirectory of itself, '/home/ian/msc_project/tmp_exp_docking/6C7Q_ligand_6C7Q_receptor_fixed/6C7Q_ligand_6C7Q_receptor_fixed'\n",
      "mv: cannot move '6C7Q_ligand_6C7Q_receptor_fixed' to a subdirectory of itself, '/home/ian/msc_project/tmp_exp_docking/6C7Q_ligand_6C7Q_receptor_fixed/6C7Q_ligand_6C7Q_receptor_fixed'\n",
      "mv: cannot move '6C7Q_ligand_6C7Q_receptor_fixed' to a subdirectory of itself, '/home/ian/msc_project/tmp_exp_docking/6C7Q_ligand_6C7Q_receptor_fixed/6C7Q_ligand_6C7Q_receptor_fixed'\n",
      "100%|| 6/6 [08:17<00:00, 82.95s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6C7Q: Estimated Free Energy of Binding = -9.999 kcal/mol\n",
      "Output saved as 6C7Q_receptor_fixed_6C7Q_ligand.out\n",
      "\n",
      "Data saved to AutoDock_w_water_data.csv\n",
      "Data saved in ./data//home/ian/msc_project/tmp_exp_docking/6C7Q_ligand_6C7Q_receptor_fixed\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "pdb_ids = ['5S9L',\n",
    "           '1KLT',\n",
    "           '1A42',\n",
    "            '4B3D', # Failed in vina\n",
    "           '3GVU',\n",
    "           '6C7Q',\n",
    "           ]\n",
    "for pdb_id in tqdm(pdb_ids):\n",
    "    # download \n",
    "    filename = download_pdb_file(pdb_id)\n",
    "    # split the protein and ligand from the PDB file\n",
    "    complex_file, receptor_file, ligand_file = pdb_to_prot_lig(pdb_id, filename, preserve_water=True,preserve_metal=True)\n",
    "    # rest of pipeline\n",
    "    vina_process_lig_prot(ligand_file, receptor_file, complex_file, preserve_water=True, preserve_metal=True, csv_out_file='AutoDock_w_water_data.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Downloading PDB structure '6ynq'...\n",
      "Fixed receptor file saved as: 6YNQ_receptor_fixed.pdb\n",
      "Extracting ligand from 6YNQ.pdb...\n",
      "Reading receptor from 6YNQ.pdb...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " CL 403   A 2   and  CL 403   B 2   are overlapping. Overlap: 19.65, Tanimoto: 1.00\n",
      "    Overlap is complete, marking  CL 403   B 2   for deletion\n",
      "    Deleted CL   :  CL 403   B 2  \n",
      "HOH 696   A 2   and HOH 696   B 2   are overlapping. Overlap: 19.65, Tanimoto: 1.00\n",
      "    Overlap is complete, marking HOH 696   B 2   for deletion\n",
      "    Deleted  O   : HOH 696   B 2  \n",
      "DPI: 0.13, RFree: 0.23, Resolution: 1.80\n",
      "Processing BU # 1 with title: 3C-LIKE PROTEINASE, chains AB\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The complex file has been saved as 6YNQ.pdb\n",
      "The receptor file has been saved as 6YNQ_receptor_fixed.pdb\n",
      "The ligand file has been saved as 6YNQ_ligand.pdb\n",
      "Error occurred: None\n",
      "Skipping 6YNQ_ligand.pdb due to an error: Command './ADFRsuite/ADFRsuite_x86_64Linux_1.0/bin/prepare_ligand -l 6YNQ_ligand.pdb -o 6YNQ_ligand.pdbqt -A hydrogens' returned non-zero exit status 1.\n",
      "Error message saved to mpro_w_water_data.csv\n",
      "Data saved in ./data//home/ian/msc_project/tmp_vina_mpro_docking/6YNQ_ligand_6YNQ_receptor_fixed\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"/home/ian/msc_project/docking/ADFRsuite/ADFRsuite_x86_64Linux_1.0/CCSBpckgs/AutoDockTools/Utilities24/prepare_ligand4.py\", line 242, in <module>\n",
      "    attach_singletons=attach_singletons)\n",
      "  File \"/home/ian/msc_project/docking/ADFRsuite/ADFRsuite_x86_64Linux_1.0/CCSBpckgs/AutoDockTools/MoleculePreparation.py\", line 1016, in __init__\n",
      "    detect_bonds_between_cycles=detect_bonds_between_cycles)\n",
      "  File \"/home/ian/msc_project/docking/ADFRsuite/ADFRsuite_x86_64Linux_1.0/CCSBpckgs/AutoDockTools/MoleculePreparation.py\", line 765, in __init__\n",
      "    delete_single_nonstd_residues=False)\n",
      "  File \"/home/ian/msc_project/docking/ADFRsuite/ADFRsuite_x86_64Linux_1.0/CCSBpckgs/AutoDockTools/MoleculePreparation.py\", line 124, in __init__\n",
      "    self.repairMol(mol, self.repair_type_list)\n",
      "  File \"/home/ian/msc_project/docking/ADFRsuite/ADFRsuite_x86_64Linux_1.0/CCSBpckgs/AutoDockTools/MoleculePreparation.py\", line 174, in repairMol\n",
      "    self.newHs = self.addHydrogens(mol)\n",
      "  File \"/home/ian/msc_project/docking/ADFRsuite/ADFRsuite_x86_64Linux_1.0/CCSBpckgs/AutoDockTools/MoleculePreparation.py\", line 187, in addHydrogens\n",
      "    HB.addHydrogens(mol)\n",
      "  File \"/home/ian/msc_project/docking/ADFRsuite/ADFRsuite_x86_64Linux_1.0/CCSBpckgs/MolKit/hydrogenBuilder.py\", line 61, in addHydrogens\n",
      "    babel.assignHybridization(mol.allAtoms)\n",
      "  File \"/home/ian/msc_project/docking/ADFRsuite/ADFRsuite_x86_64Linux_1.0/CCSBpckgs/PyBabel/atomTypes.py\", line 127, in assignHybridization\n",
      "    a.babel_atomic_number = self.get_atomic_number(a.babel_type)\n",
      "  File \"/home/ian/msc_project/docking/ADFRsuite/ADFRsuite_x86_64Linux_1.0/CCSBpckgs/PyBabel/atomTypes.py\", line 103, in get_atomic_number\n",
      "    (name,_name) )\n",
      "ValueError: Could not find atomic number for X X\n",
      "mv: cannot move '6YNQ_ligand_6YNQ_receptor_fixed' to a subdirectory of itself, '/home/ian/msc_project/tmp_vina_mpro_docking/6YNQ_ligand_6YNQ_receptor_fixed/6YNQ_ligand_6YNQ_receptor_fixed'\n",
      "mv: cannot move '6YNQ_ligand_6YNQ_receptor_fixed' to a subdirectory of itself, '/home/ian/msc_project/tmp_vina_mpro_docking/6YNQ_ligand_6YNQ_receptor_fixed/6YNQ_ligand_6YNQ_receptor_fixed'\n",
      "mv: cannot move '6YNQ_ligand_6YNQ_receptor_fixed' to a subdirectory of itself, '/home/ian/msc_project/tmp_vina_mpro_docking/6YNQ_ligand_6YNQ_receptor_fixed/6YNQ_ligand_6YNQ_receptor_fixed'\n"
     ]
    }
   ],
   "source": [
    "pdb_id = '6YNQ' # Mpro active site with ligand\n",
    "filename = download_pdb_file(pdb_id)\n",
    "# split the protein and ligand from the PDB file\n",
    "complex_file, receptor_file, ligand_file = pdb_to_prot_lig(pdb_id, filename, preserve_water=True,preserve_metal=True)\n",
    "# rest of pipeline\n",
    "vina_process_lig_prot(ligand_file, receptor_file, complex_file, preserve_water=True, preserve_metal=True, csv_out_file='mpro_w_water_data.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The ligand file has been saved as 6R8O_ligand.pdbqt\n",
      "adding gasteiger charges to peptide\n",
      "The protein file 6R8O_receptor_fixed.pdb has been converted to pdbqt format and saved in 6R8O_receptor_fixed.pdbqt\n",
      "Extracting ligand from 6R8O.pdb...\n",
      "Reading receptor from 6R8O.pdb...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "DPI: 0.05, RFree: 0.18, Resolution: 1.36\n",
      "Processing BU # 1 with title: PEPTIDYL-PROLYL CIS-TRANS ISOMERASE F, MITOCHONDRIA, chains A\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6R8O: Estimated Free Energy of Binding = -10.33 kcal/mol\n",
      "Output saved as 6R8O_receptor_fixed_6R8O_ligand.out\n",
      "\n",
      "Data saved to vina_docking_data_playground.csv\n",
      "Data saved in ./data//home/ian/msc_project/docking/6R8O_ligand_6R8O_receptor_fixed\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "mv: cannot move '6R8O_ligand_6R8O_receptor_fixed' to a subdirectory of itself, '/home/ian/msc_project/docking/6R8O_ligand_6R8O_receptor_fixed/6R8O_ligand_6R8O_receptor_fixed'\n",
      "mv: cannot move '6R8O_ligand_6R8O_receptor_fixed' to a subdirectory of itself, '/home/ian/msc_project/docking/6R8O_ligand_6R8O_receptor_fixed/6R8O_ligand_6R8O_receptor_fixed'\n",
      "mv: cannot move '6R8O_ligand_6R8O_receptor_fixed' to a subdirectory of itself, '/home/ian/msc_project/docking/6R8O_ligand_6R8O_receptor_fixed/6R8O_ligand_6R8O_receptor_fixed'\n",
      "mv: cannot move '/home/ian/msc_project/docking/6R8O_ligand_6R8O_receptor_fixed' to './data/6R8O_ligand_6R8O_receptor_fixed': Directory not empty\n"
     ]
    }
   ],
   "source": [
    "receptor_file = '6R8O_receptor_fixed.pdb'\n",
    "ligand_file = '6R8O_ligand.pdb'\n",
    "complex_file = '6R8O.pdb'\n",
    "vina_process_lig_prot(ligand_file, receptor_file, complex_file, csv_out_file='vina_docking_data_playground.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<font color='orange'>Testing with Mpro protein and a ligand</font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "1 molecule converted\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The ligand file has been converted to Updated_Mpro_data_3.pdb\n",
      "The ligand file has been saved as Updated_Mpro_data_3.pdbqt\n",
      "adding gasteiger charges to peptide\n",
      "The protein file Mpro-protein.pdb has been converted to pdbqt format and saved in Mpro-protein.pdbqt\n",
      "Extracting ligand from Mpro-x0072_0.pdb...\n",
      "Reading receptor from Mpro-x0072_0.pdb...\n",
      "Failed to extract ligand from Mpro-x0072_0.pdb using the OpenEye Toolkit,trying a different script written using OESplitMolComplex...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "DPI: 0.12, RFree: 0.23, Resolution: 1.65\n",
      "Processing BU # 1 with title: ---, chains AB\n",
      "Warning: OEPlaceHydrogens: mol DesignUnit Components has heavy atoms ' O   HOH 231 A B 1  ' and ' O   HOH 231 B B 1  ' too close 0.11 < 0.5A\n",
      "Warning: Unable to add protons to DesignUnit: ---(AB)__DU__biounit, failing with strict mode enabled\n",
      "Warning: No ligand or sites detected, please revise inputs and/or options\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mpro-protein: Estimated Free Energy of Binding = -6.421 kcal/mol\n",
      "Output saved as Mpro-protein_Updated_Mpro_data_3.out\n",
      "\n",
      "Data saved to mpro_vina_w_water_data.csv\n",
      "Data saved in ./data//home/ian/msc_project/tmp_vina_mpro_docking/Updated_Mpro_data_3_Mpro-protein\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "mv: cannot move 'Updated_Mpro_data_3_Mpro-protein' to a subdirectory of itself, '/home/ian/msc_project/tmp_vina_mpro_docking/Updated_Mpro_data_3_Mpro-protein/Updated_Mpro_data_3_Mpro-protein'\n",
      "mv: cannot move 'Updated_Mpro_data_3_Mpro-protein' to a subdirectory of itself, '/home/ian/msc_project/tmp_vina_mpro_docking/Updated_Mpro_data_3_Mpro-protein/Updated_Mpro_data_3_Mpro-protein'\n",
      "mv: cannot move '/home/ian/msc_project/tmp_vina_mpro_docking/Updated_Mpro_data_3_Mpro-protein' to './data/Updated_Mpro_data_3_Mpro-protein': Directory not empty\n"
     ]
    }
   ],
   "source": [
    "ligand_file = 'Updated_Mpro_data_3.sdf'\n",
    "# receptor_file = 'Mpro-protein.pdb'\n",
    "# complex_file = 'Mpro-x0072_0.pdb'\n",
    "os.system(f'cp Updated_Mpro_ligands/{ligand_file} .')\n",
    "os.system(f'cp Mpro_complexes/{receptor_file} .')\n",
    "os.system(f'cp Mpro_complexes/{complex_file} .')\n",
    "\n",
    "vina_process_lig_prot(ligand_file, \n",
    "                      receptor_file, \n",
    "                      complex_file, \n",
    "                      preserve_water=True, \n",
    "                      preserve_metal=True, \n",
    "                      csv_out_file='mpro_vina_w_water_data.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<font color='yellow'>Running the pipeline for lists</font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The ligand file has been saved as Mpro-x0161_0_ligand.pdbqt\n",
      "adding gasteiger charges to peptide\n",
      "The protein file Mpro-x0072_0_receptor_fixed.pdb has been converted to pdbqt format and saved in Mpro-x0072_0_receptor_fixed.pdbqt\n",
      "Mpro-x0072: Estimated Free Energy of Binding = 0.0 kcal/mol\n",
      "Output saved as Mpro-x0072_0_receptor_fixed_Mpro-x0161_0_ligand.out\n",
      "\n",
      "Data saved to vina_docking_data_playground.csv\n",
      "Data saved in data/Mpro-x0161_0_ligand_with_Mpro-x0072_0_receptor_fixed\n",
      "The ligand file has been saved as Mpro-x0161_0_ligand.pdbqt\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "cp: -r not specified; omitting directory 'Mpro-x0072_0_receptor_fixed'\n",
      "mv: cannot move '/home/ian/msc_project/docking/Mpro-x0072_0_receptor_fixed' to 'data/Mpro-x0161_0_ligand_with_Mpro-x0072_0_receptor_fixed/Mpro-x0072_0_receptor_fixed': Directory not empty\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "adding gasteiger charges to peptide\n",
      "The protein file Mpro-x0161_0_receptor_fixed.pdb has been converted to pdbqt format and saved in Mpro-x0161_0_receptor_fixed.pdbqt\n",
      "Mpro-x0161: Estimated Free Energy of Binding = 0.0 kcal/mol\n",
      "Output saved as Mpro-x0161_0_receptor_fixed_Mpro-x0161_0_ligand.out\n",
      "\n",
      "Data saved to vina_docking_data_playground.csv\n",
      "Data saved in data/Mpro-x0161_0_ligand_with_Mpro-x0161_0_receptor_fixed\n",
      "The ligand file has been saved as Mpro-x0072_0_ligand.pdbqt\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "cp: -r not specified; omitting directory 'Mpro-x0161_0_receptor_fixed'\n",
      "mv: cannot move '/home/ian/msc_project/docking/Mpro-x0161_0_receptor_fixed' to 'data/Mpro-x0161_0_ligand_with_Mpro-x0161_0_receptor_fixed/Mpro-x0161_0_receptor_fixed': Directory not empty\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "adding gasteiger charges to peptide\n",
      "The protein file Mpro-x0072_0_receptor_fixed.pdb has been converted to pdbqt format and saved in Mpro-x0072_0_receptor_fixed.pdbqt\n",
      "Mpro-x0072: Estimated Free Energy of Binding = -2.443 kcal/mol\n",
      "Output saved as Mpro-x0072_0_receptor_fixed_Mpro-x0072_0_ligand.out\n",
      "\n",
      "Data saved to vina_docking_data_playground.csv\n",
      "Data saved in data/Mpro-x0072_0_ligand_with_Mpro-x0072_0_receptor_fixed\n",
      "The ligand file has been saved as Mpro-x0072_0_ligand.pdbqt\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "cp: -r not specified; omitting directory 'Mpro-x0072_0_receptor_fixed'\n",
      "mv: cannot move '/home/ian/msc_project/docking/Mpro-x0072_0_receptor_fixed' to 'data/Mpro-x0072_0_ligand_with_Mpro-x0072_0_receptor_fixed/Mpro-x0072_0_receptor_fixed': Directory not empty\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "adding gasteiger charges to peptide\n",
      "The protein file Mpro-x0161_0_receptor_fixed.pdb has been converted to pdbqt format and saved in Mpro-x0161_0_receptor_fixed.pdbqt\n",
      "Mpro-x0161: Estimated Free Energy of Binding = -0.252 kcal/mol\n",
      "Output saved as Mpro-x0161_0_receptor_fixed_Mpro-x0072_0_ligand.out\n",
      "\n",
      "Data saved to vina_docking_data_playground.csv\n",
      "Data saved in data/Mpro-x0072_0_ligand_with_Mpro-x0161_0_receptor_fixed\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "cp: -r not specified; omitting directory 'Mpro-x0161_0_receptor_fixed'\n",
      "mv: cannot move '/home/ian/msc_project/docking/Mpro-x0161_0_receptor_fixed' to 'data/Mpro-x0072_0_ligand_with_Mpro-x0161_0_receptor_fixed/Mpro-x0161_0_receptor_fixed': Directory not empty\n"
     ]
    }
   ],
   "source": [
    "ligand_files = ['Mpro-x0161_0_ligand.pdb', \n",
    "                'Mpro-x0072_0_ligand.pdb'] \n",
    "receptor_files = ['Mpro-x0072_0_receptor_fixed.pdb',\n",
    "                 'Mpro-x0161_0_receptor_fixed.pdb']\n",
    "for ligand_file in tqdm(ligand_files):\n",
    "    for receptor_file in receptor_files:\n",
    "        vina_process_lig_prot(lig_file=ligand_file, \n",
    "                 prot_file=receptor_file,\n",
    "                 csv_out_file='vina_docking_data_playground.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<font color='yellow'>Trying out with diffusion model generated molecules</font>\n",
    "\n",
    "For these molecules, the input sdf file likely has more than 1 molecules. If this is the case, then we have to add another step to split the sdf file into several sdf files first. We use the function ``split_sdf``."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "14 molecules have been successfully separated into individual SDF files.\n",
      "The ligand file has been converted to molecules_bx_2024_06_18_154410_1.pdb\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "1 molecule converted\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The ligand file has been saved as molecules_bx_2024_06_18_154410_1.pdbqt\n",
      "adding gasteiger charges to peptide\n",
      "The protein file Mpro-x0072_0_receptor_fixed.pdb has been converted to pdbqt format and saved in Mpro-x0072_0_receptor_fixed.pdbqt\n",
      "Mpro-x0072: Estimated Free Energy of Binding = 16.739 kcal/mol\n",
      "Output saved as Mpro-x0072_0_receptor_fixed_molecules_bx_2024_06_18_154410_1.out\n",
      "\n",
      "Data saved to vina_docking_data_playground.csv\n",
      "Data saved in data/molecules_bx_2024_06_18_154410_1_with_Mpro-x0072_0_receptor_fixed\n",
      "The ligand file has been converted to molecules_bx_2024_06_18_154410_2.pdb\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "cp: -r not specified; omitting directory 'Mpro-x0072_0_receptor_fixed'\n",
      "mv: cannot move '/home/ian/msc_project/docking/Mpro-x0072_0_receptor_fixed' to 'data/molecules_bx_2024_06_18_154410_1_with_Mpro-x0072_0_receptor_fixed/Mpro-x0072_0_receptor_fixed': Directory not empty\n",
      "1 molecule converted\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The ligand file has been saved as molecules_bx_2024_06_18_154410_2.pdbqt\n",
      "adding gasteiger charges to peptide\n",
      "The protein file Mpro-x0072_0_receptor_fixed.pdb has been converted to pdbqt format and saved in Mpro-x0072_0_receptor_fixed.pdbqt\n",
      "Mpro-x0072: Estimated Free Energy of Binding = 14.988 kcal/mol\n",
      "Output saved as Mpro-x0072_0_receptor_fixed_molecules_bx_2024_06_18_154410_2.out\n",
      "\n",
      "Data saved to vina_docking_data_playground.csv\n",
      "Data saved in data/molecules_bx_2024_06_18_154410_2_with_Mpro-x0072_0_receptor_fixed\n",
      "The ligand file has been converted to molecules_bx_2024_06_18_154410_3.pdb\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "cp: -r not specified; omitting directory 'Mpro-x0072_0_receptor_fixed'\n",
      "mv: cannot move '/home/ian/msc_project/docking/Mpro-x0072_0_receptor_fixed' to 'data/molecules_bx_2024_06_18_154410_2_with_Mpro-x0072_0_receptor_fixed/Mpro-x0072_0_receptor_fixed': Directory not empty\n",
      "1 molecule converted\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The ligand file has been saved as molecules_bx_2024_06_18_154410_3.pdbqt\n",
      "adding gasteiger charges to peptide\n",
      "The protein file Mpro-x0072_0_receptor_fixed.pdb has been converted to pdbqt format and saved in Mpro-x0072_0_receptor_fixed.pdbqt\n",
      "Mpro-x0072: Estimated Free Energy of Binding = 31.436 kcal/mol\n",
      "Output saved as Mpro-x0072_0_receptor_fixed_molecules_bx_2024_06_18_154410_3.out\n",
      "\n",
      "Data saved to vina_docking_data_playground.csv\n",
      "Data saved in data/molecules_bx_2024_06_18_154410_3_with_Mpro-x0072_0_receptor_fixed\n",
      "The ligand file has been converted to molecules_bx_2024_06_18_154410_4.pdb\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "cp: -r not specified; omitting directory 'Mpro-x0072_0_receptor_fixed'\n",
      "mv: cannot move '/home/ian/msc_project/docking/Mpro-x0072_0_receptor_fixed' to 'data/molecules_bx_2024_06_18_154410_3_with_Mpro-x0072_0_receptor_fixed/Mpro-x0072_0_receptor_fixed': Directory not empty\n",
      "1 molecule converted\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The ligand file has been saved as molecules_bx_2024_06_18_154410_4.pdbqt\n",
      "adding gasteiger charges to peptide\n",
      "The protein file Mpro-x0072_0_receptor_fixed.pdb has been converted to pdbqt format and saved in Mpro-x0072_0_receptor_fixed.pdbqt\n",
      "Mpro-x0072: Estimated Free Energy of Binding = 25.424 kcal/mol\n",
      "Output saved as Mpro-x0072_0_receptor_fixed_molecules_bx_2024_06_18_154410_4.out\n",
      "\n",
      "Data saved to vina_docking_data_playground.csv\n",
      "Data saved in data/molecules_bx_2024_06_18_154410_4_with_Mpro-x0072_0_receptor_fixed\n",
      "The ligand file has been converted to molecules_bx_2024_06_18_154410_5.pdb\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "cp: -r not specified; omitting directory 'Mpro-x0072_0_receptor_fixed'\n",
      "mv: cannot move '/home/ian/msc_project/docking/Mpro-x0072_0_receptor_fixed' to 'data/molecules_bx_2024_06_18_154410_4_with_Mpro-x0072_0_receptor_fixed/Mpro-x0072_0_receptor_fixed': Directory not empty\n",
      "1 molecule converted\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The ligand file has been saved as molecules_bx_2024_06_18_154410_5.pdbqt\n",
      "adding gasteiger charges to peptide\n",
      "The protein file Mpro-x0072_0_receptor_fixed.pdb has been converted to pdbqt format and saved in Mpro-x0072_0_receptor_fixed.pdbqt\n",
      "Mpro-x0072: Estimated Free Energy of Binding = 16.036 kcal/mol\n",
      "Output saved as Mpro-x0072_0_receptor_fixed_molecules_bx_2024_06_18_154410_5.out\n",
      "\n",
      "Data saved to vina_docking_data_playground.csv\n",
      "Data saved in data/molecules_bx_2024_06_18_154410_5_with_Mpro-x0072_0_receptor_fixed\n",
      "The ligand file has been converted to molecules_bx_2024_06_18_154410_6.pdb\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "cp: -r not specified; omitting directory 'Mpro-x0072_0_receptor_fixed'\n",
      "mv: cannot move '/home/ian/msc_project/docking/Mpro-x0072_0_receptor_fixed' to 'data/molecules_bx_2024_06_18_154410_5_with_Mpro-x0072_0_receptor_fixed/Mpro-x0072_0_receptor_fixed': Directory not empty\n",
      "1 molecule converted\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The ligand file has been saved as molecules_bx_2024_06_18_154410_6.pdbqt\n",
      "adding gasteiger charges to peptide\n",
      "The protein file Mpro-x0072_0_receptor_fixed.pdb has been converted to pdbqt format and saved in Mpro-x0072_0_receptor_fixed.pdbqt\n",
      "Mpro-x0072: Estimated Free Energy of Binding = 18.963 kcal/mol\n",
      "Output saved as Mpro-x0072_0_receptor_fixed_molecules_bx_2024_06_18_154410_6.out\n",
      "\n",
      "Data saved to vina_docking_data_playground.csv\n",
      "Data saved in data/molecules_bx_2024_06_18_154410_6_with_Mpro-x0072_0_receptor_fixed\n",
      "The ligand file has been converted to molecules_bx_2024_06_18_154410_7.pdb\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "cp: -r not specified; omitting directory 'Mpro-x0072_0_receptor_fixed'\n",
      "mv: cannot move '/home/ian/msc_project/docking/Mpro-x0072_0_receptor_fixed' to 'data/molecules_bx_2024_06_18_154410_6_with_Mpro-x0072_0_receptor_fixed/Mpro-x0072_0_receptor_fixed': Directory not empty\n",
      "1 molecule converted\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The ligand file has been saved as molecules_bx_2024_06_18_154410_7.pdbqt\n",
      "adding gasteiger charges to peptide\n",
      "The protein file Mpro-x0072_0_receptor_fixed.pdb has been converted to pdbqt format and saved in Mpro-x0072_0_receptor_fixed.pdbqt\n",
      "Mpro-x0072: Estimated Free Energy of Binding = 28.215 kcal/mol\n",
      "Output saved as Mpro-x0072_0_receptor_fixed_molecules_bx_2024_06_18_154410_7.out\n",
      "\n",
      "Data saved to vina_docking_data_playground.csv\n",
      "Data saved in data/molecules_bx_2024_06_18_154410_7_with_Mpro-x0072_0_receptor_fixed\n",
      "The ligand file has been converted to molecules_bx_2024_06_18_154410_8.pdb\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "cp: -r not specified; omitting directory 'Mpro-x0072_0_receptor_fixed'\n",
      "mv: cannot move '/home/ian/msc_project/docking/Mpro-x0072_0_receptor_fixed' to 'data/molecules_bx_2024_06_18_154410_7_with_Mpro-x0072_0_receptor_fixed/Mpro-x0072_0_receptor_fixed': Directory not empty\n",
      "1 molecule converted\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The ligand file has been saved as molecules_bx_2024_06_18_154410_8.pdbqt\n",
      "adding gasteiger charges to peptide\n",
      "The protein file Mpro-x0072_0_receptor_fixed.pdb has been converted to pdbqt format and saved in Mpro-x0072_0_receptor_fixed.pdbqt\n",
      "Mpro-x0072: Estimated Free Energy of Binding = 24.821 kcal/mol\n",
      "Output saved as Mpro-x0072_0_receptor_fixed_molecules_bx_2024_06_18_154410_8.out\n",
      "\n",
      "Data saved to vina_docking_data_playground.csv\n",
      "Data saved in data/molecules_bx_2024_06_18_154410_8_with_Mpro-x0072_0_receptor_fixed\n",
      "The ligand file has been converted to molecules_bx_2024_06_18_154410_9.pdb\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "cp: -r not specified; omitting directory 'Mpro-x0072_0_receptor_fixed'\n",
      "mv: cannot move '/home/ian/msc_project/docking/Mpro-x0072_0_receptor_fixed' to 'data/molecules_bx_2024_06_18_154410_8_with_Mpro-x0072_0_receptor_fixed/Mpro-x0072_0_receptor_fixed': Directory not empty\n",
      "1 molecule converted\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The ligand file has been saved as molecules_bx_2024_06_18_154410_9.pdbqt\n",
      "adding gasteiger charges to peptide\n",
      "The protein file Mpro-x0072_0_receptor_fixed.pdb has been converted to pdbqt format and saved in Mpro-x0072_0_receptor_fixed.pdbqt\n",
      "Mpro-x0072: Estimated Free Energy of Binding = 26.111 kcal/mol\n",
      "Output saved as Mpro-x0072_0_receptor_fixed_molecules_bx_2024_06_18_154410_9.out\n",
      "\n",
      "Data saved to vina_docking_data_playground.csv\n",
      "Data saved in data/molecules_bx_2024_06_18_154410_9_with_Mpro-x0072_0_receptor_fixed\n",
      "The ligand file has been converted to molecules_bx_2024_06_18_154410_10.pdb\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "cp: -r not specified; omitting directory 'Mpro-x0072_0_receptor_fixed'\n",
      "mv: cannot move '/home/ian/msc_project/docking/Mpro-x0072_0_receptor_fixed' to 'data/molecules_bx_2024_06_18_154410_9_with_Mpro-x0072_0_receptor_fixed/Mpro-x0072_0_receptor_fixed': Directory not empty\n",
      "1 molecule converted\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The ligand file has been saved as molecules_bx_2024_06_18_154410_10.pdbqt\n",
      "adding gasteiger charges to peptide\n",
      "The protein file Mpro-x0072_0_receptor_fixed.pdb has been converted to pdbqt format and saved in Mpro-x0072_0_receptor_fixed.pdbqt\n",
      "Mpro-x0072: Estimated Free Energy of Binding = 25.968 kcal/mol\n",
      "Output saved as Mpro-x0072_0_receptor_fixed_molecules_bx_2024_06_18_154410_10.out\n",
      "\n",
      "Data saved to vina_docking_data_playground.csv\n",
      "Data saved in data/molecules_bx_2024_06_18_154410_10_with_Mpro-x0072_0_receptor_fixed\n",
      "The ligand file has been converted to molecules_bx_2024_06_18_154410_11.pdb\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "cp: -r not specified; omitting directory 'Mpro-x0072_0_receptor_fixed'\n",
      "mv: cannot move '/home/ian/msc_project/docking/Mpro-x0072_0_receptor_fixed' to 'data/molecules_bx_2024_06_18_154410_10_with_Mpro-x0072_0_receptor_fixed/Mpro-x0072_0_receptor_fixed': Directory not empty\n",
      "1 molecule converted\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The ligand file has been saved as molecules_bx_2024_06_18_154410_11.pdbqt\n",
      "adding gasteiger charges to peptide\n",
      "The protein file Mpro-x0072_0_receptor_fixed.pdb has been converted to pdbqt format and saved in Mpro-x0072_0_receptor_fixed.pdbqt\n",
      "Mpro-x0072: Estimated Free Energy of Binding = 28.271 kcal/mol\n",
      "Output saved as Mpro-x0072_0_receptor_fixed_molecules_bx_2024_06_18_154410_11.out\n",
      "\n",
      "Data saved to vina_docking_data_playground.csv\n",
      "Data saved in data/molecules_bx_2024_06_18_154410_11_with_Mpro-x0072_0_receptor_fixed\n",
      "The ligand file has been converted to molecules_bx_2024_06_18_154410_12.pdb\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "cp: -r not specified; omitting directory 'Mpro-x0072_0_receptor_fixed'\n",
      "mv: cannot move '/home/ian/msc_project/docking/Mpro-x0072_0_receptor_fixed' to 'data/molecules_bx_2024_06_18_154410_11_with_Mpro-x0072_0_receptor_fixed/Mpro-x0072_0_receptor_fixed': Directory not empty\n",
      "1 molecule converted\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The ligand file has been saved as molecules_bx_2024_06_18_154410_12.pdbqt\n",
      "adding gasteiger charges to peptide\n",
      "The protein file Mpro-x0072_0_receptor_fixed.pdb has been converted to pdbqt format and saved in Mpro-x0072_0_receptor_fixed.pdbqt\n",
      "Mpro-x0072: Estimated Free Energy of Binding = 26.477 kcal/mol\n",
      "Output saved as Mpro-x0072_0_receptor_fixed_molecules_bx_2024_06_18_154410_12.out\n",
      "\n",
      "Data saved to vina_docking_data_playground.csv\n",
      "Data saved in data/molecules_bx_2024_06_18_154410_12_with_Mpro-x0072_0_receptor_fixed\n",
      "The ligand file has been converted to molecules_bx_2024_06_18_154410_13.pdb\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "cp: -r not specified; omitting directory 'Mpro-x0072_0_receptor_fixed'\n",
      "mv: cannot move '/home/ian/msc_project/docking/Mpro-x0072_0_receptor_fixed' to 'data/molecules_bx_2024_06_18_154410_12_with_Mpro-x0072_0_receptor_fixed/Mpro-x0072_0_receptor_fixed': Directory not empty\n",
      "1 molecule converted\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The ligand file has been saved as molecules_bx_2024_06_18_154410_13.pdbqt\n",
      "adding gasteiger charges to peptide\n",
      "The protein file Mpro-x0072_0_receptor_fixed.pdb has been converted to pdbqt format and saved in Mpro-x0072_0_receptor_fixed.pdbqt\n",
      "Mpro-x0072: Estimated Free Energy of Binding = 17.754 kcal/mol\n",
      "Output saved as Mpro-x0072_0_receptor_fixed_molecules_bx_2024_06_18_154410_13.out\n",
      "\n",
      "Data saved to vina_docking_data_playground.csv\n",
      "Data saved in data/molecules_bx_2024_06_18_154410_13_with_Mpro-x0072_0_receptor_fixed\n",
      "The ligand file has been converted to molecules_bx_2024_06_18_154410_14.pdb\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "cp: -r not specified; omitting directory 'Mpro-x0072_0_receptor_fixed'\n",
      "mv: cannot move '/home/ian/msc_project/docking/Mpro-x0072_0_receptor_fixed' to 'data/molecules_bx_2024_06_18_154410_13_with_Mpro-x0072_0_receptor_fixed/Mpro-x0072_0_receptor_fixed': Directory not empty\n",
      "1 molecule converted\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The ligand file has been saved as molecules_bx_2024_06_18_154410_14.pdbqt\n",
      "adding gasteiger charges to peptide\n",
      "The protein file Mpro-x0072_0_receptor_fixed.pdb has been converted to pdbqt format and saved in Mpro-x0072_0_receptor_fixed.pdbqt\n",
      "Mpro-x0072: Estimated Free Energy of Binding = 25.664 kcal/mol\n",
      "Output saved as Mpro-x0072_0_receptor_fixed_molecules_bx_2024_06_18_154410_14.out\n",
      "\n",
      "Data saved to vina_docking_data_playground.csv\n",
      "Data saved in data/molecules_bx_2024_06_18_154410_14_with_Mpro-x0072_0_receptor_fixed\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "cp: -r not specified; omitting directory 'Mpro-x0072_0_receptor_fixed'\n",
      "mv: cannot move '/home/ian/msc_project/docking/Mpro-x0072_0_receptor_fixed' to 'data/molecules_bx_2024_06_18_154410_14_with_Mpro-x0072_0_receptor_fixed/Mpro-x0072_0_receptor_fixed': Directory not empty\n"
     ]
    }
   ],
   "source": [
    "lig_file='molecules_bx_2024_06_18_154410.sdf' # this is when the sdf file is just 1 molecule\n",
    "lig_file = split_sdf(lig_file)\n",
    "prot_file='Mpro-x0072_0_receptor_fixed.pdb' # replace with a protein file of interest\n",
    "if isinstance(lig_file,list):\n",
    "    for lig in lig_file:\n",
    "        vina_process_lig_prot(lig_file=lig, \n",
    "                 prot_file=prot_file,\n",
    "                 csv_out_file='vina_docking_data_playground.csv')\n",
    "else:\n",
    "    vina_process_lig_prot(lig_file=lig_file, \n",
    "                prot_file=prot_file,\n",
    "                csv_out_file='vina_docking_data_playground.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Testing: OE Pipeline\n",
    "\n",
    "<font color='yellow'>Testing with complexes produced from splitting PDB IDs</font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|          | 0/8 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Downloading PDB structure '6r8o'...\n",
      "Extracting ligand from 6R8O.pdb...\n",
      "Reading receptor from 6R8O.pdb...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "DPI: 0.05, RFree: 0.18, Resolution: 1.36\n",
      "Processing BU # 1 with title: PEPTIDYL-PROLYL CIS-TRANS ISOMERASE F, MITOCHONDRIA, chains A\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The complex file has been saved as 6R8O.pdb\n",
      "The receptor file has been saved as 6R8O_receptor.pdb\n",
      "The ligand file has been saved as 6R8O_ligand.pdb\n",
      "Preparing the ligand and protein files using OpenEye Toolkits...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "DPI: 0.05, RFree: 0.18, Resolution: 1.36\n",
      "Processing BU # 1 with title: PEPTIDYL-PROLYL CIS-TRANS ISOMERASE F, MITOCHONDRIA, chains A, alt: A\n",
      "   Falling back to charging protein with OEMMFF94Charges\n",
      "Processing BU # 2 with title: PEPTIDYL-PROLYL CIS-TRANS ISOMERASE F, MITOCHONDRIA, chains A, alt: B\n",
      "   Falling back to charging protein with OEMMFF94Charges\n",
      "Skipping redundant DU with alts outside the site of interest, renaming existing to collapse alts\n",
      "Discarding redundant alt DU with title PEPTIDYL-PROLYL CIS-TRANS ISOMERASE F, MITOCHONDRIA(A)altB > JV2(A-304)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Design unit was successfully made for 6R8O.pdb, output is saved to ['6R8O_DU_0.oedu'].\n",
      "The receptor design unit file has been saved as 6R8O_DU_0_receptor.oedu\n",
      "cleaning JV2(A-304)\n",
      "docking JV2(A-304)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "mv: cannot move '6R8O_ligand_6R8O_receptor' to a subdirectory of itself, '/home/ian/msc_project/oe_docking_leakypdb_tmp/6R8O_ligand_6R8O_receptor/6R8O_ligand_6R8O_receptor'\n",
      "mv: cannot move '6R8O_ligand_6R8O_receptor' to a subdirectory of itself, '/home/ian/msc_project/oe_docking_leakypdb_tmp/6R8O_ligand_6R8O_receptor/6R8O_ligand_6R8O_receptor'\n",
      "mv: cannot move '6R8O_ligand_6R8O_receptor' to a subdirectory of itself, '/home/ian/msc_project/oe_docking_leakypdb_tmp/6R8O_ligand_6R8O_receptor/6R8O_ligand_6R8O_receptor'\n",
      " 12%|        | 1/8 [01:55<13:30, 115.75s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The docked ligand file has been saved as ['6R8O_DU_0_receptor_6R8O_ligand_docked.sdf']\n",
      "Output saved as 6R8O_receptor_6R8O_ligand_0.out\n",
      "\n",
      "Data saved to OE_w_water_data.csv\n",
      "Docked ligand 6R8O_DU_0_receptor_6R8O_ligand_docked.sdf, molecule 1: Chemgauss4 score = -15.09\n",
      "Data saved in ./data//home/ian/msc_project/oe_docking_leakypdb_tmp/6R8O_ligand_6R8O_receptor\n",
      "Downloading PDB structure '5s9l'...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ian/micromamba/envs/asapdiscovery/lib/python3.10/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 7409.\n",
      "  warnings.warn(\n",
      "DPI: 0.12, RFree: 0.19, Resolution: 1.90\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Extracting ligand from 5S9L.pdb...\n",
      "Reading receptor from 5S9L.pdb...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing BU # 1 with title: ISOFORM 2 OF ECTONUCLEOTIDE, chains AB\n",
      "Warning: No rotamer of residue: GLU 573   A 1   passed a clash test, cluster will be skipped\n",
      "Found unresolved N-terminal with 23 residues before TRP 51   A 1  , with sequence FTASRIKRAEWDEGPPTVLSDSP\n",
      "Found 4 residue gap between LYS 463   A 1   and CYS 468   A 1  , with sequence PSGK\n",
      "Found 2 residue gap between GLU 573   A 1   and ASN 576   A 1  , with sequence PK\n",
      "Found unresolved C-terminal with 12 residues after GLU 861   A 1  , with sequence IGGRHHHHHHHH\n",
      "Warning: There was a problem building some missing pieces, built as much as was possible\n",
      "   Falling back to charging protein with OEMMFF94Charges\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The complex file has been saved as 5S9L.pdb\n",
      "The receptor file has been saved as 5S9L_receptor.pdb\n",
      "The ligand file has been saved as 5S9L_ligand.pdb\n",
      "Preparing the ligand and protein files using OpenEye Toolkits...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "DPI: 0.12, RFree: 0.19, Resolution: 1.90\n",
      "Warning: GetGroupAndLocation(' CA :SER725 ' chain='A' alt='A' model=0, confCode='B', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code B on  CA  in SER 725   A 1 A\n",
      "Warning: GetGroupAndLocation(' CA :SER725 ' chain='A' alt='C' model=0, confCode='B', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code B on  CA  in SER 725   A 1 C\n",
      "Warning: GetGroupAndLocation(' CB :SER725 ' chain='A' alt='A' model=0, confCode='B', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code B on  CB  in SER 725   A 1 A\n",
      "Warning: GetGroupAndLocation(' CB :SER725 ' chain='A' alt='C' model=0, confCode='B', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code B on  CB  in SER 725   A 1 C\n",
      "Warning: GetGroupAndLocation(' OG :SER725 ' chain='A' alt='A' model=0, confCode='B', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code B on  OG  in SER 725   A 1 A\n",
      "Warning: GetGroupAndLocation(' OG :SER725 ' chain='A' alt='C' model=0, confCode='B', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code B on  OG  in SER 725   A 1 C\n",
      "Warning: AltFactory problem enumerating alternate locations using altCode: B\n",
      "Warning: GetGroupAndLocation(' CA :CYS148 ' chain='A' alt='A' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CA  in CYS 148   A 1 A\n",
      "Warning: GetGroupAndLocation(' CA :CYS148 ' chain='A' alt='B' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CA  in CYS 148   A 1 B\n",
      "Warning: GetGroupAndLocation(' CB :CYS148 ' chain='A' alt='A' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CB  in CYS 148   A 1 A\n",
      "Warning: GetGroupAndLocation(' CB :CYS148 ' chain='A' alt='B' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CB  in CYS 148   A 1 B\n",
      "Warning: GetGroupAndLocation(' SG :CYS148 ' chain='A' alt='A' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  SG  in CYS 148   A 1 A\n",
      "Warning: GetGroupAndLocation(' SG :CYS148 ' chain='A' alt='B' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  SG  in CYS 148   A 1 B\n",
      "Warning: GetGroupAndLocation(' CA :CYS194 ' chain='A' alt='A' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CA  in CYS 194   A 1 A\n",
      "Warning: GetGroupAndLocation(' CA :CYS194 ' chain='A' alt='B' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CA  in CYS 194   A 1 B\n",
      "Warning: GetGroupAndLocation(' CB :CYS194 ' chain='A' alt='A' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CB  in CYS 194   A 1 A\n",
      "Warning: GetGroupAndLocation(' CB :CYS194 ' chain='A' alt='B' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CB  in CYS 194   A 1 B\n",
      "Warning: GetGroupAndLocation(' SG :CYS194 ' chain='A' alt='A' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  SG  in CYS 194   A 1 A\n",
      "Warning: GetGroupAndLocation(' SG :CYS194 ' chain='A' alt='B' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  SG  in CYS 194   A 1 B\n",
      "Warning: GetGroupAndLocation(' CA :GLU189 ' chain='A' alt='A' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CA  in GLU 189   A 1 A\n",
      "Warning: GetGroupAndLocation(' CA :GLU189 ' chain='A' alt='B' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CA  in GLU 189   A 1 B\n",
      "Warning: GetGroupAndLocation(' CB :GLU189 ' chain='A' alt='A' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CB  in GLU 189   A 1 A\n",
      "Warning: GetGroupAndLocation(' CB :GLU189 ' chain='A' alt='B' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CB  in GLU 189   A 1 B\n",
      "Warning: GetGroupAndLocation(' CG :GLU189 ' chain='A' alt='A' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CG  in GLU 189   A 1 A\n",
      "Warning: GetGroupAndLocation(' CG :GLU189 ' chain='A' alt='B' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CG  in GLU 189   A 1 B\n",
      "Warning: GetGroupAndLocation(' CD :GLU189 ' chain='A' alt='A' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CD  in GLU 189   A 1 A\n",
      "Warning: GetGroupAndLocation(' CD :GLU189 ' chain='A' alt='B' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CD  in GLU 189   A 1 B\n",
      "Warning: GetGroupAndLocation(' OE1:GLU189 ' chain='A' alt='A' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  OE1 in GLU 189   A 1 A\n",
      "Warning: GetGroupAndLocation(' OE1:GLU189 ' chain='A' alt='B' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  OE1 in GLU 189   A 1 B\n",
      "Warning: GetGroupAndLocation(' OE2:GLU189 ' chain='A' alt='A' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  OE2 in GLU 189   A 1 A\n",
      "Warning: GetGroupAndLocation(' OE2:GLU189 ' chain='A' alt='B' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  OE2 in GLU 189   A 1 B\n",
      "Warning: GetGroupAndLocation(' CA :ARG450 ' chain='A' alt='A' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CA  in ARG 450   A 1 A\n",
      "Warning: GetGroupAndLocation(' CA :ARG450 ' chain='A' alt='B' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CA  in ARG 450   A 1 B\n",
      "Warning: GetGroupAndLocation(' CB :ARG450 ' chain='A' alt='A' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CB  in ARG 450   A 1 A\n",
      "Warning: GetGroupAndLocation(' CB :ARG450 ' chain='A' alt='B' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CB  in ARG 450   A 1 B\n",
      "Warning: GetGroupAndLocation(' CG :ARG450 ' chain='A' alt='A' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CG  in ARG 450   A 1 A\n",
      "Warning: GetGroupAndLocation(' CG :ARG450 ' chain='A' alt='B' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CG  in ARG 450   A 1 B\n",
      "Warning: GetGroupAndLocation(' CD :ARG450 ' chain='A' alt='A' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CD  in ARG 450   A 1 A\n",
      "Warning: GetGroupAndLocation(' CD :ARG450 ' chain='A' alt='B' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CD  in ARG 450   A 1 B\n",
      "Warning: GetGroupAndLocation(' NE :ARG450 ' chain='A' alt='A' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  NE  in ARG 450   A 1 A\n",
      "Warning: GetGroupAndLocation(' NE :ARG450 ' chain='A' alt='B' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  NE  in ARG 450   A 1 B\n",
      "Warning: GetGroupAndLocation(' CZ :ARG450 ' chain='A' alt='A' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CZ  in ARG 450   A 1 A\n",
      "Warning: GetGroupAndLocation(' CZ :ARG450 ' chain='A' alt='B' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CZ  in ARG 450   A 1 B\n",
      "Warning: GetGroupAndLocation(' NH1:ARG450 ' chain='A' alt='A' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  NH1 in ARG 450   A 1 A\n",
      "Warning: GetGroupAndLocation(' NH1:ARG450 ' chain='A' alt='B' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  NH1 in ARG 450   A 1 B\n",
      "Warning: GetGroupAndLocation(' NH2:ARG450 ' chain='A' alt='A' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  NH2 in ARG 450   A 1 A\n",
      "Warning: GetGroupAndLocation(' NH2:ARG450 ' chain='A' alt='B' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  NH2 in ARG 450   A 1 B\n",
      "Warning: GetGroupAndLocation(' CA :SER608 ' chain='A' alt='A' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CA  in SER 608   A 1 A\n",
      "Warning: GetGroupAndLocation(' CA :SER608 ' chain='A' alt='B' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CA  in SER 608   A 1 B\n",
      "Warning: GetGroupAndLocation(' CB :SER608 ' chain='A' alt='A' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CB  in SER 608   A 1 A\n",
      "Warning: GetGroupAndLocation(' CB :SER608 ' chain='A' alt='B' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CB  in SER 608   A 1 B\n",
      "Warning: GetGroupAndLocation(' OG :SER608 ' chain='A' alt='A' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  OG  in SER 608   A 1 A\n",
      "Warning: GetGroupAndLocation(' OG :SER608 ' chain='A' alt='B' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  OG  in SER 608   A 1 B\n",
      "Warning: GetGroupAndLocation(' CA :GLU618 ' chain='A' alt='A' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CA  in GLU 618   A 1 A\n",
      "Warning: GetGroupAndLocation(' CA :GLU618 ' chain='A' alt='B' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CA  in GLU 618   A 1 B\n",
      "Warning: GetGroupAndLocation(' CB :GLU618 ' chain='A' alt='A' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CB  in GLU 618   A 1 A\n",
      "Warning: GetGroupAndLocation(' CB :GLU618 ' chain='A' alt='B' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CB  in GLU 618   A 1 B\n",
      "Warning: GetGroupAndLocation(' CG :GLU618 ' chain='A' alt='A' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CG  in GLU 618   A 1 A\n",
      "Warning: GetGroupAndLocation(' CG :GLU618 ' chain='A' alt='B' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CG  in GLU 618   A 1 B\n",
      "Warning: GetGroupAndLocation(' CD :GLU618 ' chain='A' alt='A' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CD  in GLU 618   A 1 A\n",
      "Warning: GetGroupAndLocation(' CD :GLU618 ' chain='A' alt='B' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CD  in GLU 618   A 1 B\n",
      "Warning: GetGroupAndLocation(' OE1:GLU618 ' chain='A' alt='A' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  OE1 in GLU 618   A 1 A\n",
      "Warning: GetGroupAndLocation(' OE1:GLU618 ' chain='A' alt='B' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  OE1 in GLU 618   A 1 B\n",
      "Warning: GetGroupAndLocation(' OE2:GLU618 ' chain='A' alt='A' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  OE2 in GLU 618   A 1 A\n",
      "Warning: GetGroupAndLocation(' OE2:GLU618 ' chain='A' alt='B' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  OE2 in GLU 618   A 1 B\n",
      "Warning: GetGroupAndLocation(' CA :SER659 ' chain='A' alt='A' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CA  in SER 659   A 1 A\n",
      "Warning: GetGroupAndLocation(' CA :SER659 ' chain='A' alt='B' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CA  in SER 659   A 1 B\n",
      "Warning: GetGroupAndLocation(' CB :SER659 ' chain='A' alt='A' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CB  in SER 659   A 1 A\n",
      "Warning: GetGroupAndLocation(' CB :SER659 ' chain='A' alt='B' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CB  in SER 659   A 1 B\n",
      "Warning: GetGroupAndLocation(' OG :SER659 ' chain='A' alt='A' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  OG  in SER 659   A 1 A\n",
      "Warning: GetGroupAndLocation(' OG :SER659 ' chain='A' alt='B' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  OG  in SER 659   A 1 B\n",
      "Warning: GetGroupAndLocation(' CA :MET701 ' chain='A' alt='A' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CA  in MET 701   A 1 A\n",
      "Warning: GetGroupAndLocation(' CA :MET701 ' chain='A' alt='B' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CA  in MET 701   A 1 B\n",
      "Warning: GetGroupAndLocation(' CB :MET701 ' chain='A' alt='A' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CB  in MET 701   A 1 A\n",
      "Warning: GetGroupAndLocation(' CB :MET701 ' chain='A' alt='B' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CB  in MET 701   A 1 B\n",
      "Warning: GetGroupAndLocation(' CG :MET701 ' chain='A' alt='A' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CG  in MET 701   A 1 A\n",
      "Warning: GetGroupAndLocation(' CG :MET701 ' chain='A' alt='B' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CG  in MET 701   A 1 B\n",
      "Warning: GetGroupAndLocation(' SD :MET701 ' chain='A' alt='A' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  SD  in MET 701   A 1 A\n",
      "Warning: GetGroupAndLocation(' SD :MET701 ' chain='A' alt='B' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  SD  in MET 701   A 1 B\n",
      "Warning: GetGroupAndLocation(' CE :MET701 ' chain='A' alt='A' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CE  in MET 701   A 1 A\n",
      "Warning: GetGroupAndLocation(' CE :MET701 ' chain='A' alt='B' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CE  in MET 701   A 1 B\n",
      "Warning: GetGroupAndLocation(' CA :LYS721 ' chain='A' alt='A' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CA  in LYS 721   A 1 A\n",
      "Warning: GetGroupAndLocation(' CA :LYS721 ' chain='A' alt='B' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CA  in LYS 721   A 1 B\n",
      "Warning: GetGroupAndLocation(' CB :LYS721 ' chain='A' alt='A' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CB  in LYS 721   A 1 A\n",
      "Warning: GetGroupAndLocation(' CB :LYS721 ' chain='A' alt='B' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CB  in LYS 721   A 1 B\n",
      "Warning: GetGroupAndLocation(' CG :LYS721 ' chain='A' alt='A' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CG  in LYS 721   A 1 A\n",
      "Warning: GetGroupAndLocation(' CG :LYS721 ' chain='A' alt='B' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CG  in LYS 721   A 1 B\n",
      "Warning: GetGroupAndLocation(' CD :LYS721 ' chain='A' alt='A' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CD  in LYS 721   A 1 A\n",
      "Warning: GetGroupAndLocation(' CD :LYS721 ' chain='A' alt='B' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CD  in LYS 721   A 1 B\n",
      "Warning: GetGroupAndLocation(' CE :LYS721 ' chain='A' alt='A' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CE  in LYS 721   A 1 A\n",
      "Warning: GetGroupAndLocation(' CE :LYS721 ' chain='A' alt='B' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CE  in LYS 721   A 1 B\n",
      "Warning: GetGroupAndLocation(' NZ :LYS721 ' chain='A' alt='A' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  NZ  in LYS 721   A 1 A\n",
      "Warning: GetGroupAndLocation(' NZ :LYS721 ' chain='A' alt='B' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  NZ  in LYS 721   A 1 B\n",
      "Warning: GetGroupAndLocation(' CA :ASN806 ' chain='A' alt='A' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CA  in ASN 806   A 1 A\n",
      "Warning: GetGroupAndLocation(' CA :ASN806 ' chain='A' alt='B' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CA  in ASN 806   A 1 B\n",
      "Warning: GetGroupAndLocation(' CB :ASN806 ' chain='A' alt='A' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CB  in ASN 806   A 1 A\n",
      "Warning: GetGroupAndLocation(' CB :ASN806 ' chain='A' alt='B' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CB  in ASN 806   A 1 B\n",
      "Warning: GetGroupAndLocation(' CG :ASN806 ' chain='A' alt='A' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CG  in ASN 806   A 1 A\n",
      "Warning: GetGroupAndLocation(' CG :ASN806 ' chain='A' alt='B' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CG  in ASN 806   A 1 B\n",
      "Warning: GetGroupAndLocation(' OD1:ASN806 ' chain='A' alt='A' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  OD1 in ASN 806   A 1 A\n",
      "Warning: GetGroupAndLocation(' OD1:ASN806 ' chain='A' alt='B' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  OD1 in ASN 806   A 1 B\n",
      "Warning: GetGroupAndLocation(' ND2:ASN806 ' chain='A' alt='A' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  ND2 in ASN 806   A 1 A\n",
      "Warning: GetGroupAndLocation(' ND2:ASN806 ' chain='A' alt='B' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  ND2 in ASN 806   A 1 B\n",
      "Warning: GetGroupAndLocation(' CA :SER808 ' chain='A' alt='A' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CA  in SER 808   A 1 A\n",
      "Warning: GetGroupAndLocation(' CA :SER808 ' chain='A' alt='B' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CA  in SER 808   A 1 B\n",
      "Warning: GetGroupAndLocation(' CB :SER808 ' chain='A' alt='A' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CB  in SER 808   A 1 A\n",
      "Warning: GetGroupAndLocation(' CB :SER808 ' chain='A' alt='B' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CB  in SER 808   A 1 B\n",
      "Warning: GetGroupAndLocation(' OG :SER808 ' chain='A' alt='A' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  OG  in SER 808   A 1 A\n",
      "Warning: GetGroupAndLocation(' OG :SER808 ' chain='A' alt='B' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  OG  in SER 808   A 1 B\n",
      "Warning: GetGroupAndLocation(' CA :MET821 ' chain='A' alt='A' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CA  in MET 821   A 1 A\n",
      "Warning: GetGroupAndLocation(' CA :MET821 ' chain='A' alt='B' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CA  in MET 821   A 1 B\n",
      "Warning: GetGroupAndLocation(' CB :MET821 ' chain='A' alt='A' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CB  in MET 821   A 1 A\n",
      "Warning: GetGroupAndLocation(' CB :MET821 ' chain='A' alt='B' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CB  in MET 821   A 1 B\n",
      "Warning: GetGroupAndLocation(' CG :MET821 ' chain='A' alt='A' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CG  in MET 821   A 1 A\n",
      "Warning: GetGroupAndLocation(' CG :MET821 ' chain='A' alt='B' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CG  in MET 821   A 1 B\n",
      "Warning: GetGroupAndLocation(' SD :MET821 ' chain='A' alt='A' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  SD  in MET 821   A 1 A\n",
      "Warning: GetGroupAndLocation(' SD :MET821 ' chain='A' alt='B' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  SD  in MET 821   A 1 B\n",
      "Warning: GetGroupAndLocation(' CE :MET821 ' chain='A' alt='A' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CE  in MET 821   A 1 A\n",
      "Warning: GetGroupAndLocation(' CE :MET821 ' chain='A' alt='B' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CE  in MET 821   A 1 B\n",
      "Warning: GetGroupAndLocation(' N  :SER842 ' chain='A' alt='A' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  N   in SER 842   A 1 A\n",
      "Warning: GetGroupAndLocation(' N  :SER842 ' chain='A' alt='B' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  N   in SER 842   A 1 B\n",
      "Warning: GetGroupAndLocation(' CA :SER842 ' chain='A' alt='A' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CA  in SER 842   A 1 A\n",
      "Warning: GetGroupAndLocation(' CA :SER842 ' chain='A' alt='B' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CA  in SER 842   A 1 B\n",
      "Warning: GetGroupAndLocation(' C  :SER842 ' chain='A' alt='A' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  C   in SER 842   A 1 A\n",
      "Warning: GetGroupAndLocation(' C  :SER842 ' chain='A' alt='B' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  C   in SER 842   A 1 B\n",
      "Warning: GetGroupAndLocation(' O  :SER842 ' chain='A' alt='A' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  O   in SER 842   A 1 A\n",
      "Warning: GetGroupAndLocation(' O  :SER842 ' chain='A' alt='B' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  O   in SER 842   A 1 B\n",
      "Warning: GetGroupAndLocation(' CB :SER842 ' chain='A' alt='A' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CB  in SER 842   A 1 A\n",
      "Warning: GetGroupAndLocation(' CB :SER842 ' chain='A' alt='B' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CB  in SER 842   A 1 B\n",
      "Warning: GetGroupAndLocation(' OG :SER842 ' chain='A' alt='A' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  OG  in SER 842   A 1 A\n",
      "Warning: GetGroupAndLocation(' OG :SER842 ' chain='A' alt='B' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  OG  in SER 842   A 1 B\n",
      "Warning: GetGroupAndLocation(' CA :SER844 ' chain='A' alt='A' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CA  in SER 844   A 1 A\n",
      "Warning: GetGroupAndLocation(' CA :SER844 ' chain='A' alt='B' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CA  in SER 844   A 1 B\n",
      "Warning: GetGroupAndLocation(' CB :SER844 ' chain='A' alt='A' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CB  in SER 844   A 1 A\n",
      "Warning: GetGroupAndLocation(' CB :SER844 ' chain='A' alt='B' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CB  in SER 844   A 1 B\n",
      "Warning: GetGroupAndLocation(' OG :SER844 ' chain='A' alt='A' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  OG  in SER 844   A 1 A\n",
      "Warning: GetGroupAndLocation(' OG :SER844 ' chain='A' alt='B' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  OG  in SER 844   A 1 B\n",
      "Warning: GetGroupAndLocation(' CA :SER846 ' chain='A' alt='A' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CA  in SER 846   A 1 A\n",
      "Warning: GetGroupAndLocation(' CA :SER846 ' chain='A' alt='B' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CA  in SER 846   A 1 B\n",
      "Warning: GetGroupAndLocation(' CB :SER846 ' chain='A' alt='A' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CB  in SER 846   A 1 A\n",
      "Warning: GetGroupAndLocation(' CB :SER846 ' chain='A' alt='B' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  CB  in SER 846   A 1 B\n",
      "Warning: GetGroupAndLocation(' OG :SER846 ' chain='A' alt='A' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  OG  in SER 846   A 1 A\n",
      "Warning: GetGroupAndLocation(' OG :SER846 ' chain='A' alt='B' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  OG  in SER 846   A 1 B\n",
      "Warning: GetGroupAndLocation(' O  :HOH1354 ' chain='A' alt='A' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  O   in HOH 1354   A 2 A\n",
      "Warning: GetGroupAndLocation(' O  :HOH1354 ' chain='A' alt='B' model=0, confCode='C', caseSensitive) - atom not found in alt records\n",
      "Warning: Problem setting alternate location code C on  O   in HOH 1354   A 2 B\n",
      "Warning: AltFactory problem enumerating alternate locations using altCode: C\n",
      "Processing BU # 1 with title: ISOFORM 2 OF ECTONUCLEOTIDE, chains AB, alt: A\n",
      "Warning: No rotamer of residue: GLU 573   A 1   passed a clash test, cluster will be skipped\n",
      "Found unresolved N-terminal with 23 residues before TRP 51   A 1  , with sequence FTASRIKRAEWDEGPPTVLSDSP\n",
      "Found 4 residue gap between LYS 463   A 1   and CYS 468   A 1  , with sequence PSGK\n",
      "Found 2 residue gap between GLU 573   A 1   and ASN 576   A 1  , with sequence PK\n",
      "Found unresolved C-terminal with 12 residues after GLU 861   A 1  , with sequence IGGRHHHHHHHH\n",
      "Warning: There was a problem building some missing pieces, built as much as was possible\n",
      "   Falling back to charging protein with OEMMFF94Charges\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Design unit was successfully made for 5S9L.pdb, output is saved to ['5S9L_DU_0.oedu'].\n",
      "The receptor design unit file has been saved as 5S9L_DU_0_receptor.oedu\n",
      "cleaning 6ZO(A-906)\n",
      "docking 6ZO(A-906)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "mv: cannot move '5S9L_ligand_5S9L_receptor' to a subdirectory of itself, '/home/ian/msc_project/oe_docking_leakypdb_tmp/5S9L_ligand_5S9L_receptor/5S9L_ligand_5S9L_receptor'\n",
      "mv: cannot move '5S9L_ligand_5S9L_receptor' to a subdirectory of itself, '/home/ian/msc_project/oe_docking_leakypdb_tmp/5S9L_ligand_5S9L_receptor/5S9L_ligand_5S9L_receptor'\n",
      "mv: cannot move '5S9L_ligand_5S9L_receptor' to a subdirectory of itself, '/home/ian/msc_project/oe_docking_leakypdb_tmp/5S9L_ligand_5S9L_receptor/5S9L_ligand_5S9L_receptor'\n",
      " 25%|       | 2/8 [03:24<09:59, 99.88s/it] "
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The docked ligand file has been saved as ['5S9L_DU_0_receptor_5S9L_ligand_docked.sdf']\n",
      "Output saved as 5S9L_receptor_5S9L_ligand_0.out\n",
      "\n",
      "Data saved to OE_w_water_data.csv\n",
      "Docked ligand 5S9L_DU_0_receptor_5S9L_ligand_docked.sdf, molecule 1: Chemgauss4 score = -14.36\n",
      "Data saved in ./data//home/ian/msc_project/oe_docking_leakypdb_tmp/5S9L_ligand_5S9L_receptor\n",
      "Downloading PDB structure '1klt'...\n",
      "Extracting ligand from 1KLT.pdb...\n",
      "Reading receptor from 1KLT.pdb...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "DPI: 0.22, RFree: 0.26, Resolution: 1.90\n",
      "Processing BU # 1 with title: CHYMASE, chains A\n",
      "Warning: For residue LYS 109   A 1   removing clashing solvent molecule HOH 533   A 2  \n",
      "Warning: For residue LYS 109   A 1   removing clashing solvent molecule HOH 531   A 2  \n",
      "Warning: For residue ARG 137   A 1   removing clashing solvent molecule HOH 591   A 2  \n",
      "Warning: For 'PMS', reference state application resulted in 3 results in 3D\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The complex file has been saved as 1KLT.pdb\n",
      "The receptor file has been saved as 1KLT_receptor.pdb\n",
      "The ligand file has been saved as 1KLT_ligand.pdb\n",
      "Preparing the ligand and protein files using OpenEye Toolkits...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "DPI: 0.22, RFree: 0.26, Resolution: 1.90\n",
      "Processing BU # 1 with title: CHYMASE, chains A\n",
      "Warning: For residue LYS 109   A 1   removing clashing solvent molecule HOH 533   A 2  \n",
      "Warning: For residue LYS 109   A 1   removing clashing solvent molecule HOH 531   A 2  \n",
      "Warning: For residue ARG 137   A 1   removing clashing solvent molecule HOH 591   A 2  \n",
      "Warning: For 'PMS', reference state application resulted in 3 results in 3D\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Design unit was successfully made for 1KLT.pdb, output is saved to ['1KLT_DU_0.oedu'].\n",
      "The receptor design unit file has been saved as 1KLT_DU_0_receptor.oedu\n",
      "cleaning PMS(A-400)\n",
      "docking PMS(A-400)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "mv: cannot move '1KLT_ligand_1KLT_receptor' to a subdirectory of itself, '/home/ian/msc_project/oe_docking_leakypdb_tmp/1KLT_ligand_1KLT_receptor/1KLT_ligand_1KLT_receptor'\n",
      "mv: cannot move '1KLT_ligand_1KLT_receptor' to a subdirectory of itself, '/home/ian/msc_project/oe_docking_leakypdb_tmp/1KLT_ligand_1KLT_receptor/1KLT_ligand_1KLT_receptor'\n",
      "mv: cannot move '1KLT_ligand_1KLT_receptor' to a subdirectory of itself, '/home/ian/msc_project/oe_docking_leakypdb_tmp/1KLT_ligand_1KLT_receptor/1KLT_ligand_1KLT_receptor'\n",
      " 38%|      | 3/8 [04:25<06:49, 81.99s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The docked ligand file has been saved as ['1KLT_DU_0_receptor_1KLT_ligand_docked.sdf']\n",
      "Output saved as 1KLT_receptor_1KLT_ligand_0.out\n",
      "\n",
      "Data saved to OE_w_water_data.csv\n",
      "Docked ligand 1KLT_DU_0_receptor_1KLT_ligand_docked.sdf, molecule 1: Chemgauss4 score = -7.52\n",
      "Data saved in ./data//home/ian/msc_project/oe_docking_leakypdb_tmp/1KLT_ligand_1KLT_receptor\n",
      "Downloading PDB structure '1a42'...\n",
      "Extracting ligand from 1A42.pdb...\n",
      "Reading receptor from 1A42.pdb...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning: Cannot set Rfree, due to lack of R value for CARBONIC ANHYDRASE II\n",
      "DPI: 0.00, RFree: 0.00, Resolution: 2.25\n",
      "Processing BU # 1 with title: CARBONIC ANHYDRASE II, chains A\n",
      "Found unresolved N-terminal with 2 residues before HIS 4   A 1  , with sequence SH\n",
      "Found unresolved C-terminal with 1 residues after PHE 260   A 1  , with sequence K\n",
      "Warning: There was a problem building some missing pieces, built as much as was possible\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The complex file has been saved as 1A42.pdb\n",
      "The receptor file has been saved as 1A42_receptor.pdb\n",
      "The ligand file has been saved as 1A42_ligand.pdb\n",
      "Preparing the ligand and protein files using OpenEye Toolkits...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning: Cannot set Rfree, due to lack of R value for CARBONIC ANHYDRASE II\n",
      "DPI: 0.00, RFree: 0.00, Resolution: 2.25\n",
      "Processing BU # 1 with title: CARBONIC ANHYDRASE II, chains A\n",
      "Found unresolved N-terminal with 2 residues before HIS 4   A 1  , with sequence SH\n",
      "Found unresolved C-terminal with 1 residues after PHE 260   A 1  , with sequence K\n",
      "Warning: There was a problem building some missing pieces, built as much as was possible\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Design unit was successfully made for 1A42.pdb, output is saved to ['1A42_DU_0.oedu'].\n",
      "The receptor design unit file has been saved as 1A42_DU_0_receptor.oedu\n",
      "cleaning BZU(A-555)\n",
      "docking BZU(A-555)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "mv: cannot move '1A42_ligand_1A42_receptor' to a subdirectory of itself, '/home/ian/msc_project/oe_docking_leakypdb_tmp/1A42_ligand_1A42_receptor/1A42_ligand_1A42_receptor'\n",
      "mv: cannot move '1A42_ligand_1A42_receptor' to a subdirectory of itself, '/home/ian/msc_project/oe_docking_leakypdb_tmp/1A42_ligand_1A42_receptor/1A42_ligand_1A42_receptor'\n",
      "mv: cannot move '1A42_ligand_1A42_receptor' to a subdirectory of itself, '/home/ian/msc_project/oe_docking_leakypdb_tmp/1A42_ligand_1A42_receptor/1A42_ligand_1A42_receptor'\n",
      " 50%|     | 4/8 [05:08<04:27, 66.76s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The docked ligand file has been saved as ['1A42_DU_0_receptor_1A42_ligand_docked.sdf']\n",
      "Output saved as 1A42_receptor_1A42_ligand_0.out\n",
      "\n",
      "Data saved to OE_w_water_data.csv\n",
      "Docked ligand 1A42_DU_0_receptor_1A42_ligand_docked.sdf, molecule 1: Chemgauss4 score = -9.89\n",
      "Data saved in ./data//home/ian/msc_project/oe_docking_leakypdb_tmp/1A42_ligand_1A42_receptor\n",
      "Downloading PDB structure '4b3d'...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ian/micromamba/envs/asapdiscovery/lib/python3.10/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 4092.\n",
      "  warnings.warn(\n",
      "/home/ian/micromamba/envs/asapdiscovery/lib/python3.10/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain C is discontinuous at line 4117.\n",
      "  warnings.warn(\n",
      "/home/ian/micromamba/envs/asapdiscovery/lib/python3.10/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 4142.\n",
      "  warnings.warn(\n",
      "/home/ian/micromamba/envs/asapdiscovery/lib/python3.10/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain C is discontinuous at line 4491.\n",
      "  warnings.warn(\n",
      "DPI: 0.09, RFree: 0.21, Resolution: 1.59\n",
      "Processing BU # 1 with title: DNA REPAIR AND RECOMBINATION PROTEIN RADA, chains A\n",
      "Found unresolved N-terminal with 1 residues before ALA 108   A 1  , with sequence M\n",
      "Found 11 residue gap between VAL 285   A 1   and ALA 309   A 1  , with sequence QANGGHILAHS\n",
      "Found 6 residue gap between ASP 329   A 1   and GLY 336   A 1  , with sequence APHLPE\n",
      "Warning: There was a problem building some missing pieces, built as much as was possible\n",
      "   Falling back to charging protein with OEMMFF94Charges\n",
      "Processing BU # 2 with title: DNA REPAIR AND RECOMBINATION PROTEIN RADA, chains C\n",
      "Found unresolved N-terminal with 1 residues before ALA 108   C 2  , with sequence M\n",
      "Found 10 residue gap between VAL 285   C 2   and SER 308   C 2  , with sequence QANGGHILAH\n",
      "Found 5 residue gap between ASP 329   C 2   and GLU 335   C 2  , with sequence APHLP\n",
      "Warning: There was a problem building some missing pieces, built as much as was possible\n",
      "   Falling back to charging protein with OEMMFF94Charges\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Extracting ligand from 4B3D.pdb...\n",
      "Reading receptor from 4B3D.pdb...\n",
      "The complex file has been saved as 4B3D.pdb\n",
      "The receptor file has been saved as 4B3D_receptor.pdb\n",
      "The ligand file has been saved as 4B3D_ligand.pdb\n",
      "Preparing the ligand and protein files using OpenEye Toolkits...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "DPI: 0.09, RFree: 0.21, Resolution: 1.59\n",
      "Processing BU # 1 with title: DNA REPAIR AND RECOMBINATION PROTEIN RADA, chains A, alt: A\n",
      "Found unresolved N-terminal with 1 residues before ALA 108   A 1  , with sequence M\n",
      "Found 11 residue gap between VAL 285   A 1   and ALA 309   A 1  , with sequence QANGGHILAHS\n",
      "Found 6 residue gap between ASP 329   A 1   and GLY 336   A 1  , with sequence APHLPE\n",
      "Cap on GLY 336   A 1   clashed with solvent, removing clashing solvent\n",
      "  Removing HOH 2338   A 3  \n",
      "Cap on VAL 285   A 1   clashed with solvent, removing clashing solvent\n",
      "  Removing HOH 2308   A 3  \n",
      "Cap on ASP 329   A 1   clashed with solvent, removing clashing solvent\n",
      "  Removing HOH 2334   A 3  \n",
      "Warning: There was a problem building some missing pieces, built as much as was possible\n",
      "   Falling back to charging protein with OEMMFF94Charges\n",
      "Processing BU # 2 with title: DNA REPAIR AND RECOMBINATION PROTEIN RADA, chains A, alt: B\n",
      "Found unresolved N-terminal with 1 residues before ALA 108   A 1  , with sequence M\n",
      "Found 11 residue gap between VAL 285   A 1   and ALA 309   A 1  , with sequence QANGGHILAHS\n",
      "Found 6 residue gap between ASP 329   A 1   and GLY 336   A 1  , with sequence APHLPE\n",
      "Cap on GLY 336   A 1   clashed with solvent, removing clashing solvent\n",
      "  Removing HOH 2338   A 3  \n",
      "Cap on VAL 285   A 1   clashed with solvent, removing clashing solvent\n",
      "  Removing HOH 2308   A 3  \n",
      "Cap on ASP 329   A 1   clashed with solvent, removing clashing solvent\n",
      "  Removing HOH 2334   A 3  \n",
      "Warning: There was a problem building some missing pieces, built as much as was possible\n",
      "   Falling back to charging protein with OEMMFF94Charges\n",
      "Processing BU # 3 with title: DNA REPAIR AND RECOMBINATION PROTEIN RADA, chains C, alt: A\n",
      "Found unresolved N-terminal with 1 residues before ALA 108   C 2  , with sequence M\n",
      "Found 10 residue gap between VAL 285   C 2   and SER 308   C 2  , with sequence QANGGHILAH\n",
      "Found 5 residue gap between ASP 329   C 2   and GLU 335   C 2  , with sequence APHLP\n",
      "Cap on VAL 285   C 2   clashed with solvent, removing clashing solvent\n",
      "  Removing HOH 2275   C 3  \n",
      "Cap on ASP 329   C 2   clashed with solvent, removing clashing solvent\n",
      "  Removing HOH 2191   C 3  \n",
      "Warning: There was a problem building some missing pieces, built as much as was possible\n",
      "   Falling back to charging protein with OEMMFF94Charges\n",
      "Processing BU # 4 with title: DNA REPAIR AND RECOMBINATION PROTEIN RADA, chains C, alt: B\n",
      "Found unresolved N-terminal with 1 residues before ALA 108   C 2  , with sequence M\n",
      "Found 10 residue gap between VAL 285   C 2   and SER 308   C 2  , with sequence QANGGHILAH\n",
      "Found 5 residue gap between ASP 329   C 2   and GLU 335   C 2  , with sequence APHLP\n",
      "Cap on VAL 285   C 2   clashed with solvent, removing clashing solvent\n",
      "  Removing HOH 2275   C 3  \n",
      "Cap on ASP 329   C 2   clashed with solvent, removing clashing solvent\n",
      "  Removing HOH 2191   C 3  \n",
      "Warning: There was a problem building some missing pieces, built as much as was possible\n",
      "   Falling back to charging protein with OEMMFF94Charges\n",
      "Warning: Could not find alignment that matches constraints, try again without constraints\n",
      "Superposition - RMSD: 0.00, Ref: , Fit: , SeqScore: 996\n",
      "Superposition - RMSD: 0.04, Ref: A, Fit: A, SeqScore: 996\n",
      "Superposition - RMSD: 0.04, Ref: A, Fit: A, SeqScore: 996\n",
      "Superposition - RMSD: 0.34, Ref: A, Fit: C, SeqScore: 976\n",
      "Superposition - RMSD: 0.34, Ref: A, Fit: C, SeqScore: 976\n",
      "Superposition - RMSD: 0.34, Ref: A, Fit: C, SeqScore: 976\n",
      "Superposition - RMSD: 0.34, Ref: A, Fit: C, SeqScore: 976\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Design unit was successfully made for 4B3D.pdb, output is saved to ['4B3D_DU_0.oedu', '4B3D_DU_1.oedu', '4B3D_DU_2.oedu', '4B3D_DU_3.oedu', '4B3D_DU_4.oedu', '4B3D_DU_5.oedu', '4B3D_DU_6.oedu', '4B3D_DU_7.oedu'].\n",
      "The receptor design unit file has been saved as 4B3D_DU_0_receptor.oedu\n",
      "The receptor design unit file has been saved as 4B3D_DU_1_receptor.oedu\n",
      "The receptor design unit file has been saved as 4B3D_DU_2_receptor.oedu\n",
      "The receptor design unit file has been saved as 4B3D_DU_3_receptor.oedu\n",
      "The receptor design unit file has been saved as 4B3D_DU_4_receptor.oedu\n",
      "The receptor design unit file has been saved as 4B3D_DU_5_receptor.oedu\n",
      "The receptor design unit file has been saved as 4B3D_DU_6_receptor.oedu\n",
      "The receptor design unit file has been saved as 4B3D_DU_7_receptor.oedu\n",
      "cleaning 5MI(A-1351)\n",
      "docking 5MI(A-1351)\n",
      "cleaning 5MI(A-1351)\n",
      "docking 5MI(A-1351)\n",
      "cleaning 5MI(A-1351)\n",
      "docking 5MI(A-1351)\n",
      "cleaning 5MI(A-1351)\n",
      "docking 5MI(A-1351)\n",
      "cleaning 5MI(A-1351)\n",
      "docking 5MI(A-1351)\n",
      "cleaning 5MI(A-1351)\n",
      "docking 5MI(A-1351)\n",
      "cleaning 5MI(A-1351)\n",
      "docking 5MI(A-1351)\n",
      "cleaning 5MI(A-1351)\n",
      "docking 5MI(A-1351)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "mv: cannot move '4B3D_ligand_4B3D_receptor' to a subdirectory of itself, '/home/ian/msc_project/oe_docking_leakypdb_tmp/4B3D_ligand_4B3D_receptor/4B3D_ligand_4B3D_receptor'\n",
      "mv: cannot move '4B3D_ligand_4B3D_receptor' to a subdirectory of itself, '/home/ian/msc_project/oe_docking_leakypdb_tmp/4B3D_ligand_4B3D_receptor/4B3D_ligand_4B3D_receptor'\n",
      "mv: cannot move '4B3D_ligand_4B3D_receptor' to a subdirectory of itself, '/home/ian/msc_project/oe_docking_leakypdb_tmp/4B3D_ligand_4B3D_receptor/4B3D_ligand_4B3D_receptor'\n",
      " 62%|   | 5/8 [06:56<04:04, 81.50s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The docked ligand file has been saved as ['4B3D_DU_0_receptor_4B3D_ligand_docked.sdf', '4B3D_DU_1_receptor_4B3D_ligand_docked.sdf', '4B3D_DU_2_receptor_4B3D_ligand_docked.sdf', '4B3D_DU_3_receptor_4B3D_ligand_docked.sdf', '4B3D_DU_4_receptor_4B3D_ligand_docked.sdf', '4B3D_DU_5_receptor_4B3D_ligand_docked.sdf', '4B3D_DU_6_receptor_4B3D_ligand_docked.sdf', '4B3D_DU_7_receptor_4B3D_ligand_docked.sdf']\n",
      "Output saved as 4B3D_receptor_4B3D_ligand_0.out\n",
      "\n",
      "Data saved to OE_w_water_data.csv\n",
      "Docked ligand 4B3D_DU_0_receptor_4B3D_ligand_docked.sdf, molecule 1: Chemgauss4 score = -7.92\n",
      "Output saved as 4B3D_receptor_4B3D_ligand_0.out\n",
      "\n",
      "Data saved to OE_w_water_data.csv\n",
      "Docked ligand 4B3D_DU_1_receptor_4B3D_ligand_docked.sdf, molecule 1: Chemgauss4 score = -7.86\n",
      "Output saved as 4B3D_receptor_4B3D_ligand_0.out\n",
      "\n",
      "Data saved to OE_w_water_data.csv\n",
      "Docked ligand 4B3D_DU_2_receptor_4B3D_ligand_docked.sdf, molecule 1: Chemgauss4 score = -8.09\n",
      "Output saved as 4B3D_receptor_4B3D_ligand_0.out\n",
      "\n",
      "Data saved to OE_w_water_data.csv\n",
      "Docked ligand 4B3D_DU_3_receptor_4B3D_ligand_docked.sdf, molecule 1: Chemgauss4 score = -7.98\n",
      "Output saved as 4B3D_receptor_4B3D_ligand_0.out\n",
      "\n",
      "Data saved to OE_w_water_data.csv\n",
      "Docked ligand 4B3D_DU_4_receptor_4B3D_ligand_docked.sdf, molecule 1: Chemgauss4 score = -4.81\n",
      "Output saved as 4B3D_receptor_4B3D_ligand_0.out\n",
      "\n",
      "Data saved to OE_w_water_data.csv\n",
      "Docked ligand 4B3D_DU_5_receptor_4B3D_ligand_docked.sdf, molecule 1: Chemgauss4 score = -4.78\n",
      "Output saved as 4B3D_receptor_4B3D_ligand_0.out\n",
      "\n",
      "Data saved to OE_w_water_data.csv\n",
      "Docked ligand 4B3D_DU_6_receptor_4B3D_ligand_docked.sdf, molecule 1: Chemgauss4 score = -4.03\n",
      "Output saved as 4B3D_receptor_4B3D_ligand_0.out\n",
      "\n",
      "Data saved to OE_w_water_data.csv\n",
      "Docked ligand 4B3D_DU_7_receptor_4B3D_ligand_docked.sdf, molecule 1: Chemgauss4 score = -4.25\n",
      "Data saved in ./data//home/ian/msc_project/oe_docking_leakypdb_tmp/4B3D_ligand_4B3D_receptor\n",
      "Downloading PDB structure '3gvu'...\n",
      "Extracting ligand from 3GVU.pdb...\n",
      "Reading receptor from 3GVU.pdb...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "DPI: 0.15, RFree: 0.23, Resolution: 2.05\n",
      "Processing BU # 1 with title: TYROSINE-PROTEIN KINASE ABL2, chains A\n",
      "Warning: None of the residue conformations passed a clash test\n",
      "Warning: For residue MET 278   A 1   removing clashing solvent molecule HOH 203   A 2  \n",
      "Warning: For residue LYS 402   A 1   removing clashing solvent molecule HOH 231   A 2  \n",
      "Warning: For residue LYS 402   A 1   removing clashing solvent molecule HOH 136   A 2  \n",
      "Warning: For residue SER 546   A 1   removing clashing solvent molecule HOH 549   A 2  \n",
      "Warning: Sequence alignment created an incorrect match for chain A, skipping gap detection. Try again with adjusted gap and extend penalties.\n",
      "Warning: There was a problem building some missing pieces, built as much as was possible\n",
      "   Falling back to charging protein with OEMMFF94Charges\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The complex file has been saved as 3GVU.pdb\n",
      "The receptor file has been saved as 3GVU_receptor.pdb\n",
      "The ligand file has been saved as 3GVU_ligand.pdb\n",
      "Preparing the ligand and protein files using OpenEye Toolkits...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "DPI: 0.15, RFree: 0.23, Resolution: 2.05\n",
      "Processing BU # 1 with title: TYROSINE-PROTEIN KINASE ABL2, chains A, alt: A\n",
      "Warning: None of the residue conformations passed a clash test\n",
      "Warning: For residue MET 278   A 1   removing clashing solvent molecule HOH 203   A 2  \n",
      "Warning: For residue LYS 402   A 1   removing clashing solvent molecule HOH 231   A 2  \n",
      "Warning: For residue LYS 402   A 1   removing clashing solvent molecule HOH 136   A 2  \n",
      "Warning: For residue SER 546   A 1   removing clashing solvent molecule HOH 549   A 2  \n",
      "Warning: Sequence alignment created an incorrect match for chain A, skipping gap detection. Try again with adjusted gap and extend penalties.\n",
      "Warning: There was a problem building some missing pieces, built as much as was possible\n",
      "Warning: AddExplicitHydrogensForProteins: H   SER277(A) has problem with hydrogen count or geometry. Please fix if necessary.\n",
      "Warning: AddExplicitHydrogensForProteins: H   ASP545(A) has problem with hydrogen count or geometry. Please fix if necessary.\n",
      "Warning: AddExplicitHydrogensForProteins: H   SER277(A) has problem with hydrogen count or geometry. Please fix if necessary.\n",
      "Warning: AddExplicitHydrogensForProteins: H   ASP545(A) has problem with hydrogen count or geometry. Please fix if necessary.\n",
      "   Falling back to charging protein with OEMMFF94Charges\n",
      "Warning: Failed to charge protein\n",
      "Warning: Unable to add partial charges and radii to DesignUnit: TYROSINE-PROTEIN KINASE ABL2(A)altA__DU__biounit\n",
      "Processing BU # 2 with title: TYROSINE-PROTEIN KINASE ABL2, chains A, alt: B\n",
      "Warning: None of the residue conformations passed a clash test\n",
      "Warning: For residue LYS 402   A 1   removing clashing solvent molecule HOH 231   A 2  \n",
      "Warning: For residue LYS 402   A 1   removing clashing solvent molecule HOH 136   A 2  \n",
      "Warning: For residue SER 546   A 1   removing clashing solvent molecule HOH 549   A 2  \n",
      "Warning: Sequence alignment created an incorrect match for chain A, skipping gap detection. Try again with adjusted gap and extend penalties.\n",
      "Warning: There was a problem building some missing pieces, built as much as was possible\n",
      "Warning: AddExplicitHydrogensForProteins: H   ASP545(A) has problem with hydrogen count or geometry. Please fix if necessary.\n",
      "Warning: AddExplicitHydrogensForProteins: H   ASP545(A) has problem with hydrogen count or geometry. Please fix if necessary.\n",
      "   Falling back to charging protein with OEMMFF94Charges\n",
      "Warning: Failed to charge protein\n",
      "Warning: Unable to add partial charges and radii to DesignUnit: TYROSINE-PROTEIN KINASE ABL2(A)altB__DU__biounit\n",
      "Skipping redundant DU with alts outside the site of interest, renaming existing to collapse alts\n",
      "Discarding redundant alt DU with title TYROSINE-PROTEIN KINASE ABL2(A)altB > STI(A-1001)\n",
      "Skipping redundant DU with alts outside the site of interest, renaming existing to collapse alts\n",
      "Discarding redundant alt DU with title TYROSINE-PROTEIN KINASE ABL2(A)altB > STI(A-1002)\n",
      "Warning: Could not find alignment that matches constraints, try again without constraints\n",
      "Superposition - RMSD: 0.00, Ref: , Fit: , SeqScore: 1446\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Design unit was successfully made for 3GVU.pdb, output is saved to ['3GVU_DU_0.oedu', '3GVU_DU_1.oedu'].\n",
      "The receptor design unit file has been saved as 3GVU_DU_0_receptor.oedu\n",
      "The receptor design unit file has been saved as 3GVU_DU_1_receptor.oedu\n",
      "cleaning STI(A-1001)\n",
      "docking STI(A-1001)\n",
      "cleaning STI(A-1001)\n",
      "docking STI(A-1001)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "mv: cannot move '3GVU_ligand_3GVU_receptor' to a subdirectory of itself, '/home/ian/msc_project/oe_docking_leakypdb_tmp/3GVU_ligand_3GVU_receptor/3GVU_ligand_3GVU_receptor'\n",
      "mv: cannot move '3GVU_ligand_3GVU_receptor' to a subdirectory of itself, '/home/ian/msc_project/oe_docking_leakypdb_tmp/3GVU_ligand_3GVU_receptor/3GVU_ligand_3GVU_receptor'\n",
      "mv: cannot move '3GVU_ligand_3GVU_receptor' to a subdirectory of itself, '/home/ian/msc_project/oe_docking_leakypdb_tmp/3GVU_ligand_3GVU_receptor/3GVU_ligand_3GVU_receptor'\n",
      " 75%|  | 6/8 [09:13<03:21, 100.55s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The docked ligand file has been saved as ['3GVU_DU_0_receptor_3GVU_ligand_docked.sdf', '3GVU_DU_1_receptor_3GVU_ligand_docked.sdf']\n",
      "Output saved as 3GVU_receptor_3GVU_ligand_0.out\n",
      "\n",
      "Data saved to OE_w_water_data.csv\n",
      "Docked ligand 3GVU_DU_0_receptor_3GVU_ligand_docked.sdf, molecule 1: Chemgauss4 score = -25.99\n",
      "Output saved as 3GVU_receptor_3GVU_ligand_0.out\n",
      "\n",
      "Data saved to OE_w_water_data.csv\n",
      "Docked ligand 3GVU_DU_1_receptor_3GVU_ligand_docked.sdf, molecule 1: Chemgauss4 score = -5.08\n",
      "Data saved in ./data//home/ian/msc_project/oe_docking_leakypdb_tmp/3GVU_ligand_3GVU_receptor\n",
      "Downloading PDB structure '6c7q'...\n",
      "Extracting ligand from 6C7Q.pdb...\n",
      "Reading receptor from 6C7Q.pdb...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "DPI: 0.08, RFree: 0.23, Resolution: 1.51\n",
      "Processing BU # 1 with title: BROMODOMAIN-CONTAINING PROTEIN 4, chains A\n",
      "Warning: For residue GLU 383   A 1   removing clashing solvent molecule HOH 686   A 2  \n",
      "Found unresolved N-terminal with 19 residues before LYS 349   A 1  , with sequence SNAKDVPDSQQHPAPEKSS\n",
      "Warning: There was a problem building some missing pieces, built as much as was possible\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The complex file has been saved as 6C7Q.pdb\n",
      "The receptor file has been saved as 6C7Q_receptor.pdb\n",
      "The ligand file has been saved as 6C7Q_ligand.pdb\n",
      "Preparing the ligand and protein files using OpenEye Toolkits...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "DPI: 0.08, RFree: 0.23, Resolution: 1.51\n",
      "Processing BU # 1 with title: BROMODOMAIN-CONTAINING PROTEIN 4, chains A, alt: A\n",
      "Warning: For residue GLU 383   A 1   removing clashing solvent molecule HOH 686   A 2  \n",
      "Found unresolved N-terminal with 19 residues before LYS 349   A 1  , with sequence SNAKDVPDSQQHPAPEKSS\n",
      "Warning: There was a problem building some missing pieces, built as much as was possible\n",
      "Processing BU # 2 with title: BROMODOMAIN-CONTAINING PROTEIN 4, chains A, alt: B\n",
      "Warning: For residue GLU 383   A 1   removing clashing solvent molecule HOH 686   A 2  \n",
      "Found unresolved N-terminal with 19 residues before LYS 349   A 1  , with sequence SNAKDVPDSQQHPAPEKSS\n",
      "Warning: There was a problem building some missing pieces, built as much as was possible\n",
      "Superposition - RMSD: 0.00, Ref: A, Fit: A, SeqScore: 575\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Design unit was successfully made for 6C7Q.pdb, output is saved to ['6C7Q_DU_0.oedu', '6C7Q_DU_1.oedu'].\n",
      "The receptor design unit file has been saved as 6C7Q_DU_0_receptor.oedu\n",
      "The receptor design unit file has been saved as 6C7Q_DU_1_receptor.oedu\n",
      "cleaning EO1(A-501)\n",
      "docking EO1(A-501)\n",
      "cleaning EO1(A-501)\n",
      "docking EO1(A-501)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "mv: cannot move '6C7Q_ligand_6C7Q_receptor' to a subdirectory of itself, '/home/ian/msc_project/oe_docking_leakypdb_tmp/6C7Q_ligand_6C7Q_receptor/6C7Q_ligand_6C7Q_receptor'\n",
      "mv: cannot move '6C7Q_ligand_6C7Q_receptor' to a subdirectory of itself, '/home/ian/msc_project/oe_docking_leakypdb_tmp/6C7Q_ligand_6C7Q_receptor/6C7Q_ligand_6C7Q_receptor'\n",
      "mv: cannot move '6C7Q_ligand_6C7Q_receptor' to a subdirectory of itself, '/home/ian/msc_project/oe_docking_leakypdb_tmp/6C7Q_ligand_6C7Q_receptor/6C7Q_ligand_6C7Q_receptor'\n",
      " 88%| | 7/8 [11:20<01:49, 109.12s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The docked ligand file has been saved as ['6C7Q_DU_0_receptor_6C7Q_ligand_docked.sdf', '6C7Q_DU_1_receptor_6C7Q_ligand_docked.sdf']\n",
      "Output saved as 6C7Q_receptor_6C7Q_ligand_0.out\n",
      "\n",
      "Data saved to OE_w_water_data.csv\n",
      "Docked ligand 6C7Q_DU_0_receptor_6C7Q_ligand_docked.sdf, molecule 1: Chemgauss4 score = -9.55\n",
      "Output saved as 6C7Q_receptor_6C7Q_ligand_0.out\n",
      "\n",
      "Data saved to OE_w_water_data.csv\n",
      "Docked ligand 6C7Q_DU_1_receptor_6C7Q_ligand_docked.sdf, molecule 1: Chemgauss4 score = -9.65\n",
      "Data saved in ./data//home/ian/msc_project/oe_docking_leakypdb_tmp/6C7Q_ligand_6C7Q_receptor\n",
      "Downloading PDB structure '4o61'...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ian/micromamba/envs/asapdiscovery/lib/python3.10/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 3888.\n",
      "  warnings.warn(\n",
      "/home/ian/micromamba/envs/asapdiscovery/lib/python3.10/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'U' for Atom (name=UNK) with given element 'X'\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/ian/micromamba/envs/asapdiscovery/lib/python3.10/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 3926.\n",
      "  warnings.warn(\n",
      "/home/ian/micromamba/envs/asapdiscovery/lib/python3.10/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 3964.\n",
      "  warnings.warn(\n",
      "/home/ian/micromamba/envs/asapdiscovery/lib/python3.10/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 4093.\n",
      "  warnings.warn(\n",
      "DPI: 0.14, RFree: 0.23, Resolution: 1.90\n",
      "Processing BU # 1 with title: RNA DEMETHYLASE ALKBH5, chains A\n",
      "Warning: Incomplete backbone in THR 294   A 1  , removing residue\n",
      "Warning: For residue ARG 83   A 1   removing clashing solvent molecule HOH 506   A 3  \n",
      "Warning: For residue ARG 93   A 1   removing clashing solvent molecule HOH 489   A 3  \n",
      "Warning: For residue GLN 97   A 1   removing clashing solvent molecule HOH 488   A 3  \n",
      "Warning: For residue LYS 231   A 1   removing clashing solvent molecule HOH 513   A 3  \n",
      "Found unresolved N-terminal with 3 residues before LEU 76   A 1  , with sequence GQQ\n",
      "Found 7 residue gap between TYR 141   A 1   and GLY 149   A 1  , with sequence GAQLQKR\n",
      "Found unresolved C-terminal with 1 residues after GLU 293   A 1  , with sequence T\n",
      "Warning: There was a problem building some missing pieces, built as much as was possible\n",
      "   Falling back to charging protein with OEMMFF94Charges\n",
      "Warning: Failed to charge protein\n",
      "Warning: Failed to charge miscellaneous components\n",
      "Warning: Unable to add partial charg"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Extracting ligand from 4O61.pdb...\n",
      "Reading receptor from 4O61.pdb...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "es and radii to DesignUnit: RNA DEMETHYLASE ALKBH5(A)__DU__biounit\n",
      "Processing BU # 2 with title: RNA DEMETHYLASE ALKBH5, chains B\n",
      "Warning: For residue LYS 84   B 2   removing clashing solvent molecule HOH 489 A B 3  \n",
      "Warning: For residue ARG 93   B 2   removing clashing solvent molecule HOH 459   B 3  \n",
      "Warning: For residue LYS 231   B 2   removing clashing solvent molecule HOH 421   B 3  \n",
      "Warning: For residue LYS 231   B 2   removing clashing solvent molecule HOH 537   B 3  \n",
      "Found unresolved N-terminal with 4 residues before GLN 77   B 2  , with sequence GQQL\n",
      "Found 8 residue gap between THR 140   B 2   and GLY 149   B 2  , with sequence YGAQLQKR\n",
      "Found unresolved C-terminal with 1 residues after GLU 293   B 2  , with sequence T\n",
      "Warning: There was a problem building some missing pieces, built as much as was possible\n",
      "   Falling back to charging protein with OEMMFF94Charges\n",
      "Warning: Failed to charge miscellaneous components\n",
      "Warning: Unable to add partial charges and radii to DesignUnit: RNA DEMETHYLASE ALKBH5(B)__DU__biounit\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The complex file has been saved as 4O61.pdb\n",
      "The receptor file has been saved as 4O61_receptor.pdb\n",
      "The ligand file has been saved as 4O61_ligand.pdb\n",
      "Preparing the ligand and protein files using OpenEye Toolkits...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "DPI: 0.14, RFree: 0.23, Resolution: 1.90\n",
      "Processing BU # 1 with title: RNA DEMETHYLASE ALKBH5, chains A, alt: A\n",
      "Warning: Incomplete backbone in THR 294   A 1  , removing residue\n",
      "Warning: For residue ARG 83   A 1   removing clashing solvent molecule HOH 506   A 3  \n",
      "Warning: For residue ARG 93   A 1   removing clashing solvent molecule HOH 489   A 3  \n",
      "Warning: For residue GLN 97   A 1   removing clashing solvent molecule HOH 488   A 3  \n",
      "Warning: For residue LYS 231   A 1   removing clashing solvent molecule HOH 513   A 3  \n",
      "Found unresolved N-terminal with 3 residues before LEU 76   A 1  , with sequence GQQ\n",
      "Found 7 residue gap between TYR 141   A 1   and GLY 149   A 1  , with sequence GAQLQKR\n",
      "Found unresolved C-terminal with 1 residues after GLU 293   A 1  , with sequence T\n",
      "Warning: There was a problem building some missing pieces, built as much as was possible\n",
      "   Falling back to charging protein with OEMMFF94Charges\n",
      "Warning: Failed to charge miscellaneous components\n",
      "Warning: Unable to add partial charges and radii to DesignUnit: RNA DEMETHYLASE ALKBH5(A)altA__DU__biounit\n",
      "Processing BU # 2 with title: RNA DEMETHYLASE ALKBH5, chains A, alt: B\n",
      "Warning: Incomplete backbone in THR 294   A 1  , removing residue\n",
      "Warning: For residue ARG 83   A 1   removing clashing solvent molecule HOH 506   A 3  \n",
      "Warning: For residue ARG 93   A 1   removing clashing solvent molecule HOH 489   A 3  \n",
      "Warning: For residue GLN 97   A 1   removing clashing solvent molecule HOH 488   A 3  \n",
      "Warning: For residue LYS 231   A 1   removing clashing solvent molecule HOH 513   A 3  \n",
      "Found unresolved N-terminal with 3 residues before LEU 76   A 1  , with sequence GQQ\n",
      "Found 7 residue gap between TYR 141   A 1   and GLY 149   A 1  , with sequence GAQLQKR\n",
      "Found unresolved C-terminal with 1 residues after GLU 293   A 1  , with sequence T\n",
      "Warning: There was a problem building some missing pieces, built as much as was possible\n",
      "   Falling back to charging protein with OEMMFF94Charges\n",
      "Warning: Failed to charge miscellaneous components\n",
      "Warning: Unable to add partial charges and radii to DesignUnit: RNA DEMETHYLASE ALKBH5(A)altB__DU__biounit\n",
      "Processing BU # 3 with title: RNA DEMETHYLASE ALKBH5, chains B, alt: A\n",
      "Warning: For residue LYS 84   B 2   removing clashing solvent molecule HOH 489 A B 3  \n",
      "Warning: For residue ARG 93   B 2   removing clashing solvent molecule HOH 459   B 3  \n",
      "Warning: For residue LYS 231   B 2   removing clashing solvent molecule HOH 421   B 3  \n",
      "Warning: For residue LYS 231   B 2   removing clashing solvent molecule HOH 537   B 3  \n",
      "Found unresolved N-terminal with 4 residues before GLN 77   B 2  , with sequence GQQL\n",
      "Found 8 residue gap between THR 140   B 2   and GLY 149   B 2  , with sequence YGAQLQKR\n",
      "Found unresolved C-terminal with 1 residues after GLU 293   B 2  , with sequence T\n",
      "Warning: There was a problem building some missing pieces, built as much as was possible\n",
      "   Falling back to charging protein with OEMMFF94Charges\n",
      "Warning: Failed to charge miscellaneous components\n",
      "Warning: Unable to add partial charges and radii to DesignUnit: RNA DEMETHYLASE ALKBH5(B)altA__DU__biounit\n",
      "Processing BU # 4 with title: RNA DEMETHYLASE ALKBH5, chains B, alt: B\n",
      "Warning: For residue LYS 84   B 2   removing clashing solvent molecule HOH 489 A B 3  \n",
      "Warning: For residue ARG 93   B 2   removing clashing solvent molecule HOH 459   B 3  \n",
      "Warning: For residue LYS 102   B 2   removing clashing solvent molecule HOH 487   B 3  \n",
      "Warning: For residue LYS 231   B 2   removing clashing solvent molecule HOH 421   B 3  \n",
      "Warning: For residue LYS 231   B 2   removing clashing solvent molecule HOH 537   B 3  \n",
      "Found unresolved N-terminal with 4 residues before GLN 77   B 2  , with sequence GQQL\n",
      "Found 8 residue gap between THR 140   B 2   and GLY 149   B 2  , with sequence YGAQLQKR\n",
      "Found unresolved C-terminal with 1 residues after GLU 293   B 2  , with sequence T\n",
      "Warning: There was a problem building some missing pieces, built as much as was possible\n",
      "   Falling back to charging protein with OEMMFF94Charges\n",
      "Warning: Failed to charge miscellaneous components\n",
      "Warning: Unable to add partial charges and radii to DesignUnit: RNA DEMETHYLASE ALKBH5(B)altB__DU__biounit\n",
      "Skipping redundant DU with alts outside the site of interest, renaming existing to collapse alts\n",
      "Discarding redundant alt DU with title RNA DEMETHYLASE ALKBH5(A)altB > UNL(A-303)\n",
      "Skipping redundant DU with alts outside the site of interest, renaming existing to collapse alts\n",
      "Discarding redundant alt DU with title RNA DEMETHYLASE ALKBH5(B)altB > UNL(B-303)\n",
      "Superposition - RMSD: 0.24, Ref: A, Fit: B, SeqScore: 1049\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Design unit was successfully made for 4O61.pdb, output is saved to ['4O61_DU_0.oedu', '4O61_DU_1.oedu'].\n",
      "The receptor design unit file has been saved as 4O61_DU_0_receptor.oedu\n",
      "The receptor design unit file has been saved as 4O61_DU_1_receptor.oedu\n",
      "cleaning UNL(A-303)\n",
      "Molecule is fragmented. Skipping.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "mv: cannot move '4O61_ligand_4O61_receptor' to a subdirectory of itself, '/home/ian/msc_project/oe_docking_leakypdb_tmp/4O61_ligand_4O61_receptor/4O61_ligand_4O61_receptor'\n",
      "mv: cannot move '4O61_ligand_4O61_receptor' to a subdirectory of itself, '/home/ian/msc_project/oe_docking_leakypdb_tmp/4O61_ligand_4O61_receptor/4O61_ligand_4O61_receptor'\n",
      "mv: cannot move '4O61_ligand_4O61_receptor' to a subdirectory of itself, '/home/ian/msc_project/oe_docking_leakypdb_tmp/4O61_ligand_4O61_receptor/4O61_ligand_4O61_receptor'\n",
      "100%|| 8/8 [13:14<00:00, 99.35s/it] "
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cleaning UNL(A-303)\n",
      "Molecule is fragmented. Skipping.\n",
      "The docked ligand file has been saved as ['4O61_DU_0_receptor_4O61_ligand_docked.sdf', '4O61_DU_1_receptor_4O61_ligand_docked.sdf']\n",
      "Error in 4O61: File error: Invalid input file 4O61_DU_0_receptor_4O61_ligand_docked.sdf\n",
      "4O61: No valid energy value available.\n",
      "Skipping 4O61_ligand.pdb due to an error: File error: Invalid input file 4O61_DU_0_receptor_4O61_ligand_docked.sdf\n",
      "Error message saved to OE_w_water_data.csv\n",
      "Data saved in ./data//home/ian/msc_project/oe_docking_leakypdb_tmp/4O61_ligand_4O61_receptor\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "pdb_ids = ['6R8O',\n",
    "           '5S9L',\n",
    "           '1KLT',\n",
    "           '1A42',\n",
    "           '4B3D', # Failed in vina\n",
    "           '3GVU',\n",
    "           '6C7Q',\n",
    "           '4O61', # Failed ADFR    \n",
    "           ]\n",
    "for pdb_id in tqdm(pdb_ids):\n",
    "    # download \n",
    "    filename = download_pdb_file(pdb_id)\n",
    "    # split the protein and ligand from the PDB file\n",
    "    complex_file, receptor_file, ligand_file = pdb_to_prot_lig(pdb_id, filename, preserve_water=True,preserve_metal=True)\n",
    "    # rest of pipeline\n",
    "    oe_process_lig_prot(ligand_file, receptor_file, complex_file, preserve_water=True, preserve_metal=True, csv_out_file='OE_w_water_data.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<font color='yellow'>Testing the pipeline using Mpro data</font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ian/micromamba/envs/asapdiscovery/lib/python3.10/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 2838.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fixed receptor file saved as: Mpro-x0072_0_receptor_fixed.pdb\n",
      "Extracting ligand from Mpro-x0072_0.pdb...\n",
      "Reading receptor from Mpro-x0072_0.pdb...\n",
      "Failed to extract ligand from Mpro-x0072_0.pdb using the OpenEye Toolkit,trying a different script written using OESplitMolComplex...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "DPI: 0.12, RFree: 0.23, Resolution: 1.65\n",
      "Processing BU # 1 with title: ---, chains AB\n",
      "Warning: OEPlaceHydrogens: mol DesignUnit Components has heavy atoms ' O   HOH 231 A B 1  ' and ' O   HOH 231 B B 1  ' too close 0.11 < 0.5A\n",
      "Warning: Unable to add protons to DesignUnit: ---(AB)__DU__biounit, failing with strict mode enabled\n",
      "Warning: No ligand or sites detected, please revise inputs and/or options\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The receptor file has been saved as Mpro-x0072_0_receptor_fixed.pdb\n",
      "The ligand file has been saved as Mpro-x0072_0_ligand.pdb\n",
      "Preparing the ligand and protein files using OpenEye Toolkits...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "DPI: 0.12, RFree: 0.23, Resolution: 1.65\n",
      "Processing BU # 1 with title: ---, chains AB, alt: A\n",
      "Warning: OEPlaceHydrogens: mol DesignUnit Components has heavy atoms ' O   HOH 231 A B 1  ' and ' O   HOH 231 B B 1  ' too close 0.11 < 0.5A\n",
      "Warning: Unable to add protons to DesignUnit: ---(AB)altA__DU__biounit\n",
      "Warning: Due to lack of protons, unable to add partial charges and radii to DesignUnit: ---(AB)altA__DU__biounit\n",
      "Processing BU # 2 with title: ---, chains AB, alt: B\n",
      "Warning: OEPlaceHydrogens: mol DesignUnit Components has heavy atoms ' O   HOH 231 A B 1  ' and ' O   HOH 231 B B 1  ' too close 0.11 < 0.5A\n",
      "Warning: Unable to add protons to DesignUnit: ---(AB)altB__DU__biounit\n",
      "Warning: Due to lack of protons, unable to add partial charges and radii to DesignUnit: ---(AB)altB__DU__biounit\n",
      "Warning: Hydrogens have not been placed, can not add OEInteractionHints\n",
      "Warning: Unable to add interactions to Design Unit: ---(AB)altA > LIG(A-1101)\n",
      "Warning: Hydrogens have not been placed, can not add OEInteractionHints\n",
      "Warning: Unable to add interactions to Design Unit: ---(AB)altA > LIG(B-1101)\n",
      "Warning: Hydrogens have not been placed, can not add OEInteractionHints\n",
      "Warning: Unable to add interactions to Design Unit: ---(AB)altB > LIG(A-1101)\n",
      "Warning: Hydrogens have not been placed, can not add OEInteractionHints\n",
      "Warning: Unable to add interactions to Design Unit: ---(AB)altB > LIG(B-1101)\n",
      "Skipping redundant DU with alts outside the site of interest, renaming existing to collapse alts\n",
      "Discarding redundant alt DU with title ---(AB)altB > LIG(A-1101)\n",
      "Skipping redundant DU with alts outside the site of interest, renaming existing to collapse alts\n",
      "Discarding redundant alt DU with title ---(AB)altB > LIG(B-1101)\n",
      "Warning: Could not find alignment that matches constraints, try again without constraints\n",
      "Superposition - RMSD: 0.00, Ref: , Fit: , SeqScore: 3102\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Design unit was successfully made for Mpro-x0072_0.pdb, output is saved to ['Mpro-x0072_0_DU_0.oedu', 'Mpro-x0072_0_DU_1.oedu'].\n",
      "The receptor design unit file has been saved as Mpro-x0072_0_DU_0_receptor.oedu\n",
      "The receptor design unit file has been saved as Mpro-x0072_0_DU_1_receptor.oedu\n",
      "cleaning LIG\n",
      "docking LIG\n",
      "cleaning LIG\n",
      "docking LIG\n",
      "The docked ligand file has been saved as ['Mpro-x0072_0_DU_0_ligand_docked.sdf', 'Mpro-x0072_0_DU_1_ligand_docked.sdf']\n",
      "Output saved as Mpro-x0072_0_receptor_fixed_Mpro-x0072_0_ligand_0.out\n",
      "\n",
      "Data saved to oe_docking_data_playground.csv\n",
      "Docked ligand Mpro-x0072_0_DU_0_ligand_docked.sdf, molecule 1: Chemgauss4 score = -6.91\n",
      "Output saved as Mpro-x0072_0_receptor_fixed_Mpro-x0072_0_ligand_0.out\n",
      "\n",
      "Data saved to oe_docking_data_playground.csv\n",
      "Docked ligand Mpro-x0072_0_DU_1_ligand_docked.sdf, molecule 1: Chemgauss4 score = -6.92\n",
      "Data saved in data/Mpro-x0072_0_ligand_with_Mpro-x0072_0_receptor_fixed\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "cp: -r not specified; omitting directory 'Mpro-x0072_0_receptor_fixed'\n",
      "mv: cannot move '/home/ian/msc_project/docking/6R8O' to 'data/Mpro-x0072_0_ligand_with_Mpro-x0072_0_receptor_fixed/6R8O': Directory not empty\n"
     ]
    }
   ],
   "source": [
    "prot_file, lig_file = pdb_to_prot_lig(pdb_id='Mpro-x0072_0', filename='Mpro-x0072_0.pdb')\n",
    "complex_file='Mpro-x0072_0.pdb'\n",
    "oe_process_lig_prot(lig_file, prot_file, complex_file, csv_out_file='oe_docking_data_playground.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<font color='yellow'>Trying out with diffusion model generated molecules</font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ian/micromamba/envs/asapdiscovery/lib/python3.10/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 2838.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fixed receptor file saved as: Mpro-x0072_0_receptor_fixed.pdb\n",
      "Extracting ligand from Mpro-x0072_0.pdb...\n",
      "Reading receptor from Mpro-x0072_0.pdb...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "DPI: 0.12, RFree: 0.23, Resolution: 1.65\n",
      "Processing BU # 1 with title: ---, chains AB\n",
      "Warning: OEPlaceHydrogens: mol DesignUnit Components has heavy atoms ' O   HOH 231 A B 1  ' and ' O   HOH 231 B B 1  ' too close 0.11 < 0.5A\n",
      "Warning: Unable to add protons to DesignUnit: ---(AB)__DU__biounit, failing with strict mode enabled\n",
      "Warning: No ligand or sites detected, please revise inputs and/or options\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Failed to extract ligand from Mpro-x0072_0.pdb using the OpenEye Toolkit,trying a different script written using OESplitMolComplex...\n",
      "The receptor file has been saved as Mpro-x0072_0_receptor_fixed.pdb\n",
      "The ligand file has been saved as Mpro-x0072_0_ligand.pdb\n",
      "Preparing the ligand and protein files using OpenEye Toolkits...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "DPI: 0.12, RFree: 0.23, Resolution: 1.65\n",
      "Processing BU # 1 with title: ---, chains AB, alt: A\n",
      "Warning: OEPlaceHydrogens: mol DesignUnit Components has heavy atoms ' O   HOH 231 A B 1  ' and ' O   HOH 231 B B 1  ' too close 0.11 < 0.5A\n",
      "Warning: Unable to add protons to DesignUnit: ---(AB)altA__DU__biounit\n",
      "Warning: Due to lack of protons, unable to add partial charges and radii to DesignUnit: ---(AB)altA__DU__biounit\n",
      "Processing BU # 2 with title: ---, chains AB, alt: B\n",
      "Warning: OEPlaceHydrogens: mol DesignUnit Components has heavy atoms ' O   HOH 231 A B 1  ' and ' O   HOH 231 B B 1  ' too close 0.11 < 0.5A\n",
      "Warning: Unable to add protons to DesignUnit: ---(AB)altB__DU__biounit\n",
      "Warning: Due to lack of protons, unable to add partial charges and radii to DesignUnit: ---(AB)altB__DU__biounit\n",
      "Warning: Hydrogens have not been placed, can not add OEInteractionHints\n",
      "Warning: Unable to add interactions to Design Unit: ---(AB)altA > LIG(A-1101)\n",
      "Warning: Hydrogens have not been placed, can not add OEInteractionHints\n",
      "Warning: Unable to add interactions to Design Unit: ---(AB)altA > LIG(B-1101)\n",
      "Warning: Hydrogens have not been placed, can not add OEInteractionHints\n",
      "Warning: Unable to add interactions to Design Unit: ---(AB)altB > LIG(A-1101)\n",
      "Warning: Hydrogens have not been placed, can not add OEInteractionHints\n",
      "Warning: Unable to add interactions to Design Unit: ---(AB)altB > LIG(B-1101)\n",
      "Skipping redundant DU with alts outside the site of interest, renaming existing to collapse alts\n",
      "Discarding redundant alt DU with title ---(AB)altB > LIG(A-1101)\n",
      "Skipping redundant DU with alts outside the site of interest, renaming existing to collapse alts\n",
      "Discarding redundant alt DU with title ---(AB)altB > LIG(B-1101)\n",
      "Warning: Could not find alignment that matches constraints, try again without constraints\n",
      "Superposition - RMSD: 0.00, Ref: , Fit: , SeqScore: 3102\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Design unit was successfully made for Mpro-x0072_0.pdb, output is saved to ['Mpro-x0072_0_DU_0.oedu', 'Mpro-x0072_0_DU_1.oedu'].\n",
      "The receptor design unit file has been saved as Mpro-x0072_0_DU_0_receptor.oedu\n",
      "The receptor design unit file has been saved as Mpro-x0072_0_DU_1_receptor.oedu\n",
      "Skipping molecules_bx_2024_06_18_154410.sdf due to an error: expected str, bytes or os.PathLike object, not NoneType\n",
      "Error message saved to oe_docking_data_playground.csv\n",
      "Data saved in ./data//home/ian/msc_project/docking/molecules_bx_2024_06_18_154410_Mpro-x0072_0_receptor_fixed\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "mv: cannot move 'molecules_bx_2024_06_18_154410_Mpro-x0072_0_receptor_fixed' to a subdirectory of itself, '/home/ian/msc_project/docking/molecules_bx_2024_06_18_154410_Mpro-x0072_0_receptor_fixed/molecules_bx_2024_06_18_154410_Mpro-x0072_0_receptor_fixed'\n",
      "mv: cannot move 'molecules_bx_2024_06_18_154410_Mpro-x0072_0_receptor_fixed' to a subdirectory of itself, '/home/ian/msc_project/docking/molecules_bx_2024_06_18_154410_Mpro-x0072_0_receptor_fixed/molecules_bx_2024_06_18_154410_Mpro-x0072_0_receptor_fixed'\n",
      "mv: cannot move 'molecules_bx_2024_06_18_154410_Mpro-x0072_0_receptor_fixed' to a subdirectory of itself, '/home/ian/msc_project/docking/molecules_bx_2024_06_18_154410_Mpro-x0072_0_receptor_fixed/molecules_bx_2024_06_18_154410_Mpro-x0072_0_receptor_fixed'\n",
      "mv: cannot move '/home/ian/msc_project/docking/molecules_bx_2024_06_18_154410_Mpro-x0072_0_receptor_fixed' to './data/molecules_bx_2024_06_18_154410_Mpro-x0072_0_receptor_fixed': Directory not empty\n"
     ]
    }
   ],
   "source": [
    "os.system('cp 2024_06_18_154410/molecules_bx_2024_06_18_154410.sdf ./')\n",
    "os.system('cp Mpro_complexes/Mpro-x0072_0.pdb ./')\n",
    "# split the protein and ligand from the PDB file\n",
    "receptor_file, ligand_file = pdb_to_prot_lig(pdb_id ='Mpro-x0072_0', filename = 'Mpro-x0072_0.pdb')\n",
    "#--------\n",
    "lig_file='molecules_bx_2024_06_18_154410.sdf' # this contains 14 molecules\n",
    "prot_file='Mpro-x0072_0_receptor_fixed.pdb' # replace with a protein file of interest\n",
    "complex_file='Mpro-x0072_0.pdb' # replace with a reference protein-ligand complex\n",
    "oe_process_lig_prot(lig_file, prot_file, complex_file, csv_out_file='oe_docking_data_playground.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<font color='yellow'>Running the pipeline for lists</font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|          | 0/2 [00:00<?, ?it/s]DPI: 0.12, RFree: 0.23, Resolution: 1.65\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Preparing the ligand and protein files using OpenEye Toolkits...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing BU # 1 with title: ---, chains AB, alt: A\n",
      "Warning: OEPlaceHydrogens: mol DesignUnit Components has heavy atoms ' O   HOH 231 A B 1  ' and ' O   HOH 231 B B 1  ' too close 0.11 < 0.5A\n",
      "Warning: Unable to add protons to DesignUnit: ---(AB)altA__DU__biounit\n",
      "Warning: Due to lack of protons, unable to add partial charges and radii to DesignUnit: ---(AB)altA__DU__biounit\n",
      "Processing BU # 2 with title: ---, chains AB, alt: B\n",
      "Warning: OEPlaceHydrogens: mol DesignUnit Components has heavy atoms ' O   HOH 231 A B 1  ' and ' O   HOH 231 B B 1  ' too close 0.11 < 0.5A\n",
      "Warning: Unable to add protons to DesignUnit: ---(AB)altB__DU__biounit\n",
      "Warning: Due to lack of protons, unable to add partial charges and radii to DesignUnit: ---(AB)altB__DU__biounit\n",
      "Warning: Hydrogens have not been placed, can not add OEInteractionHints\n",
      "Warning: Unable to add interactions to Design Unit: ---(AB)altA > LIG(A-1101)\n",
      "Warning: Hydrogens have not been placed, can not add OEInteractionHints\n",
      "Warning: Unable to add interactions to Design Unit: ---(AB)altA > LIG(B-1101)\n",
      "Warning: Hydrogens have not been placed, can not add OEInteractionHints\n",
      "Warning: Unable to add interactions to Design Unit: ---(AB)altB > LIG(A-1101)\n",
      "Warning: Hydrogens have not been placed, can not add OEInteractionHints\n",
      "Warning: Unable to add interactions to Design Unit: ---(AB)altB > LIG(B-1101)\n",
      "Skipping redundant DU with alts outside the site of interest, renaming existing to collapse alts\n",
      "Discarding redundant alt DU with title ---(AB)altB > LIG(A-1101)\n",
      "Skipping redundant DU with alts outside the site of interest, renaming existing to collapse alts\n",
      "Discarding redundant alt DU with title ---(AB)altB > LIG(B-1101)\n",
      "Warning: Could not find alignment that matches constraints, try again without constraints\n",
      "Superposition - RMSD: 0.00, Ref: , Fit: , SeqScore: 3102\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Design unit was successfully made for Mpro-x0072_0.pdb, output is saved to ['Mpro-x0072_0_DU_0.oedu', 'Mpro-x0072_0_DU_1.oedu'].\n",
      "The receptor design unit file has been saved as Mpro-x0072_0_DU_0_receptor.oedu\n",
      "The receptor design unit file has been saved as Mpro-x0072_0_DU_1_receptor.oedu\n",
      "cleaning LIG(B-1101)\n",
      "docking LIG(B-1101)\n",
      "cleaning LIG(B-1101)\n",
      "docking LIG(B-1101)\n",
      "The docked ligand file has been saved as ['Mpro-x0072_0_DU_0_ligand_docked.sdf', 'Mpro-x0072_0_DU_1_ligand_docked.sdf']\n",
      "Output saved as Mpro-x0072_0_receptor_fixed_Mpro-x0161_0_ligand_0.out\n",
      "\n",
      "Data saved to oe_docking_data_playground.csv\n",
      "Docked ligand Mpro-x0072_0_DU_0_ligand_docked.sdf, molecule 1: Chemgauss4 score = -5.89\n",
      "Output saved as Mpro-x0072_0_receptor_fixed_Mpro-x0161_0_ligand_0.out\n",
      "\n",
      "Data saved to oe_docking_data_playground.csv\n",
      "Docked ligand Mpro-x0072_0_DU_1_ligand_docked.sdf, molecule 1: Chemgauss4 score = -5.97\n",
      "Data saved in data/Mpro-x0161_0_ligand_with_Mpro-x0072_0_receptor_fixed\n",
      "Preparing the ligand and protein files using OpenEye Toolkits...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "mv: cannot move '/home/ian/msc_project/docking/6R8O' to 'data/Mpro-x0161_0_ligand_with_Mpro-x0072_0_receptor_fixed/6R8O': Directory not empty\n",
      "DPI: 0.12, RFree: 0.23, Resolution: 1.65\n",
      "Processing BU # 1 with title: ---, chains AB, alt: A\n",
      "Warning: OEPlaceHydrogens: mol DesignUnit Components has heavy atoms ' O   HOH 231 A B 1  ' and ' O   HOH 231 B B 1  ' too close 0.11 < 0.5A\n",
      "Warning: Unable to add protons to DesignUnit: ---(AB)altA__DU__biounit\n",
      "Warning: Due to lack of protons, unable to add partial charges and radii to DesignUnit: ---(AB)altA__DU__biounit\n",
      "Processing BU # 2 with title: ---, chains AB, alt: B\n",
      "Warning: OEPlaceHydrogens: mol DesignUnit Components has heavy atoms ' O   HOH 231 A B 1  ' and ' O   HOH 231 B B 1  ' too close 0.11 < 0.5A\n",
      "Warning: Unable to add protons to DesignUnit: ---(AB)altB__DU__biounit\n",
      "Warning: Due to lack of protons, unable to add partial charges and radii to DesignUnit: ---(AB)altB__DU__biounit\n",
      "Warning: Hydrogens have not been placed, can not add OEInteractionHints\n",
      "Warning: Unable to add interactions to Design Unit: ---(AB)altA > LIG(A-1101)\n",
      "Warning: Hydrogens have not been placed, can not add OEInteractionHints\n",
      "Warning: Unable to add interactions to Design Unit: ---(AB)altA > LIG(B-1101)\n",
      "Warning: Hydrogens have not been placed, can not add OEInteractionHints\n",
      "Warning: Unable to add interactions to Design Unit: ---(AB)altB > LIG(A-1101)\n",
      "Warning: Hydrogens have not been placed, can not add OEInteractionHints\n",
      "Warning: Unable to add interactions to Design Unit: ---(AB)altB > LIG(B-1101)\n",
      "Skipping redundant DU with alts outside the site of interest, renaming existing to collapse alts\n",
      "Discarding redundant alt DU with title ---(AB)altB > LIG(A-1101)\n",
      "Skipping redundant DU with alts outside the site of interest, renaming existing to collapse alts\n",
      "Discarding redundant alt DU with title ---(AB)altB > LIG(B-1101)\n",
      "Warning: Could not find alignment that matches constraints, try again without constraints\n",
      "Superposition - RMSD: 0.00, Ref: , Fit: , SeqScore: 3102\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Design unit was successfully made for Mpro-x0072_0.pdb, output is saved to ['Mpro-x0072_0_DU_0.oedu', 'Mpro-x0072_0_DU_1.oedu'].\n",
      "The receptor design unit file has been saved as Mpro-x0072_0_DU_0_receptor.oedu\n",
      "The receptor design unit file has been saved as Mpro-x0072_0_DU_1_receptor.oedu\n",
      "cleaning LIG(B-1101)\n",
      "docking LIG(B-1101)\n",
      "cleaning LIG(B-1101)\n",
      "docking LIG(B-1101)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 50%|     | 1/2 [02:24<02:24, 144.37s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The docked ligand file has been saved as ['Mpro-x0072_0_DU_0_ligand_docked.sdf', 'Mpro-x0072_0_DU_1_ligand_docked.sdf']\n",
      "Output saved as Mpro-x0161_0_receptor_fixed_Mpro-x0161_0_ligand_0.out\n",
      "\n",
      "Data saved to oe_docking_data_playground.csv\n",
      "Docked ligand Mpro-x0072_0_DU_0_ligand_docked.sdf, molecule 1: Chemgauss4 score = -5.89\n",
      "Output saved as Mpro-x0161_0_receptor_fixed_Mpro-x0161_0_ligand_0.out\n",
      "\n",
      "Data saved to oe_docking_data_playground.csv\n",
      "Docked ligand Mpro-x0072_0_DU_1_ligand_docked.sdf, molecule 1: Chemgauss4 score = -5.97\n",
      "Data saved in data/Mpro-x0161_0_ligand_with_Mpro-x0161_0_receptor_fixed\n",
      "Preparing the ligand and protein files using OpenEye Toolkits...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "DPI: 0.12, RFree: 0.23, Resolution: 1.65\n",
      "Processing BU # 1 with title: ---, chains AB, alt: A\n",
      "Warning: OEPlaceHydrogens: mol DesignUnit Components has heavy atoms ' O   HOH 231 A B 1  ' and ' O   HOH 231 B B 1  ' too close 0.11 < 0.5A\n",
      "Warning: Unable to add protons to DesignUnit: ---(AB)altA__DU__biounit\n",
      "Warning: Due to lack of protons, unable to add partial charges and radii to DesignUnit: ---(AB)altA__DU__biounit\n",
      "Processing BU # 2 with title: ---, chains AB, alt: B\n",
      "Warning: OEPlaceHydrogens: mol DesignUnit Components has heavy atoms ' O   HOH 231 A B 1  ' and ' O   HOH 231 B B 1  ' too close 0.11 < 0.5A\n",
      "Warning: Unable to add protons to DesignUnit: ---(AB)altB__DU__biounit\n",
      "Warning: Due to lack of protons, unable to add partial charges and radii to DesignUnit: ---(AB)altB__DU__biounit\n",
      "Warning: Hydrogens have not been placed, can not add OEInteractionHints\n",
      "Warning: Unable to add interactions to Design Unit: ---(AB)altA > LIG(A-1101)\n",
      "Warning: Hydrogens have not been placed, can not add OEInteractionHints\n",
      "Warning: Unable to add interactions to Design Unit: ---(AB)altA > LIG(B-1101)\n",
      "Warning: Hydrogens have not been placed, can not add OEInteractionHints\n",
      "Warning: Unable to add interactions to Design Unit: ---(AB)altB > LIG(A-1101)\n",
      "Warning: Hydrogens have not been placed, can not add OEInteractionHints\n",
      "Warning: Unable to add interactions to Design Unit: ---(AB)altB > LIG(B-1101)\n",
      "Skipping redundant DU with alts outside the site of interest, renaming existing to collapse alts\n",
      "Discarding redundant alt DU with title ---(AB)altB > LIG(A-1101)\n",
      "Skipping redundant DU with alts outside the site of interest, renaming existing to collapse alts\n",
      "Discarding redundant alt DU with title ---(AB)altB > LIG(B-1101)\n",
      "Warning: Could not find alignment that matches constraints, try again without constraints\n",
      "Superposition - RMSD: 0.00, Ref: , Fit: , SeqScore: 3102\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Design unit was successfully made for Mpro-x0072_0.pdb, output is saved to ['Mpro-x0072_0_DU_0.oedu', 'Mpro-x0072_0_DU_1.oedu'].\n",
      "The receptor design unit file has been saved as Mpro-x0072_0_DU_0_receptor.oedu\n",
      "The receptor design unit file has been saved as Mpro-x0072_0_DU_1_receptor.oedu\n",
      "cleaning LIG\n",
      "docking LIG\n",
      "cleaning LIG\n",
      "docking LIG\n",
      "The docked ligand file has been saved as ['Mpro-x0072_0_DU_0_ligand_docked.sdf', 'Mpro-x0072_0_DU_1_ligand_docked.sdf']\n",
      "Output saved as Mpro-x0072_0_receptor_fixed_Mpro-x0072_0_ligand_0.out\n",
      "\n",
      "Data saved to oe_docking_data_playground.csv\n",
      "Docked ligand Mpro-x0072_0_DU_0_ligand_docked.sdf, molecule 1: Chemgauss4 score = -6.91\n",
      "Output saved as Mpro-x0072_0_receptor_fixed_Mpro-x0072_0_ligand_0.out\n",
      "\n",
      "Data saved to oe_docking_data_playground.csv\n",
      "Docked ligand Mpro-x0072_0_DU_1_ligand_docked.sdf, molecule 1: Chemgauss4 score = -6.92\n",
      "Data saved in data/Mpro-x0072_0_ligand_with_Mpro-x0072_0_receptor_fixed\n",
      "Preparing the ligand and protein files using OpenEye Toolkits...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "mv: cannot move '/home/ian/msc_project/docking/6R8O' to 'data/Mpro-x0072_0_ligand_with_Mpro-x0072_0_receptor_fixed/6R8O': Directory not empty\n",
      "DPI: 0.12, RFree: 0.23, Resolution: 1.65\n",
      "Processing BU # 1 with title: ---, chains AB, alt: A\n",
      "Warning: OEPlaceHydrogens: mol DesignUnit Components has heavy atoms ' O   HOH 231 A B 1  ' and ' O   HOH 231 B B 1  ' too close 0.11 < 0.5A\n",
      "Warning: Unable to add protons to DesignUnit: ---(AB)altA__DU__biounit\n",
      "Warning: Due to lack of protons, unable to add partial charges and radii to DesignUnit: ---(AB)altA__DU__biounit\n",
      "Processing BU # 2 with title: ---, chains AB, alt: B\n",
      "Warning: OEPlaceHydrogens: mol DesignUnit Components has heavy atoms ' O   HOH 231 A B 1  ' and ' O   HOH 231 B B 1  ' too close 0.11 < 0.5A\n",
      "Warning: Unable to add protons to DesignUnit: ---(AB)altB__DU__biounit\n",
      "Warning: Due to lack of protons, unable to add partial charges and radii to DesignUnit: ---(AB)altB__DU__biounit\n",
      "Warning: Hydrogens have not been placed, can not add OEInteractionHints\n",
      "Warning: Unable to add interactions to Design Unit: ---(AB)altA > LIG(A-1101)\n",
      "Warning: Hydrogens have not been placed, can not add OEInteractionHints\n",
      "Warning: Unable to add interactions to Design Unit: ---(AB)altA > LIG(B-1101)\n",
      "Warning: Hydrogens have not been placed, can not add OEInteractionHints\n",
      "Warning: Unable to add interactions to Design Unit: ---(AB)altB > LIG(A-1101)\n",
      "Warning: Hydrogens have not been placed, can not add OEInteractionHints\n",
      "Warning: Unable to add interactions to Design Unit: ---(AB)altB > LIG(B-1101)\n",
      "Skipping redundant DU with alts outside the site of interest, renaming existing to collapse alts\n",
      "Discarding redundant alt DU with title ---(AB)altB > LIG(A-1101)\n",
      "Skipping redundant DU with alts outside the site of interest, renaming existing to collapse alts\n",
      "Discarding redundant alt DU with title ---(AB)altB > LIG(B-1101)\n",
      "Warning: Could not find alignment that matches constraints, try again without constraints\n",
      "Superposition - RMSD: 0.00, Ref: , Fit: , SeqScore: 3102\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Design unit was successfully made for Mpro-x0072_0.pdb, output is saved to ['Mpro-x0072_0_DU_0.oedu', 'Mpro-x0072_0_DU_1.oedu'].\n",
      "The receptor design unit file has been saved as Mpro-x0072_0_DU_0_receptor.oedu\n",
      "The receptor design unit file has been saved as Mpro-x0072_0_DU_1_receptor.oedu\n",
      "cleaning LIG\n",
      "docking LIG\n",
      "cleaning LIG\n",
      "docking LIG\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|| 2/2 [04:45<00:00, 142.68s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The docked ligand file has been saved as ['Mpro-x0072_0_DU_0_ligand_docked.sdf', 'Mpro-x0072_0_DU_1_ligand_docked.sdf']\n",
      "Output saved as Mpro-x0161_0_receptor_fixed_Mpro-x0072_0_ligand_0.out\n",
      "\n",
      "Data saved to oe_docking_data_playground.csv\n",
      "Docked ligand Mpro-x0072_0_DU_0_ligand_docked.sdf, molecule 1: Chemgauss4 score = -6.91\n",
      "Output saved as Mpro-x0161_0_receptor_fixed_Mpro-x0072_0_ligand_0.out\n",
      "\n",
      "Data saved to oe_docking_data_playground.csv\n",
      "Docked ligand Mpro-x0072_0_DU_1_ligand_docked.sdf, molecule 1: Chemgauss4 score = -6.92\n",
      "Data saved in data/Mpro-x0072_0_ligand_with_Mpro-x0161_0_receptor_fixed\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "ligand_files = ['Mpro-x0161_0_ligand.pdb', \n",
    "                'Mpro-x0072_0_ligand.pdb'] \n",
    "receptor_files = ['Mpro-x0072_0_receptor_fixed.pdb',\n",
    "                 'Mpro-x0161_0_receptor_fixed.pdb']\n",
    "complex_file = 'Mpro-x0072_0.pdb'\n",
    "for ligand_file in tqdm(ligand_files):\n",
    "    for receptor_file in receptor_files:\n",
    "        oe_process_lig_prot(lig_file=ligand_file, \n",
    "                 prot_file=receptor_file,\n",
    "                 complex_file = complex_file,\n",
    "                 csv_out_file='oe_docking_data_playground.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# <font color='green'>Playground</font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob\n",
    "ligand_files = glob.glob('L4001_*.sdf')\n",
    "receptor_file = 'Mpro-x0072_0_receptor_fixed.pdb'\n",
    "complex_file = 'Mpro-x0072_0.pdb'\n",
    "for ligand_file in tqdm(ligand_files):\n",
    "    oe_process_lig_prot(lig_file=ligand_file, \n",
    "                prot_file=receptor_file,\n",
    "                complex_file = complex_file,\n",
    "                csv_out_file='oe_docking_data_playground.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Number of conformers: 5\n",
      "Best Receptor pose flag: True\n",
      "posing \n",
      "Receptor used: CHYMASE(A) >  pose probability: 0.050000\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.system('python ./OpenEye/GenerateMultiplePose.py -receptor 1KLT_DU_0_receptor.oedu -numPoses 5 -in 1KLT_DU_0_receptor_L1001_0_docked.sdf -out output_poses.oedu')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Conversion complete. Saved to output_poses.sdf\n",
      "Conversion complete. Saved to 1KLT_DU_0_receptor.pdb\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning: Unable to set unknown read format (UNDEFINED) in input stream, using existing format (Canonical stereo SMILES)\n",
      "Warning: Unable to set unknown read format (UNDEFINED) in input stream, using existing format (Canonical stereo SMILES)\n"
     ]
    }
   ],
   "source": [
    "from openeye import oechem\n",
    "# TODO: THIS FUNCTION IS NOT WORKING YET!!!!\n",
    "def convert_oedu_to_sdf(input_file, output_file):\n",
    "    ifs = oechem.oemolistream()\n",
    "    if not ifs.open(input_file):\n",
    "        oechem.OEThrow.Fatal(\"Unable to open %s\" % input_file)\n",
    "    \n",
    "    ofs = oechem.oemolostream()\n",
    "    if not ofs.open(output_file):\n",
    "        oechem.OEThrow.Fatal(\"Unable to open %s\" % output_file)\n",
    "    \n",
    "    if not ifs.SetFormat(oechem.OEFormat_OEB):\n",
    "        oechem.OEThrow.Fatal(\"Unable to set input format to OEB for %s\" % input_file)\n",
    "    \n",
    "    if not ofs.SetFormat(oechem.OEFormat_SDF):\n",
    "        oechem.OEThrow.Fatal(\"Unable to set output format to SDF for %s\" % output_file)\n",
    "    \n",
    "    for mol in ifs.GetOEGraphMols():\n",
    "        oechem.OEWriteMolecule(ofs, mol)\n",
    "    \n",
    "    ifs.close()\n",
    "    ofs.close()\n",
    "    print(\"Conversion complete. Saved to %s\" % output_file)\n",
    "\n",
    "# Example usage\n",
    "# input_file = \"output_poses.oedu\"  # Ensure this file is in the correct format\n",
    "# output_file = \"output_poses.sdf\"\n",
    "# convert_oedu_to_sdf(input_file, output_file)\n",
    "\n",
    "\n",
    "# # Example usage\n",
    "# input_file = \"1KLT_DU_0_receptor.oedu\"\n",
    "# output_file = \"1KLT_DU_0_receptor.pdb\"\n",
    "# convert_oedu_to_sdf(input_file, output_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Downloading PDB structure '5s9l'...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ian/micromamba/envs/asapdiscovery/lib/python3.10/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 7409.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "# trying out my modified receptor select function to see whether water is preserved when instructed to\n",
    "pdb_id = '5s9l'\n",
    "complex_filename = download_pdb_file(pdb_id)\n",
    "structure = get_structure(complex_filename)\n",
    "save_receptor(structure, '5S9L_receptor_w_water_metal.pdb', preserve_water=True, preserve_metal=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "#TODO: fix issues with not getting peptide ligands (low priority)\n",
    "#TODO: fix issues with ligand extraction: some ligands present several conformations\n",
    "#TODO: determine whether a ligand is small molecule or peptide (low priority)\n",
    "#TODO: log-scale dG values, checking whether kd/ki values are experimentally determined or predicted (high priority)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
